Transcription factor analysis

Read in TF data
```{r}
hsa.tf <- read.table("~/storage/HumanMacaqueMouseCompAnaysis/TFs/hsa.tf.list.txt",header=TRUE)
```

Read in 1-1-1 orthologs
```{r}
humMacMouOrth.1to1.in <- readRDS("~/storage/HumanMacaqueMouseCompAnaysis/interdata/humMacMouOrth.1to1.in.Rds")
```


Read in SPG centroids
```{r}
hum.spg.cen <- read.table("~/storage/HumanMacaqueMouseCompAnaysis/interdata/SPG_centroids/HumanSPGCluster_AllGene_Centroids.txt", header = TRUE, row.names = 1)
mac.spg.cen <- read.table("~/storage/HumanMacaqueMouseCompAnaysis/interdata/SPG_centroids/MonkeySPGCluster_AllGene_Centroids.txt", header = TRUE, row.names = 1)
mou.spg.cen <- read.table("~/storage/HumanMacaqueMouseCompAnaysis/interdata/SPG_centroids/MouseSPGCluster_AllGene_Centroids.txt", header = TRUE, row.names = 1)
```


Read in Germ 20 centroids
```{r}
hum.germ.cen <- read.table("~/storage/HumanMacaqueMouseCompAnaysis/interdata/hum.20.warpped.centroids.text",skip = 1,header = FALSE,row.names = 1)
mac.germ.cen <- read.table("~/storage/HumanMacaqueMouseCompAnaysis/interdata/mac.20.warpped.centroids.text",skip = 1,header = FALSE,row.names = 1)
mou.germ.cen <- read.table("~/storage/HumanMacaqueMouseCompAnaysis/interdata/mou.20.warpped.centroids.text",skip = 1,header = FALSE,row.names = 1)
```

```{r}
hum.26.cen.1ortho <- cbind(hum.spg.cen[humMacMouOrth.1to1.in$external_gene_name.x,],hum.germ.cen[humMacMouOrth.1to1.in$external_gene_name.x,])
mac.26.cen.1ortho <- cbind(mac.spg.cen[humMacMouOrth.1to1.in$mmulatta_homolog_associated_gene_name,],mac.germ.cen[humMacMouOrth.1to1.in$mmulatta_homolog_associated_gene_name,])
mou.26.cen.1ortho <- cbind(mou.spg.cen[humMacMouOrth.1to1.in$mmusculus_homolog_associated_gene_name,],mou.germ.cen[humMacMouOrth.1to1.in$mmusculus_homolog_associated_gene_name,])
```

```{r}
tfs.ortho <- humMacMouOrth.1to1.in[humMacMouOrth.1to1.in$external_gene_name.x %in% hsa.tf$Symbol,]
hum.26.cen.tfs <- hum.26.cen.1ortho[tfs.ortho$external_gene_name.x,]
mac.26.cen.tfs <- mac.26.cen.1ortho[tfs.ortho$mmulatta_homolog_associated_gene_name,]
mou.26.cen.tfs <- mou.26.cen.1ortho[tfs.ortho$mmusculus_homolog_associated_gene_name,]
```

```{r}
mac.hum.cor <- c()
mou.hum.cor <- c()
mac.hum.tf.cor <- c()
mou.hum.tf.cor <- c()
for(i in c(1:26)){
  maccor <- cor(hum.26.cen.1ortho[,i],mac.26.cen.1ortho[,i], method = "spearman")
  mac.hum.cor <- c(mac.hum.cor, maccor)
  moucor <- cor(hum.26.cen.1ortho[,i],mou.26.cen.1ortho[,i], method = "spearman")
  mou.hum.cor <- c(mou.hum.cor, moucor)
  maccor.tf <- cor(hum.26.cen.tfs[,i], mac.26.cen.tfs[,i], method = "spearman")
  mac.hum.tf.cor <- c(mac.hum.tf.cor, maccor.tf)
  moucor.tf <- cor(hum.26.cen.tfs[,i], mou.26.cen.tfs[,i], method = "spearman")
  mou.hum.tf.cor <- c(mou.hum.tf.cor, moucor.tf)
}
```

```{r}
df <- data.frame(MacHum=mac.hum.cor, MacHumTF=mac.hum.tf.cor, MouHum=mou.hum.cor, MouHumTF=mou.hum.tf.cor)
df$stage = c(1:26)
ggplot(data = df) + 
  geom_rect(aes(xmin = -Inf, xmax = 6, ymin = -Inf, ymax = Inf),
                   fill = "#f9ecdc", alpha = 0.03)+  
  geom_rect(aes(xmin = 6, xmax = Inf, ymin = -Inf, ymax = Inf),
                   fill = "#cfe1f7", alpha = 0.03)+
  geom_point(aes(stage,MacHum),color="#5a97e3") + geom_smooth(aes(stage,MacHum),color="#5a97e3",se=FALSE)  + 
  geom_point(aes(stage,MacHumTF),color="#5a97e3",shape=2) + geom_smooth(aes(stage,MacHumTF),se=FALSE,color="#5a97e3", linetype = "dashed") + 
  geom_point(aes(stage,MouHum),color="#e35a97") + geom_smooth(aes(stage,MouHum),se=FALSE,color="#e35a97")  + 
  geom_point(aes(stage,MouHumTF),color="#e35a97",shape=2) + geom_smooth(aes(stage,MouHumTF),se=FALSE, linetype = "dashed",color="#e35a97") +
  ylab("Rank Correlation") +
  xlab("Developmental Stages") +
  scale_colour_discrete(breaks=c("Female", "Male"), labels=c("Woman", "Man"))
```

```{r fig.width=9, fig.height=5}
df <- data.frame(cor=c(mac.hum.cor, mac.hum.tf.cor,mou.hum.cor,mou.hum.tf.cor), stage=rep(c(1:26),4),Species=c(rep("Monkey",52),rep("Mouse",52)), GeneSet=rep(c(rep("All Orthologs",26),rep("TFs in Orthologs",26)),2))
ggplot(data = df) + 
  geom_rect(aes(xmin = -Inf, xmax = 6.5, ymin = -Inf, ymax = Inf),
                   fill = "#f9ecdc", alpha = 0.03)+  
  geom_rect(aes(xmin = 6.5, xmax = Inf, ymin = -Inf, ymax = Inf),
                   fill = "#cfe1f7", alpha = 0.03)+
  geom_point(aes(stage,cor,color=Species, shape=GeneSet)) + geom_smooth(aes(stage,cor,group=interaction(Species, GeneSet),color=Species,linetype=GeneSet),se=FALSE)  + 
  ylab("Rank Correlation") +
  xlab("Developmental Stages") +
  scale_shape_manual(values=c(19, 2)) +
  scale_color_manual(values=c("#5a97e3","#e3625a"))
```

Subsample the same number of genes?
```{r}
humMacMouOrth.1to1.in.sub <- humMacMouOrth.1to1.in[sample(c(1:dim(humMacMouOrth.1to1.in)[1]), 835),]
```

```{r}
hum.26.cen.1ortho.sub <- cbind(hum.spg.cen[humMacMouOrth.1to1.in.sub$external_gene_name.x,],hum.germ.cen[humMacMouOrth.1to1.in.sub$external_gene_name.x,])
mac.26.cen.1ortho.sub <- cbind(mac.spg.cen[humMacMouOrth.1to1.in.sub$mmulatta_homolog_associated_gene_name,],mac.germ.cen[humMacMouOrth.1to1.in.sub$mmulatta_homolog_associated_gene_name,])
mou.26.cen.1ortho.sub <- cbind(mou.spg.cen[humMacMouOrth.1to1.in.sub$mmusculus_homolog_associated_gene_name,],mou.germ.cen[humMacMouOrth.1to1.in.sub$mmusculus_homolog_associated_gene_name,])
```

```{r}
tfs.ortho <- humMacMouOrth.1to1.in[humMacMouOrth.1to1.in$external_gene_name.x %in% hsa.tf$Symbol,]
hum.26.cen.tfs <- cbind(hum.spg.cen[tfs.ortho$external_gene_name.x,], hum.germ.cen[tfs.ortho$external_gene_name.x, ])
mac.26.cen.tfs <- cbind(mac.spg.cen[tfs.ortho$mmulatta_homolog_associated_gene_name,],mac.germ.cen[tfs.ortho$mmulatta_homolog_associated_gene_name,])
mou.26.cen.tfs <- cbind(mou.spg.cen[tfs.ortho$mmusculus_homolog_associated_gene_name,],mou.germ.cen[tfs.ortho$mmusculus_homolog_associated_gene_name,])
```

```{r}
mac.hum.cor <- c()
mou.hum.cor <- c()
mac.hum.tf.cor <- c()
mou.hum.tf.cor <- c()
for(i in c(1:26)){
  maccor <- cor(hum.26.cen.1ortho[,i],mac.26.cen.1ortho[,i], method = "spearman")
  mac.hum.cor <- c(mac.hum.cor, maccor)
  moucor <- cor(hum.26.cen.1ortho[,i],mou.26.cen.1ortho[,i], method = "spearman")
  mou.hum.cor <- c(mou.hum.cor, moucor)
  maccor.tf <- cor(hum.26.cen.tfs[,i], mac.26.cen.tfs[,i], method = "spearman")
  mac.hum.tf.cor <- c(mac.hum.tf.cor, maccor.tf)
  moucor.tf <- cor(hum.26.cen.tfs[,i], mou.26.cen.tfs[,i], method = "spearman")
  mou.hum.tf.cor <- c(mou.hum.tf.cor, moucor.tf)
}
```


```{r fig.width=9, fig.height=5}
df <- data.frame(cor=c(mac.hum.cor, mac.hum.tf.cor,mou.hum.cor,mou.hum.tf.cor), stage=rep(c(1:26),4),Species=c(rep("Monkey",52),rep("Mouse",52)), GeneSet=rep(c(rep("All Orthologs",26),rep("TFs in Orthologs",26)),2))
ggplot(data = df) + 
  geom_rect(aes(xmin = -Inf, xmax = 6.5, ymin = -Inf, ymax = Inf),
                   fill = "#f9ecdc", alpha = 0.03)+  
  geom_rect(aes(xmin = 6.5, xmax = Inf, ymin = -Inf, ymax = Inf),
                   fill = "#cfe1f7", alpha = 0.03)+
  geom_point(aes(stage,cor,color=Species, shape=GeneSet)) + geom_smooth(aes(stage,cor,group=interaction(Species, GeneSet),color=Species,linetype=GeneSet),se=FALSE)  + 
  ylab("Rank Correlation") +
  xlab("Developmental Stages") +
  scale_shape_manual(values=c(19, 2)) +
  scale_color_manual(values=c("#5a97e3","#e3625a"))
```

Mark TFs for Orthologs in three species connecting plots
```{r}
germ.connec <- read.table("~/storage/HumanMacaqueMouseCompAnaysis/interdata/Germ_kmeans/GermOrthologConnection.txt",header = TRUE)
germ.connec$TF <- germ.connec$humgen %in% hsa.tf$Symbol
write.table(germ.connec, file = "~/storage/HumanMacaqueMouseCompAnaysis/interdata/Germ_kmeans/GermOrthologConnection_withTF.txt", sep = "\t", quote = FALSE, col.names = TRUE, row.names = FALSE)
```

```{r}
spg.connect <- read.table("~/storage/HumanMacaqueMouseCompAnaysis/interdata/SPG_GeneKmeans/OrthologConnection.txt",header = TRUE)
spg.connect$TF <- spg.connect$humclu %in% hsa.tf$Symbol
write.table(spg.connect, file = "~/storage/HumanMacaqueMouseCompAnaysis/interdata/SPG_GeneKmeans/OrthologConnection_withTF.txt", sep = "\t", quote = FALSE, col.names = TRUE, row.names = FALSE)
```

