# CCA alignment and clustering for SPGs from three species

```{r}
humMacMouOrth.1to1.in <- readRDS("../interdata/humMacMouOrth.1to1.in.Rds")
dge.hum.spg.raw <- readRDS("~/storage/HumanMacaqueMouseCompAnaysis/interdata/dge.hum.spg.raw.rds")
dge.mac.spg.raw <- readRDS("~/storage/HumanMacaqueMouseCompAnaysis/interdata/dge.mac.spg.raw.rds")
dge.mou.spg.raw <- readRDS("~/storage/HumanMacaqueMouseCompAnaysis/interdata/dge.mou.spg.raw.rds")
library(Seurat)
library(seriation)
library(xlsx)
```

mouse seurat object
```{r}
dge.mou.spg.raw.sorted.named <- dge.mou.spg.raw[humMacMouOrth.1to1.in$mmusculus_homolog_associated_gene_name,]
rownames(dge.mou.spg.raw.sorted.named) = humMacMouOrth.1to1.in$external_gene_name.x
dge.mou.spg <- CreateSeuratObject(raw.data = dge.mou.spg.raw.sorted.named)
dge.mou.spg <- NormalizeData(object = dge.mou.spg, normalization.method = "LogNormalize", 
    scale.factor = 10000)
dge.mou.spg <- ScaleData(object = dge.mou.spg,vars.to.regress = c("nUMI"))
dge.mou.spg <- FindVariableGenes(object = dge.mou.spg,mean.function = ExpMean, dispersion.function = LogVMR, x.low.cutoff = 0.0125, y.cutoff = 0)
```

##rename macaque
```{r}
dge.mac.spg.raw.renamed <- dge.mac.spg.raw
rownames(dge.mac.spg.raw.renamed) <- rownames(dge.hum.spg.raw)
dge.mac.spg.renamed <- CreateSeuratObject(raw.data = dge.mac.spg.raw.renamed)
dge.mac.spg.renamed <- NormalizeData(object = dge.mac.spg.renamed, normalization.method = "LogNormalize", scale.factor = 10000)
dge.mac.spg.renamed <- FindVariableGenes(object = dge.mac.spg.renamed,mean.function = ExpMean, dispersion.function = LogVMR, x.low.cutoff = 0.0125, y.cutoff = 0)
dge.mac.spg.renamed <- ScaleData(object = dge.mac.spg.renamed,vars.to.regress = c("nUMI"))
```

```{r}
dge.hum.spg <- CreateSeuratObject(raw.data = dge.hum.spg.raw)
dge.hum.spg <- NormalizeData(object = dge.hum.spg, normalization.method = "LogNormalize", 
    scale.factor = 10000)
dge.hum.spg <- FindVariableGenes(object = dge.hum.spg,mean.function = ExpMean, dispersion.function = LogVMR, x.low.cutoff = 0.0125, y.cutoff = 0)
dge.hum.spg <- ScaleData(object = dge.hum.spg,vars.to.regress = c("nUMI"))
```

Highly variable genes
```{r}
hvg.spg.1 <- rownames(x = head(x = dge.hum.spg@hvg.info, n = 2000))
hvg.spg.2 <- rownames(x = head(x = dge.mac.spg.renamed@hvg.info, n = 2000))
hvg.spg.3 <- rownames(x = head(x = dge.mou.spg@hvg.info, n = 2000))
hvg.union <- union(x = hvg.spg.1, y = hvg.spg.2)
hvg.union <- union(x = hvg.union, y = hvg.spg.3)
```

```{r}
dge.hum.spg@meta.data[,"species"] <- "Human"
dge.mac.spg.renamed@meta.data[,"species"] <- "Monkey"
dge.mou.spg@meta.data[,"species"] <- "Mouse"
```

```{r}
spg.cca <- RunMultiCCA(object.list = list(dge.hum.spg, dge.mac.spg.renamed, dge.mou.spg),genes.use = hvg.union, num.ccs = 15, niter = 20)
```


```{r}
MetageneBicorPlot(spg.cca, grouping.var = "species", dims.eval = 1:15, display.progress = TRUE)
```


```{r}
#spg.cca <- CalcVarExpRatio(object = spg.cca, reduction.type = "pca", grouping.var = "species", dims.use = 1:8)
spg.cca <- AlignSubspace(object = spg.cca, reduction.type = "cca", grouping.var = "species", dims.align = 1:8)

spg.cca <- RunPCA(object = spg.cca, reduction.use = "cca.aligned", dims.use = 1:8)
spg.cca <- RunTSNE(object = spg.cca, reduction.use = "cca.aligned", dims.use = 1:8)
spg.cca <- RunUMAP(object = spg.cca, reduction.use = "cca.aligned", dims.use = 1:8)
```

```{r fig.width=5, fig.height=4}
TSNEPlot(object = spg.cca, do.return = TRUE, pt.size = 1, group.by = "species")
DimPlot(object = spg.cca, reduction.use = "cca.aligned", pt.size = 1, group.by = "species")
DimPlot(object = spg.cca, reduction.use = "umap", pt.size = 1, group.by = "species")
```

```{r}
spg.cca <- FindClusters(object = spg.cca, reduction.type = "cca.aligned", dims.use = 1:8, save.SNN = TRUE, resolution = 0.6)
```

```{r fig.width=5, fig.height=4}
TSNEPlot(object = spg.cca, do.return = TRUE, pt.size = 1, group.by = "res.0.6")
DimPlot(object = spg.cca, reduction.use = "umap", pt.size = 1, group.by = "res.0.6")
DimPlot(object = spg.cca, reduction.use = "cca.aligned", pt.size = 1, group.by = "res.0.6")
```

change the label of the clusters
```{r}
spg.cca@meta.data$res.0.6.ord <- plyr::mapvalues(spg.cca@meta.data$res.0.6, c(4,2,0,1,5,3,7,6),c(1,2,3,4,5,6,7,8))
```


Cluster centroid for SPG
```{r}
spg.cen.0.6 <- c()
clus <- unique(spg.cca@meta.data$res.0.6.ord)
for(i in clus){
  cen <- apply(spg.cca@data[,spg.cca@meta.data$res.0.6.ord == i], 1, mean)
  spg.cen.0.6 <- cbind(spg.cen.0.6, cen)
}
colnames(spg.cen.0.6) <- clus
```


```{r fig.width=6,fig.height=5}
#library(ComplexHeatmap)
spg.cor.0.6 <- cor(spg.cen.0.6, method="spearman")
a = seriate(spg.cor.0.6,method="BEA_TSP")
ord = get_order(a)
Heatmap(spg.cor.0.6[ord,ord], cluster_rows = FALSE, cluster_columns = FALSE)
```

```{r fig.width=5, fig.height=4}
cols <- c("#477900",
"#9378ff",
"#01ce74",
"#ff2b85",
"#b0d266",
"#733697",
"#ff9c1d",
"#0190c8",
"#ffa19e")
TSNEPlot(object = spg.cca, do.return = TRUE, pt.size = 1, group.by = "res.0.6.ord", colors.use = cols)
DimPlot(object = spg.cca, reduction.use = "umap", pt.size = 1, group.by = "res.0.6.ord", cols.use = cols)
DimPlot(object = spg.cca, reduction.use = "cca.aligned", pt.size = 1, group.by = "res.0.6.ord", cols.use = cols)
```
Save spg.cca object
```{r}
saveRDS(spg.cca, file = "~/storage/HumanMacaqueMouseCompAnaysis/interdata/spg.cca.rds")
```



##Read in mouse germ cell 12 clusters
```{r}
germ12clu <- read.table("~/storage/HumanMacaqueMouseCompAnaysis/mouse/MouseAdultST24_12GermClusters_ClusterID.txt")
mouse4clus <- read.table("~/storage/HumanMacaqueMouseCompAnaysis/ClusterID_SPGcellsUMI1k_4clusters.txt",header = FALSE)
```

```{r}
spg.cca@meta.data$mouseclus <- 0
spg.cca@meta.data$mouseclus[spg.cca@cell.names %in% mouse4clus$V1[mouse4clus$V2==1]] = "1a"
spg.cca@meta.data$mouseclus[spg.cca@cell.names %in% mouse4clus$V1[mouse4clus$V2==2]] = "1b"
spg.cca@meta.data$mouseclus[spg.cca@cell.names %in% mouse4clus$V1[mouse4clus$V2==3]] = "1c"
spg.cca@meta.data$mouseclus[spg.cca@cell.names %in% mouse4clus$V1[mouse4clus$V2==4]] = "1d"
spg.cca@meta.data$mouseclus[spg.cca@cell.names %in% rownames(germ12clu)[germ12clu$x == 2]] = 2
spg.cca@meta.data$mouseclus[spg.cca@cell.names %in% rownames(germ12clu)[germ12clu$x == 3]] = 3
```

```{r fig.width=6, fig.height=5}
DimPlot(object = spg.cca, reduction.use = "tsne", pt.size = 1, group.by = "mouseclus", cols.use = c("grey","#49adad","#cb5b42","#67a64e","#8d70c9","#b69340","#c8588c"))
```

```{r}
spg.cca <- SetIdent(spg.cca, ident.use = spg.cca@meta.data$res.0.6.ord)
```


##. Find Markers
```{r}
markers.9clus <- list()
for(i in c(1:8)){
  markers.9clus[[i]] <- FindConservedMarkers(spg.cca, ident.1 = i, grouping.var = "species",min.cells.group=2)
}
```

Write to excel
```{r}
for(i in c(1:8)){
if(i==1){
  write.xlsx2(x = markers.9clus[[i]], file = "/home/xianingz/storage/HumanMacaqueMouseCompAnaysis/interdata/SPG_markers/SPG_8Clusters_Markers.xlsx", sheetName = paste0("Cluster", i), col.names = TRUE, row.names = TRUE, append = FALSE)
}else{
  write.xlsx2(x = markers.9clus[[i]], file = "/home/xianingz/storage/HumanMacaqueMouseCompAnaysis/interdata/SPG_markers/SPG_8Clusters_Markers.xlsx", sheetName = paste0("Cluster", i), col.names = TRUE, row.names = TRUE, append = TRUE)
}
}
```

```{r}
markers.sep.9clus <- list()
for(i in c(1:8)){
  markers.sep.9clus[[i]] <- FindMarkers(spg.cca, ident.1 = i, grouping.var = "species")
}
```

```{r}
for(i in c(1:8)){
if(i==1){
  write.xlsx2(x = markers.sep.9clus[[i]], file = "/home/xianingz/storage/HumanMacaqueMouseCompAnaysis/interdata/SPG_markers/SPG_8Clusters_Markers_all.xlsx", sheetName = paste0("Cluster", i), col.names = TRUE, row.names = TRUE, append = FALSE)
}else{
  write.xlsx2(x = markers.sep.9clus[[i]], file = "/home/xianingz/storage/HumanMacaqueMouseCompAnaysis/interdata/SPG_markers/SPG_8Clusters_Markers_all.xlsx", sheetName = paste0("Cluster", i), col.names = TRUE, row.names = TRUE, append = TRUE)
}
}
```

##. SPG marker Plots
```{r fig.width=10,fig.height=10}
spg.markers <- c("PIWIL4", "TSPAN33", "MORC1","ID4","FGFR3","DMRT1","DMRTB1","SOLH2H","SYCP3","SYCP1","SPO11","MEIOB","ACRV1","CATSPER1","CATSPER3","PRM1","PRM2","TNP1","TNP2","ZCWPW1","STRA8","GFRA1")
spg.markers <- intersect(spg.markers, rownames(spg.cca@raw.data))
FeaturePlot(object = spg.cca,features.plot = spg.markers,reduction.use = "tsne",pt.size = 0.5,nCol = 4)
VlnPlot(object = spg.cca, features.plot = spg.markers,nCol = 4,group.by = "res.0.6.ord",point.size.use=0)
```
##. SPG marker RA receptor 
```{r fig.width=10,fig.height=2.5}
spg.markers <- c("RARA","RARB","RARG")
spg.markers <- intersect(spg.markers, rownames(spg.cca@raw.data))
FeaturePlot(object = spg.cca,features.plot = spg.markers,reduction.use = "tsne",pt.size = 0.5,nCol = 4)
VlnPlot(object = spg.cca, features.plot = spg.markers,nCol = 3,group.by = "res.0.6.ord",point.size.use=0)
```
##. Cluster Centroids

###. calculate SPG cluster centroid for each species
```{r}
##Cluster 9 removed because of doublets.
spg.exp.all.data <- exp(spg.cca@data) - 1
hum.spg.cen <- matrix(0,nrow=11023, ncol=8)
mac.spg.cen <- matrix(0,nrow=11023, ncol=8)
mou.spg.cen <- matrix(0,nrow=11023, ncol=8)
for(i in c(1:8)){
  hum.cells <- spg.cca@cell.names[spg.cca@meta.data$res.0.6.ord == i & spg.cca@meta.data$species == "Human"]
  mac.cells <- spg.cca@cell.names[spg.cca@meta.data$res.0.6.ord == i & spg.cca@meta.data$species == "Monkey"]
  mou.cells <- spg.cca@cell.names[spg.cca@meta.data$res.0.6.ord == i & spg.cca@meta.data$species == "Mouse"]
  hum.spg.cen[,i] <- apply(spg.exp.all.data[,hum.cells], 1, mean)
  mac.spg.cen[,i] <- apply(spg.exp.all.data[,mac.cells], 1, mean)
  mou.spg.cen[,i] <- apply(spg.exp.all.data[,mou.cells], 1, mean)
}
sum(humMacMouOrth.1to1.in$external_gene_name.x != rownames(spg.cca@data)) ##confirm gene names are ordered the same.
rownames(hum.spg.cen) <- rownames(spg.cca@data)
rownames(mac.spg.cen) <- humMacMouOrth.1to1.in$mmulatta_homolog_associated_gene_name
rownames(mou.spg.cen) <- humMacMouOrth.1to1.in$mmusculus_homolog_associated_gene_name
saveRDS(hum.spg.cen, file = "~/storage/HumanMacaqueMouseCompAnaysis/interdata/hum.spg.cen.rds")
saveRDS(mac.spg.cen, file = "~/storage/HumanMacaqueMouseCompAnaysis/interdata/mac.spg.cen.rds")
saveRDS(mou.spg.cen, file = "~/storage/HumanMacaqueMouseCompAnaysis/interdata/mou.spg.cen.rds")
```

###. Write to xlxs
```{r}
write.xlsx2(hum.spg.cen[,c(1:6)], file = "/home/xianingz/storage/HumanMacaqueMouseCompAnaysis/interdata/SPG_centroids/HumanSPGClusterCentroids.xlsx", sheetName = "Human", append = FALSE, row.names = TRUE)
write.xlsx2(mac.spg.cen[,c(1:6)], file = "/home/xianingz/storage/HumanMacaqueMouseCompAnaysis/interdata/SPG_centroids/MonkeySPGClusterCentroids.xlsx", sheetName = "Monkey", append = FALSE, row.names = TRUE)
write.xlsx2(mou.spg.cen[,c(1:6)], file = "/home/xianingz/storage/HumanMacaqueMouseCompAnaysis/interdata/SPG_centroids/MouseSPGClusterCentroids.xlsx", sheetName = "Mouse", append = FALSE, row.names = TRUE)
```

Centroids for all genes for three species
extract full mouse SPG dataset
```{r}
mou.raw <- read.table("~/storage/HumanMacaqueMouseCompAnaysis/countsOriginalData/mouse_data/xaa", header = TRUE,row.names = 1,sep="\t")
#For all SPG cells
spg.cellnames <- spg.cca@cell.names[spg.cca@meta.data$species=="Mouse"]
selectcells <- colnames(mou.raw) %in% spg.cellnames
dge.mou.spg.full.raw <- mou.raw[,selectcells]
filelist <- paste0("~/storage/HumanMacaqueMouseCompAnaysis/countsOriginalData/mouse_data/", list.files("~/storage/HumanMacaqueMouseCompAnaysis/countsOriginalData/mouse_data"))
for(i in filelist[2:length(filelist)]){
print(i)
mou.raw <- read.table(i, row.names = 1, sep="\t",header = FALSE)
mou.raw <- mou.raw[,selectcells]
colnames(mou.raw) <- colnames(dge.mou.spg.full.raw)
dge.mou.spg.full.raw <- rbind(dge.mou.spg.full.raw, mou.raw)
}
rm(mou.raw)
saveRDS(dge.mou.spg.full.raw, file = "~/storage/HumanMacaqueMouseCompAnaysis/interdata/dge.mou.spg.full.raw.rds")
```

```{r}
spg.cca <- readRDS("~/storage/HumanMacaqueMouseCompAnaysis/interdata/spg.cca.rds")
dge.hum.spg.full.raw <- readRDS("~/storage/HumanMacaqueMouseCompAnaysis/interdata/dge.hum.spg.full.raw.rds")
dge.mac.spg.full.raw <- readRDS("~/storage/HumanMacaqueMouseCompAnaysis/interdata/dge.mac.spg.full.raw.rds")
```

```{r}
##Cluster centroid for all genes.
dge.mac.spg.full.norm <- apply(dge.mac.spg.full.raw,2,function(x) x/sum(x)*10000)
dge.hum.spg.full.norm <- apply(dge.hum.spg.full.raw,2,function(x) x/sum(x)*10000)
dge.mou.spg.full.norm <- apply(dge.mou.spg.full.raw,2,function(x) x/sum(x)*10000)
hum.spg.all.cen <- matrix(0,nrow=dim(dge.hum.spg.full.norm)[1], ncol=8)
mac.spg.all.cen <- matrix(0,nrow=dim(dge.mac.spg.full.norm)[1], ncol=8)
mou.spg.all.cen <- matrix(0,nrow=dim(dge.mou.spg.full.norm)[1], ncol=8)
#mou.spg.all.cen <- matrix(0,nrow=11029, ncol=8)
for(i in c(1:8)){
  hum.cells <- spg.cca@cell.names[spg.cca@meta.data$res.0.6.ord == i & spg.cca@meta.data$species == "Human"]
  mac.cells <- spg.cca@cell.names[spg.cca@meta.data$res.0.6.ord == i & spg.cca@meta.data$species == "Monkey"]
  mou.cells <- spg.cca@cell.names[spg.cca@meta.data$res.0.6.ord == i & spg.cca@meta.data$species == "Mouse"]
  hum.spg.all.cen[,i] <- apply(dge.hum.spg.full.norm[,hum.cells], 1, mean)
  mac.spg.all.cen[,i] <- apply(dge.mac.spg.full.norm[,mac.cells], 1, mean)
  mou.spg.all.cen[,i] <- apply(dge.mou.spg.full.norm[,mou.cells], 1, mean)
}
rownames(hum.spg.all.cen) <- rownames(dge.hum.spg.full.raw)
rownames(mac.spg.all.cen) <- rownames(dge.mac.spg.full.raw)
rownames(mou.spg.all.cen) <- rownames(dge.mou.spg.full.raw)
saveRDS(hum.spg.all.cen, file = "~/storage/HumanMacaqueMouseCompAnaysis/interdata//hum.spg.all.cen.rds")
saveRDS(mac.spg.all.cen, file = "~/storage/HumanMacaqueMouseCompAnaysis/interdata//mac.spg.all.cen.rds")
saveRDS(mou.spg.all.cen, file = "~/storage/HumanMacaqueMouseCompAnaysis/interdata//mou.spg.all.cen.rds")
```

##Write to xlxs
```{r}
write.xlsx2(hum.spg.all.cen[,c(1:6)], file = "/home/xianingz/storage/HumanMacaqueMouseCompAnaysis/interdata/SPG_centroids/HumanSPGCluster_AllGene_Centroids.xlsx", sheetName = "Human", append = FALSE, row.names = TRUE)
write.xlsx2(mac.spg.all.cen[,c(1:6)], file = "/home/xianingz/storage/HumanMacaqueMouseCompAnaysis/interdata/SPG_centroids/MonkeySPGCluster_AllGene_Centroids.xlsx", sheetName = "Monkey", append = FALSE, row.names = TRUE)
write.xlsx2(mou.spg.all.cen[,c(1:6)], file = "/home/xianingz/storage/HumanMacaqueMouseCompAnaysis/interdata/SPG_centroids/MouseSPGCluster_AllGene_Centroids.xlsx", sheetName = "Mouse", append = FALSE, row.names = TRUE)
```

##Write to txt
```{r}
write.table(hum.spg.all.cen[,c(1:6)], file = "/home/xianingz/storage/HumanMacaqueMouseCompAnaysis/interdata/SPG_centroids/HumanSPGCluster_AllGene_Centroids.txt", quote = FALSE, row.names = TRUE, sep="\t")
write.table(mac.spg.all.cen[,c(1:6)], file = "/home/xianingz/storage/HumanMacaqueMouseCompAnaysis/interdata/SPG_centroids/MonkeySPGCluster_AllGene_Centroids.txt", quote = FALSE, row.names = TRUE, sep="\t")
write.table(mou.spg.all.cen[,c(1:6)], file = "/home/xianingz/storage/HumanMacaqueMouseCompAnaysis/interdata/SPG_centroids/MouseSPGCluster_AllGene_Centroids.txt", quote = FALSE, row.names = TRUE, sep="\t")
```

Centroid rank correlation for three species
```{r fig.width=5, fig.height=4.5}
hum.spg.cor <- cor(hum.spg.all.cen[,c(1:6)], method="spearman")
Heatmap(hum.spg.cor, cluster_rows = FALSE, cluster_columns = FALSE)
```
```{r fig.width=5, fig.height=4.5}
mac.spg.cor <- cor(mac.spg.all.cen[,c(1:6)], method="spearman")
Heatmap(mac.spg.cor, cluster_rows = FALSE, cluster_columns = FALSE)
```
```{r fig.width=5, fig.height=4.5}
mou.spg.cor <- cor(mou.spg.all.cen[,c(1:6)], method="spearman")
Heatmap(mou.spg.cor, cluster_rows = FALSE, cluster_columns = FALSE)
```

##. Plot all the clusters in three species
```{r fig.width=20, fig.height=9}
plist <- list()
for(i in c(1:8)){
  spg.cca@meta.data$var <- (spg.cca@meta.data$res.0.6.ord==i & spg.cca@meta.data$species=="Human")*i
  p <- TSNEPlot(object = spg.cca, do.return = TRUE, pt.size = 1, group.by = "var", colors.use = c("grey",cols[i]), plot.order = c(i,"0"))
  plist[[i]] = p
}
do.call("grid.arrange", c(plist, ncol=4))
```
```{r fig.width=20, fig.height=9}
plist <- list()
for(i in c(1:8)){
  spg.cca@meta.data$var <- (spg.cca@meta.data$res.0.6.ord==i & spg.cca@meta.data$species=="Monkey")*i
  p <- TSNEPlot(object = spg.cca, do.return = TRUE, pt.size = 1, group.by = "var", colors.use = c("grey",cols[i]), plot.order = c(i,"0"))
  plist[[i]] = p
}
do.call("grid.arrange", c(plist, ncol=4))
```




```{r fig.width=20, fig.height=9}
plist <- list()
for(i in c(1:8)){
  spg.cca@meta.data$var <- (spg.cca@meta.data$res.0.6.ord==i & spg.cca@meta.data$species=="Mouse")*i
  p <- TSNEPlot(object = spg.cca, do.return = TRUE, pt.size = 1, group.by = "var", colors.use = c("grey",cols[i]), plot.order = c(i,"0"))
  plist[[i]] = p
}
do.call("grid.arrange", c(plist, ncol=4))
```

##. Markers for clusters 1 and 2 in mouse
```{r}
spg.cca@meta.data$speclus <- paste0(spg.cca@meta.data$species,"_", spg.cca@meta.data$res.0.6.ord)
spg.cca <- SetAllIdent(spg.cca, id = "speclus")
marker1over2 <- FindMarkers(spg.cca, ident.1 = "Mouse_1", ident.2 = "Mouse_2", logfc.threshold = 0.6, min.diff.pct = 0.1)
```

```{r}
rownames(humMacMouOrth.1to1.in) = humMacMouOrth.1to1.in$external_gene_name.x
rownames(marker1over2) <- humMacMouOrth.1to1.in[rownames(marker1over2),]$mmusculus_homolog_associated_gene_name
```

```{r}
write.table(marker1over2, file = "~/storage/HumanMacaqueMouseCompAnaysis/interdata/MakersMouseClust1vs2.txt",quote = FALSE, row.names = TRUE, col.names = TRUE, sep = "\t")
```

##Call markers using all genes for each species
###Human

spg.raw file used below are files that removed cluster 7-8 in SPG CCA clusterign 
```{r}
dge.hum.spg <- CreateSeuratObject(raw.data = dge.hum.spg.raw)
dge.hum.spg <- NormalizeData(object = dge.hum.spg, normalization.method = "LogNormalize", 
    scale.factor = 10000)
dge.hum.spg <- ScaleData(object = dge.hum.spg,vars.to.regress = c("nUMI"))
```

```{r}
dge.hum.spg@meta.data$clus <- spg.cca@meta.data[dge.hum.spg@cell.names,]$res.0.6.ord
dge.hum.spg <- SetIdent(dge.hum.spg, ident.use = dge.hum.spg@meta.data$clus)
```

```{r}
hum.spg.markers <- list()
for(i in c(1:6)){
  hum.spg.markers[[i]] <- FindMarkers(dge.hum.spg, ident.1 = i, logfc.threshold = 0.25, min.pct = 0.1, only.pos = TRUE)
}
```
```{r}
for(i in c(1:6)){
if(i==1){
  write.xlsx2(x = hum.spg.markers[[i]], file = "/home/xianingz/storage/HumanMacaqueMouseCompAnaysis/interdata/SPG_markers/SPG_hum_6Clusters_Markers_allGene_lgfc0.25pct10.xlsx", sheetName = paste0("Cluster", i), col.names = TRUE, row.names = TRUE, append = FALSE)
}else{
  write.xlsx2(x = hum.spg.markers[[i]], file = "/home/xianingz/storage/HumanMacaqueMouseCompAnaysis/interdata/SPG_markers/SPG_hum_6Clusters_Markers_allGene_lgfc0.25pct10.xlsx", sheetName = paste0("Cluster", i), col.names = TRUE, row.names = TRUE, append = TRUE)
}
}
```

```{r}
dge.mac.spg <- CreateSeuratObject(raw.data = dge.mac.spg.raw)
dge.mac.spg <- NormalizeData(object = dge.mac.spg, normalization.method = "LogNormalize", 
    scale.factor = 10000)
dge.mac.spg <- ScaleData(object = dge.mac.spg,vars.to.regress = c("nUMI"))

dge.mac.spg@meta.data$clus <- spg.cca@meta.data[dge.mac.spg@cell.names,]$res.0.6.ord
dge.mac.spg <- SetIdent(dge.mac.spg, ident.use = dge.mac.spg@meta.data$clus)

mac.spg.markers <- list()
for(i in c(1:6)){
  mac.spg.markers[[i]] <- FindMarkers(dge.mac.spg, ident.1 = i, logfc.threshold = 0.25, min.pct = 0.1, only.pos = TRUE)
}

for(i in c(1:6)){
if(i==1){
  write.xlsx2(x = mac.spg.markers[[i]], file = "/home/xianingz/storage/HumanMacaqueMouseCompAnaysis/interdata/SPG_markers/SPG_mac_6Clusters_Markers_allGene_lgfc0.25pct10.xlsx", sheetName = paste0("Cluster", i), col.names = TRUE, row.names = TRUE, append = FALSE)
}else{
  write.xlsx2(x = mac.spg.markers[[i]], file = "/home/xianingz/storage/HumanMacaqueMouseCompAnaysis/interdata/SPG_markers/SPG_mac_6Clusters_Markers_allGene_lgfc0.25pct10.xlsx", sheetName = paste0("Cluster", i), col.names = TRUE, row.names = TRUE, append = TRUE)
}
}
```

```{r}
dge.mou.spg <- CreateSeuratObject(raw.data = dge.mou.spg.raw)
dge.mou.spg <- NormalizeData(object = dge.mou.spg, normalization.method = "LogNormalize", 
    scale.factor = 10000)
dge.mou.spg <- ScaleData(object = dge.mou.spg,vars.to.regress = c("nUMI"))

dge.mou.spg@meta.data$clus <- spg.cca@meta.data[dge.mou.spg@cell.names,]$res.0.6.ord
dge.mou.spg <- SetIdent(dge.mou.spg, ident.use = dge.mou.spg@meta.data$clus)

mou.spg.markers <- list()
for(i in c(1:6)){
  mou.spg.markers[[i]] <- FindMarkers(dge.mou.spg, ident.1 = i, logfc.threshold = 0.25, min.pct = 0.1, only.pos = TRUE)
}

for(i in c(1:6)){
if(i==1){
  write.xlsx2(x = mou.spg.markers[[i]], file = "/home/xianingz/storage/HumanMacaqueMouseCompAnaysis/interdata/SPG_markers/SPG_mou_6Clusters_Markers_allGene_lgfc0.25pct10.xlsx", sheetName = paste0("Cluster", i), col.names = TRUE, row.names = TRUE, append = FALSE)
}else{
  write.xlsx2(x = mou.spg.markers[[i]], file = "/home/xianingz/storage/HumanMacaqueMouseCompAnaysis/interdata/SPG_markers/SPG_mou_6Clusters_Markers_allGene_lgfc0.25pct10.xlsx", sheetName = paste0("Cluster", i), col.names = TRUE, row.names = TRUE, append = TRUE)
}
}
```

Plot markers
```{r}
markers <- c("THY1", "SALL4", "PLZF", "ZBTB16", "ID4", "GFRA1", "ITGA6", "EPCAM", "PAX7", "BMI1", "EOMES", "UTF1", "UCHL1", "LIN28A", "LIN28B", "PIWIL4", "MORC1", "CDK17", "TSPAN33", "MAGEB2", "ITGA6", "MKI67", "L1TD1", "ID1", "ID3", "STRA8", "GPC3", "GPC4", "GPX1", "TCF3", "FMR1", "DNAJB6", "LENG8", "C19orf84", "ZNF428", "PSIP1", "SSFA2", "APOE", "CRYM")
```


```{r}
humMacMouOrth.1to1.in <- readRDS("~/storage/HumanMacaqueMouseCompAnaysis/interdata/humMacMouOrth.1to1.in.Rds")
spg.cca <- readRDS("~/storage/HumanMacaqueMouseCompAnaysis/interdata/spg.cca.rds")
```

```{r}
redblue100.alpha<-rgb(read.table("~/storage/HumanMacaqueMouseCompAnaysis/redblue100.txt",sep='\t',row.names=1,header=T),alpha=0.8)

set.ifnull=function(x,y) {
  if(is.null(x)) return(y)
  return(x)
}
translate.dim.code=function(reduction.use) {
  if (reduction.use=="pca") return.code="PC"
  if (reduction.use=="ica") return.code="IC"
  if (reduction.use=="tsne") return.code="tSNE_"
  if (reduction.use=="mds") return.code="MDS"
  if (reduction.use=="umap") return.code="UMAP"
  return(return.code)
}

plotgene <- function(gene){
feature=features.plot=gene

dim.1 = 1; dim.2 = 2; cells.use = NULL; pt.size = 1;
pch.use = 16; reduction.use = "tsne";
use.imputed = FALSE; no.axes = TRUE; no.legend = FALSE

cells.use <- set.ifnull(cells.use, spg.cca@cell.names[spg.cca@meta.data$res.0.6.ord %in% c(1:6)])
dim.code <- translate.dim.code(reduction.use)
dim.codes <- paste(dim.code, c(dim.1, dim.2), sep = "")
data.plot <- as.data.frame(spg.cca@dr$tsne@cell.embeddings[cells.use,])

x1 <- paste(dim.code, dim.1, sep = "")
x2 <- paste(dim.code, dim.2, sep = "")

data.plot$x <- data.plot[cells.use, x1]
data.plot$y <- data.plot[cells.use, x2]
data.plot$pt.size <- pt.size
data.use <- data.frame(t(spg.cca@data[feature,cells.use]))
rownames(data.use)=feature

data.gene0 <- na.omit(data.frame(data.use[feature, ]))
data.plot$gene <- t(data.gene0)

st6<- data.plot
st6<-st6[order(st6[,6]),]
z<-st6[,6]

jpeg(paste0("~/storage/HumanMacaqueMouseCompAnaysis/interdata/SPG_markers/plots/spg.cca","_",feature,"_40-100redblue.jpeg"),height=850,width=800,res=300)
par(mar=c(0.5,0.5,2,0.5),mgp=c(1,0.5,0))
z<-st6[,6]
zcolor <- redblue100.alpha[40:100][(z - min(z))/diff(range(z))*60 + 1]
plot(st6[,1],st6[,2], col=zcolor,pch=19,cex=0.3,cex.main=1.5,axes=F,main=features.plot,xlab="",ylab="")
dev.off()
}
```

```{r}
for(i in intersect(markers, humMacMouOrth.1to1.in$external_gene_name.x)){
  plotgene(i)
}
```

```{r}
markers2 <- c("RARA","RARB","RARG","RXRA","RXRB","ALDH1A1","ALDH1A2","ALDH1A3")
for(i in intersect(markers2, humMacMouOrth.1to1.in$external_gene_name.x)){
  plotgene(i)
}
```
"RARA"    "RARB"    "RARG"    "RXRA"    "RXRB"    "ALDH1A2" "ALDH1A3" added to the three species SPG


```{r fig.width=10,fig.height=20}
pdf("~/storage/HumanMacaqueMouseCompAnaysis/interdata/SPG_markers/AlignedThreeSpecies_Violin.pdf",width = 10, height = 20)
VlnPlot(object = spg.cca, features.plot = intersect(markers,humMacMouOrth.1to1.in$external_gene_name.x),nCol = 4,group.by = "res.0.6.ord",point.size.use=0, ident.include = c(1:6))
dev.off()
```

```{r}
dge.hum.spg.norm <- apply(dge.hum.spg.raw,2,function(x) x/sum(x)*1e4)
dge.mac.spg.norm <- apply(dge.mac.spg.raw,2,function(x) x/sum(x)*1e4)
dge.mou.spg.norm <- apply(dge.mou.spg.raw,2,function(x) x/sum(x)*1e4)
```


```{r}
plts <- list()
j=0
for(i in intersect(markers, rownames(dge.hum.spg.norm))){
  j = j+1
  dat = data.frame(expr = log(dge.hum.spg.norm[i,]+1), ident = spg.cca@ident[colnames(dge.hum.spg.norm)])
  plts[[j]] <- ggplot(dat, aes(x=ident,y=expr, fill=ident))+geom_violin(scale = "width") + xlab("Identity") + ylab("") + ggtitle(i) + theme(legend.position = "none")
}
```

```{r}
library(gridExtra)
```

```{r fig.width=10, fig.height=30}
pdf("~/storage/HumanMacaqueMouseCompAnaysis/interdata/SPG_markers/Hum_Violin.pdf",width = 10, height = 30)
do.call("grid.arrange", c(plts, ncol=4))
dev.off()
```

```{r}
plts <- list()
j=0
#mac.markers <- humMacMouOrth.1to1.in$mmulatta_homolog_associated_gene_name[humMacMouOrth.1to1.in$external_gene_name.x %in% markers]
for(i in intersect(markers, rownames(dge.mac.spg.norm))){
  j = j+1
  dat = data.frame(expr = log(dge.mac.spg.norm[i,]+1), ident = spg.cca@ident[colnames(dge.mac.spg.norm)])
  plts[[j]] <- ggplot(dat, aes(x=ident,y=expr, fill=ident))+geom_violin(scale = "width") + xlab("Identity") + ylab("") + ggtitle(i) + theme(legend.position = "none")
}

pdf("~/storage/HumanMacaqueMouseCompAnaysis/interdata/SPG_markers/Mac_Violin.pdf",width = 10, height = 30)
do.call("grid.arrange", c(plts, ncol=4))
dev.off()
```

```{r}
plts <- list()
j=0
mou.markers <- humMacMouOrth.1to1.in$mmusculus_homolog_associated_gene_name[humMacMouOrth.1to1.in$external_gene_name.x %in% markers]
for(i in mou.markers){
  j = j+1
  dat = data.frame(expr = log(dge.mou.spg.norm[i,]+1), ident = spg.cca@ident[colnames(dge.mou.spg.norm)])
  plts[[j]] <- ggplot(dat, aes(x=ident,y=expr, fill=ident))+geom_violin(scale = "width") + xlab("Identity") + ylab("") + ggtitle(i) + theme(legend.position = "none")
}

pdf("~/storage/HumanMacaqueMouseCompAnaysis/interdata/SPG_markers/Mou_Violin.pdf",width = 10, height = 30)
do.call("grid.arrange", c(plts, ncol=4))
dev.off()
```


