#. Kmeans for SPG Clusters

Read in SPG cluster centroid dataset.
```{r}
hum.spg.cen <- readRDS("~/storage/HumanMacaqueMouseCompAnaysis/interdata/hum.spg.cen.rds")
mac.spg.cen <- readRDS("~/storage/HumanMacaqueMouseCompAnaysis/interdata/mac.spg.cen.rds")
mou.spg.cen <- readRDS("~/storage/HumanMacaqueMouseCompAnaysis/interdata/mou.spg.cen.rds")
#Process 
hum.spg.cen.6 <- hum.spg.cen[,1:6]
mac.spg.cen.6 <- mac.spg.cen[,1:6]
mou.spg.cen.6 <- mou.spg.cen[,1:6]
rownames(hum.spg.cen.6) <- rownames(spg.cca@data)
rownames(mac.spg.cen.6) <- rownames(spg.cca@data)
rownames(mou.spg.cen.6) <- rownames(spg.cca@data)
hum.spg.cen.6.log <- log(hum.spg.cen.6+1)
mac.spg.cen.6.log <- log(mac.spg.cen.6+1)
mou.spg.cen.6.log <- log(mou.spg.cen.6+1)
hum.spg.cen.6.std <- t(apply(hum.spg.cen.6.log, 1, function(x) (x-mean(x))/sd(x)))
mac.spg.cen.6.std <- t(apply(mac.spg.cen.6.log, 1, function(x) (x-mean(x))/sd(x)))
mou.spg.cen.6.std <- t(apply(mou.spg.cen.6.log, 1, function(x) (x-mean(x))/sd(x)))
``` 

Select HVGs for each species.
```{r fig.width=5,fig.height=5}
g.mean <- apply(hum.spg.cen.6.log,1,mean)
g.var <- apply(hum.spg.cen.6.log,1,var)
plot(g.mean, g.var/g.mean, cex=0.5)
hum.spg.hvg <- names(g.mean)[g.mean > 0.2 & g.var/g.mean > 0.1]
abline(h=0.1)
abline(v=0.2)
length(hum.spg.hvg)
```

```{r fig.width=5,fig.height=5}
g.mean <- apply(mac.spg.cen.6.log,1,mean)
g.var <- apply(mac.spg.cen.6.log,1,var)
plot(g.mean, g.var/g.mean, cex=0.5)
mac.spg.hvg <- names(g.mean)[g.mean > 0.2 & g.var/g.mean > 0.1]
abline(h=0.1)
abline(v=0.2)
length(mac.spg.hvg)
```

```{r fig.width=5,fig.height=5}
g.mean <- apply(mou.spg.cen.6.log,1,mean)
g.var <- apply(mou.spg.cen.6.log,1,var)
plot(g.mean, g.var/g.mean, cex=0.5)
mou.spg.hvg <- names(g.mean)[g.mean > 0.2 & g.var/g.mean > 0.1]
abline(h=0.1)
abline(v=0.2)
length(mou.spg.hvg)
```


##. Asking for 6 clusters from kmeans
```{r}
colnames(hum.spg.cen.6.std) <- c(1:6)
set.seed(123)
hum.spg.kmean <- kmeans(hum.spg.cen.6.std[hum.spg.hvg,],6)
for(i in c(1:6)){
dr = Heatmap(hum.spg.cen.6.std[names(which(hum.spg.kmean$cluster==i)),], cluster_rows = FALSE, cluster_columns = FALSE,show_row_names = FALSE,row_title = i)
draw(dr)
}
```

```{r}
colnames(mac.spg.cen.6.std) <- c(1:6)
set.seed(123)
mac.spg.kmean <- kmeans(mac.spg.cen.6.std[mac.spg.hvg,],6)
for(i in c(1:6)){
dr = Heatmap(mac.spg.cen.6.std[names(which(mac.spg.kmean$cluster==i)),], cluster_rows = FALSE, cluster_columns = FALSE,show_row_names = FALSE,row_title = i)
draw(dr)
}
```
```{r}
set.seed(123)
colnames(mou.spg.cen.6.std) <- c(1:6)
mou.spg.kmean <- kmeans(mou.spg.cen.6.std[mou.spg.hvg,],6)
for(i in c(1:6)){
dr = Heatmap(mou.spg.cen.6.std[names(which(mou.spg.kmean$cluster==i)),], cluster_rows = FALSE, cluster_columns = FALSE,show_row_names = FALSE,row_title = i)
draw(dr)
}
```

##. Asking for 3 clusters from kmeans
```{r}
colnames(hum.spg.cen.6.std) <- c(1:6)
set.seed(123)
hum.spg.kmean <- kmeans(hum.spg.cen.6.std[hum.spg.hvg,],3)
for(i in c(1:3)){
dr = Heatmap(hum.spg.cen.6.std[names(which(hum.spg.kmean$cluster==i)),], cluster_rows = FALSE, cluster_columns = FALSE,show_row_names = FALSE,row_title = i)
draw(dr)
}
```
```{r}
colnames(mac.spg.cen.6.std) <- c(1:6)
set.seed(123)
mac.spg.kmean <- kmeans(mac.spg.cen.6.std[mac.spg.hvg,],3)
for(i in c(1:3)){
dr = Heatmap(mac.spg.cen.6.std[names(which(mac.spg.kmean$cluster==i)),], cluster_rows = FALSE, cluster_columns = FALSE,show_row_names = FALSE,row_title = i)
draw(dr)
}
```
```{r}
set.seed(123)
colnames(mou.spg.cen.6.std) <- c(1:6)
mou.spg.kmean <- kmeans(mou.spg.cen.6.std[mou.spg.hvg,],3)
for(i in c(1:3)){
dr = Heatmap(mou.spg.cen.6.std[names(which(mou.spg.kmean$cluster==i)),], cluster_rows = FALSE, cluster_columns = FALSE,show_row_names = FALSE,row_title = i)
draw(dr)
}
```

##. Asking for 4 clusters from kmeans
```{r}
colnames(hum.spg.cen.6.std) <- c(1:6)
set.seed(123)
hum.spg.kmean <- kmeans(hum.spg.cen.6.std[hum.spg.hvg,],4)
for(i in c(1:4)){
dr = Heatmap(hum.spg.cen.6.std[names(which(hum.spg.kmean$cluster==i)),], cluster_rows = FALSE, cluster_columns = FALSE,show_row_names = FALSE,row_title = i)
draw(dr)
}
```
```{r}
colnames(mac.spg.cen.6.std) <- c(1:6)
set.seed(123)
mac.spg.kmean <- kmeans(mac.spg.cen.6.std[mac.spg.hvg,],4)
for(i in c(1:4)){
dr = Heatmap(mac.spg.cen.6.std[names(which(mac.spg.kmean$cluster==i)),], cluster_rows = FALSE, cluster_columns = FALSE,show_row_names = FALSE,row_title = i)
draw(dr)
}
```
```{r}
set.seed(123)
colnames(mou.spg.cen.6.std) <- c(1:6)
mou.spg.kmean <- kmeans(mou.spg.cen.6.std[mou.spg.hvg,],4)
for(i in c(1:4)){
dr = Heatmap(mou.spg.cen.6.std[names(which(mou.spg.kmean$cluster==i)),], cluster_rows = FALSE, cluster_columns = FALSE,show_row_names = FALSE,row_title = i)
draw(dr)
}
```

Write Kmeans to file
```{r}
len <- max(max(table(hum.spg.kmean$cluster)))
hum.hvgs.kmean.write <- c()
stages <- c(3,2,1,4)
for(i in stages){
  genes <- names(which(hum.spg.kmean$cluster==i))
  hum.hvgs.kmean.write <- cbind(hum.hvgs.kmean.write, c(genes, rep("", len-length(genes) )))
}
write.table(hum.hvgs.kmean.write, file="~/storage/HumanMacaqueMouseCompAnaysis/interdata/SPG_GeneKmeans/hum.spg.kmean.gene.list.txt", quote = FALSE, row.names = FALSE,sep = "\t")

len <- max(max(table(mac.spg.kmean$cluster)))
mac.hvgs.kmean.write <- c()
stages <- c(2,3,1,4)
for(i in stages){
  genes <- humMacMouOrth.1to1.in$mmulatta_homolog_associated_gene_name[humMacMouOrth.1to1.in$external_gene_name.x %in% names(which(mac.spg.kmean$cluster==i))]
  mac.hvgs.kmean.write <- cbind(mac.hvgs.kmean.write, c(genes, rep("", len-length(genes) )))
}
write.table(mac.hvgs.kmean.write, file="~/storage/HumanMacaqueMouseCompAnaysis/interdata/SPG_GeneKmeans/mac.spg.kmean.gene.list.txt", quote = FALSE, row.names = FALSE,sep = "\t")

len <- max(max(table(mou.spg.kmean$cluster)))
mou.hvgs.kmean.write <- c()
stages <- c(1,4,2,3)
for(i in stages){
  genes <- humMacMouOrth.1to1.in$mmusculus_homolog_associated_gene_name[humMacMouOrth.1to1.in$external_gene_name.x %in% names(which(mou.spg.kmean$cluster==i))]
  mou.hvgs.kmean.write <- cbind(mou.hvgs.kmean.write, c(genes, rep("", len-length(genes) )))
}
write.table(mou.hvgs.kmean.write, file="~/storage/HumanMacaqueMouseCompAnaysis/interdata/SPG_GeneKmeans/mou.spg.kmeans.gene.list.txt", quote = FALSE, row.names = FALSE,sep = "\t")
```

Check genes that have high expression in both ends
```{r}
mac.genes.hb <- names(which(mac.spg.kmean$cluster==4))
mac.genes.hb <- mac.genes.hb[mac.spg.cen.6.std[mac.genes.hb,1] > 0.9 & mac.spg.cen.6.std[mac.genes.hb,6] > 0.9]
```

```{r fig.width=10, fig.height=20}
colfunc <- colorRampPalette(c("blue", "white", "red"), bias=5)
FeaturePlot(object = spg.cca, features.plot = mac.genes.hb,cols.use = colfunc(100),cells.use = spg.cca@cell.names[spg.cca@meta.data$species=="Monkey"], dark.theme = TRUE, pt.size = 0.5)
```


overlapping HVGs
```{r}
int.hvgs <- intersect(hum.spg.hvg, mac.spg.hvg)
int.hvgs <- intersect(int.hvgs, mou.spg.hvg)
```

Expression Correlation acorss 6 clusters
```{r}
spg.gene.cor <- c()
for(i in rownames(hum.spg.cen.6)){
  spg.hummac.cor <- cor(hum.spg.cen.6[i,],mac.spg.cen.6[i,], method = "spearman")
  spg.hummou.cor <- cor(hum.spg.cen.6[i,],mou.spg.cen.6[i,], method = "spearman")
  spg.macmou.cor <- cor(mac.spg.cen.6[i,],mou.spg.cen.6[i,], method = "spearman")
  spg.gene.cor <- rbind(spg.gene.cor, c(spg.hummac.cor, spg.hummou.cor, spg.macmou.cor))
}
rownames(spg.gene.cor) <- rownames(hum.spg.cen.6)
```

```{r fig.height=5, fig.width=5}
hist(spg.gene.cor[,1], breaks = 200, main = "Human vs. Monkey", xlab="Gene expression correlation")
hist(spg.gene.cor[,2], breaks = 200, main = "Human vs. Mouse", xlab="Gene expression correlation")
hist(spg.gene.cor[,3], breaks = 200, main = "Monkey vs. Mouse", xlab="Gene expression correlation")
```

#Correlation with sequence identity and dN/dS
```{r}
spg.gene.cor <- cbind(spg.gene.cor, gene.cor[,c(4:9)])
colnames(spg.gene.cor) <- colnames(gene.cor)
```

```{r fig.width=5, fig.height=10}
hummac.cor <- data.frame(spg.gene.cor[,c("HumMacCor", "HumMacdNdS", "HumMacSeq")])
hummac.cor <- hummac.cor[!is.na(hummac.cor$HumMacCor) & !is.na(hummac.cor$HumMacdNdS) & !is.infinite(hummac.cor$HumMacdNdS),]
hummac.cor$seqdensity <- get_density(hummac.cor$HumMacSeq, hummac.cor$HumMacCor, n = 100)
cortest <- cor.test(hummac.cor$HumMacSeq, hummac.cor$HumMacCor, method="spearman")
p <- ggplot(hummac.cor) + geom_point(aes(x=HumMacSeq, y=HumMacCor, color = seqdensity)) + scale_color_viridis()+ ggtitle(paste0("Human vs. Monkey","\ncor: ",round(cortest$estimate,3), " pvalue:", format(cortest$p.value,digits=4)) )+ theme(legend.position="none") + xlab("Identity") + ylab("Expression Conservability") + theme(plot.title = element_text(hjust = 0.5, face = "bold"))
p1 <- ggMarginal(p, type="density")

hummac.cor$dnsdensity <- get_density(hummac.cor$HumMacdNdS, hummac.cor$HumMacCor, n = 100)
cortest <- cor.test(hummac.cor$HumMacdNdS, hummac.cor$HumMacCor, method="spearman")
p <- ggplot(hummac.cor) + geom_point(aes(x=HumMacdNdS, y=HumMacCor, color = dnsdensity)) + scale_color_viridis()+ ggtitle(paste0("Human vs. Monkey","\ncor: ",round(cortest$estimate,3), " pvalue:", format(cortest$p.value,digits=4)) ) + theme(legend.position="none") + xlab("dN/dS") + ylab("Expression Conservability") + theme(plot.title = element_text(hjust = 0.5, face = "bold"))
p2 <- ggMarginal(p, type="density")

grid.arrange(p1, p2, ncol=1)
```

```{r fig.width=5, fig.height=10}
hummou.cor <- data.frame(spg.gene.cor[,c("HumMouCor", "HumMoudNdS", "HumMouSeq")])
hummou.cor <- hummou.cor[!is.na(hummou.cor$HumMouCor) & !is.na(hummou.cor$HumMoudNdS) & !is.infinite(hummou.cor$HumMoudNdS),]
hummou.cor$seqdensity <- get_density(hummou.cor$HumMouSeq, hummou.cor$HumMouCor, n = 100)
cortest <- cor.test(hummou.cor$HumMouSeq, hummou.cor$HumMouCor, method="spearman")
p <- ggplot(hummou.cor) + geom_point(aes(x=HumMouSeq, y=HumMouCor, color = seqdensity)) + scale_color_viridis()+ ggtitle(paste0("Human vs. Mouse","\ncor: ",round(cortest$estimate,3), " pvalue:", format(cortest$p.value,digits=4)) )+ theme(legend.position="none") + xlab("Identity") + ylab("Expression Conservability") + theme(plot.title = element_text(hjust = 0.5, face = "bold"))
p1 <- ggMarginal(p, type="density")

hummou.cor$dnsdensity <- get_density(hummou.cor$HumMoudNdS, hummou.cor$HumMouCor, n = 100)
cortest <- cor.test(hummou.cor$HumMoudNdS, hummou.cor$HumMouCor, method="spearman")
p <- ggplot(hummou.cor) + geom_point(aes(x=HumMoudNdS, y=HumMouCor, color = dnsdensity)) + scale_color_viridis()+ ggtitle(paste0("Human vs. Mouse","\ncor: ",round(cortest$estimate,3), " pvalue:", format(cortest$p.value,digits=4)) ) + theme(legend.position="none") + xlab("dN/dS") + ylab("Expression Conservability") + theme(plot.title = element_text(hjust = 0.5, face = "bold"))
p2 <- ggMarginal(p, type="density")

grid.arrange(p1, p2, ncol=1)
```

```{r fig.width=5, fig.height=10}
macmou.cor <- data.frame(spg.gene.cor[,c("MacMouCor", "MacMoudNdS", "MacMouSeq")])
macmou.cor <- macmou.cor[!is.na(macmou.cor$MacMouCor) & !is.na(macmou.cor$MacMoudNdS) & !is.infinite(macmou.cor$MacMoudNdS),]
macmou.cor$seqdensity <- get_density(macmou.cor$MacMouSeq, macmou.cor$MacMouCor, n = 100)
cortest <- cor.test(macmou.cor$MacMouSeq, macmou.cor$MacMouCor, method="spearman")
p <- ggplot(macmou.cor) + geom_point(aes(x=MacMouSeq, y=MacMouCor, color = seqdensity)) + scale_color_viridis()+ ggtitle(paste0("Monkey vs. Mouse","\ncor: ",round(cortest$estimate,3), " pvalue:", format(cortest$p.value,digits=4)) )+ theme(legend.position="none") + xlab("Identity") + ylab("Expression Conservability") + theme(plot.title = element_text(hjust = 0.5, face = "bold"))
p1 <- ggMarginal(p, type="density")

macmou.cor$dnsdensity <- get_density(macmou.cor$MacMoudNdS, macmou.cor$MacMouCor, n = 100)
cortest <- cor.test(macmou.cor$MacMoudNdS, macmou.cor$MacMouCor, method="spearman")
p <- ggplot(macmou.cor) + geom_point(aes(x=MacMoudNdS, y=MacMouCor, color = dnsdensity)) + scale_color_viridis()+ ggtitle(paste0("Monkey vs. Mouse","\ncor: ",round(cortest$estimate,3), " pvalue:", format(cortest$p.value,digits=4)) ) + theme(legend.position="none") + xlab("dN/dS") + ylab("Expression Conservability") + theme(plot.title = element_text(hjust = 0.5, face = "bold"))
p2 <- ggMarginal(p, type="density")

grid.arrange(p1, p2, ncol=1)
```


kmeans for all genes
```{r}
hum.spg.cen <- read.table("~/storage/HumanMacaqueMouseCompAnaysis/interdata/SPG_centroids/HumanSPGCluster_AllGene_Centroids.txt",header = TRUE, row.names = 1)
mac.spg.cen <- read.table("~/storage/HumanMacaqueMouseCompAnaysis/interdata/SPG_centroids/MonkeySPGCluster_AllGene_Centroids.txt",header = TRUE, row.names = 1)
mou.spg.cen <- read.table("~/storage/HumanMacaqueMouseCompAnaysis/interdata/SPG_centroids/MouseSPGCluster_AllGene_Centroids.txt",header = TRUE, row.names = 1)

hum.spg.cen.lg = log(hum.spg.cen + 1)
mac.spg.cen.lg = log(mac.spg.cen + 1)
mou.spg.cen.lg = log(mou.spg.cen + 1)

hum.spg.cen.std = t(apply(hum.spg.cen.lg, 1, function(x) (x-mean(x))/sd(x)))
mac.spg.cen.std = t(apply(mac.spg.cen.lg, 1, function(x) (x-mean(x))/sd(x)))
mou.spg.cen.std = t(apply(mou.spg.cen.lg, 1, function(x) (x-mean(x))/sd(x)))
```

HVG

```{r}
get_density <- function(x, y, ...) {
  dens <- MASS::kde2d(x, y, ...)
  ix <- findInterval(x, dens$x)
  iy <- findInterval(y, dens$y)
  ii <- cbind(ix, iy)
  return(dens$z[ii])
}

```

```{r}
library(viridis)
library(ggExtra)
library(gridExtra)
```


plot the density.
```{r fig.width=5,fig.height=5}
g.mean <- apply(hum.spg.cen.lg,1,mean)
g.var <- apply(hum.spg.cen.lg,1,var)

g.var <- g.var[g.mean!=0]
g.mean <- g.mean[g.mean!=0]
hum.spg.hvg <- names(g.mean)[g.mean > 0.1 & g.var/g.mean > 0.1]
length(hum.spg.hvg)
hum.gen <- data.frame(mean = g.mean, varmean = g.var/g.mean)
hum.gen$den <- get_density(hum.gen$mean, hum.gen$varmean, n=100)

p <- ggplot(hum.gen) + geom_point(aes(x=mean, y=varmean, color = log(den+1)),size=0.5) + scale_color_viridis()+ ggtitle("Human HVGs") + theme(legend.position="none") + xlab("Mean") + ylab("Variation/Mean") + theme(plot.title = element_text(hjust = 0.5, face = "bold")) + geom_hline(yintercept = 0.1) + geom_vline(xintercept = 0.1)
p1 <- ggMarginal(p, type="density")
grid.arrange(p1, ncol=1)
```



```{r fig.width=5,fig.height=5}
g.mean <- apply(mac.spg.cen.lg,1,mean)
g.var <- apply(mac.spg.cen.lg,1,var)
g.var <- g.var[g.mean!=0]
g.mean <- g.mean[g.mean!=0]

mac.spg.hvg <- names(g.mean)[g.mean > 0.1 & g.var/g.mean > 0.1]
length(mac.spg.hvg)
mac.gen <- data.frame(mean = g.mean, varmean = g.var/g.mean)
mac.gen$den <- get_density(mac.gen$mean, mac.gen$varmean, n=100)

p <- ggplot(mac.gen) + geom_point(aes(x=mean, y=varmean, color = log(den+1)),size=0.5) + scale_color_viridis()+ ggtitle("Monkey HVGs") + theme(legend.position="none") + xlab("Mean") + ylab("Variation/Mean") + theme(plot.title = element_text(hjust = 0.5, face = "bold")) + geom_hline(yintercept = 0.1) + geom_vline(xintercept = 0.1)
p2 <- ggMarginal(p, type="density")
grid.arrange(p2, ncol=1)
```

```{r fig.width=5,fig.height=5}
g.mean <- apply(mou.spg.cen.lg,1,mean)
g.var <- apply(mou.spg.cen.lg,1,var)
g.var <- g.var[g.mean!=0]
g.mean <- g.mean[g.mean!=0]

mou.spg.hvg <- names(g.mean)[g.mean > 0.1 & g.var/g.mean > 0.06]
length(mou.spg.hvg)

mou.gen <- data.frame(mean = g.mean, varmean = g.var/g.mean)
mou.gen$den <- get_density(mou.gen$mean, mou.gen$varmean, n=100)

p <- ggplot(mou.gen) + geom_point(aes(x=mean, y=varmean, color = log(den+1)),size=0.5) + scale_color_viridis()+ ggtitle("Mouse HVGs") + theme(legend.position="none") + xlab("Mean") + ylab("Variation/Mean") + theme(plot.title = element_text(hjust = 0.5, face = "bold")) + geom_hline(yintercept = 0.06) + geom_vline(xintercept = 0.1)
p2 <- ggMarginal(p, type="density")
grid.arrange(p2, ncol=1)
```

##. Asking for 3 clusters from kmeans
```{r}
colnames(hum.spg.cen.std) <- c(1:6)
set.seed(123)
hum.spg.kmean <- kmeans(hum.spg.cen.std[hum.spg.hvg,],3)
for(i in c(1:3)){
dr = Heatmap(hum.spg.cen.std[names(which(hum.spg.kmean$cluster==i)),], cluster_rows = FALSE, cluster_columns = FALSE,show_row_names = FALSE,row_title = i)
draw(dr)
}
```
```{r}
colnames(mac.spg.cen.std) <- c(1:6)
set.seed(123)
mac.spg.kmean <- kmeans(mac.spg.cen.std[mac.spg.hvg,],3)
for(i in c(1:3)){
dr = Heatmap(mac.spg.cen.std[names(which(mac.spg.kmean$cluster==i)),], cluster_rows = FALSE, cluster_columns = FALSE,show_row_names = FALSE,row_title = i)
draw(dr)
}
```
```{r}
set.seed(123)
colnames(mou.spg.cen.std) <- c(1:6)
mou.spg.kmean <- kmeans(mou.spg.cen.std[mou.spg.hvg,],3)
for(i in c(1:3)){
dr = Heatmap(mou.spg.cen.std[names(which(mou.spg.kmean$cluster==i)),], cluster_rows = FALSE, cluster_columns = FALSE,show_row_names = FALSE,row_title = i)
draw(dr)
}
```

##. Asking for 4 clusters from kmeans
```{r}
colnames(hum.spg.cen.std) <- c(1:6)
set.seed(123)
hum.spg.kmean <- kmeans(hum.spg.cen.std[hum.spg.hvg,],4)
for(i in c(1:4)){
dr = Heatmap(hum.spg.cen.std[names(which(hum.spg.kmean$cluster==i)),], cluster_rows = FALSE, cluster_columns = FALSE,show_row_names = FALSE,row_title = i)
draw(dr)
}
```
```{r}
colnames(mac.spg.cen.std) <- c(1:6)
set.seed(123)
mac.spg.kmean <- kmeans(mac.spg.cen.std[mac.spg.hvg,],4)
for(i in c(1:4)){
dr = Heatmap(mac.spg.cen.std[names(which(mac.spg.kmean$cluster==i)),], cluster_rows = FALSE, cluster_columns = FALSE,show_row_names = FALSE,row_title = i)
draw(dr)
}
```
```{r}
set.seed(123)
colnames(mou.spg.cen.std) <- c(1:6)
mou.spg.kmean <- kmeans(mou.spg.cen.std[mou.spg.hvg,],4)
for(i in c(1:4)){
dr = Heatmap(mou.spg.cen.std[names(which(mou.spg.kmean$cluster==i)),], cluster_rows = FALSE, cluster_columns = FALSE,show_row_names = FALSE,row_title = i)
draw(dr)
}
```

Write Kmeans to file
```{r}
len <- max(max(table(hum.spg.kmean$cluster)))
hum.hvgs.kmean.write <- c()
stages <- c(3,1,4,2)
for(i in stages){
  genes <- names(which(hum.spg.kmean$cluster==i))
  hum.hvgs.kmean.write <- cbind(hum.hvgs.kmean.write, c(genes, rep("", len-length(genes) )))
}
write.table(hum.hvgs.kmean.write, file="~/storage/HumanMacaqueMouseCompAnaysis/interdata/SPG_GeneKmeans/hum.spg.kmean.allgene.list.txt", quote = FALSE, row.names = FALSE,sep = "\t")

len <- max(max(table(mac.spg.kmean$cluster)))
mac.hvgs.kmean.write <- c()
stages <- c(2,4,3,1)
for(i in stages){
  genes <- names(which(mac.spg.kmean$cluster==i))
  mac.hvgs.kmean.write <- cbind(mac.hvgs.kmean.write, c(genes, rep("", len-length(genes) )))
}
write.table(mac.hvgs.kmean.write, file="~/storage/HumanMacaqueMouseCompAnaysis/interdata/SPG_GeneKmeans/mac.spg.kmean.allgene.list.txt", quote = FALSE, row.names = FALSE,sep = "\t")

len <- max(max(table(mou.spg.kmean$cluster)))
mou.hvgs.kmean.write <- c()
stages <- c(3,1,4,2)
for(i in stages){
  genes <- names(which(mou.spg.kmean$cluster==i))
  mou.hvgs.kmean.write <- cbind(mou.hvgs.kmean.write, c(genes, rep("", len-length(genes) )))
}
write.table(mou.hvgs.kmean.write, file="~/storage/HumanMacaqueMouseCompAnaysis/interdata/SPG_GeneKmeans/mou.spg.kmean.allgene.list.txt", quote = FALSE, row.names = FALSE,sep = "\t")
```

##.Connecting plot
```{r}
hum.spg.kmeans.genes <- c()
stages <- c(3,1,4,2)
for(i in c(1:4)){
  hum.spg.kmeans.genes <- rbind(hum.spg.kmeans.genes, data.frame(genes = names(which(hum.spg.kmean$cluster==stages[i])), stages=rep(i, sum(hum.spg.kmean$cluster==stages[i]))))
}

mac.spg.kmeans.genes <- c()
stages <- c(2,4,3,1)
for(i in c(1:4)){
  mac.spg.kmeans.genes <- rbind(mac.spg.kmeans.genes, data.frame(genes = names(which(mac.spg.kmean$cluster==stages[i])), stages=rep(i, sum(mac.spg.kmean$cluster==stages[i]))))
}

mou.spg.kmeans.genes <- c()
stages <- c(3,1,4,2)
for(i in c(1:4)){
  mou.spg.kmeans.genes <- rbind(mou.spg.kmeans.genes, data.frame(genes = names(which(mou.spg.kmean$cluster==stages[i])), stages=rep(i, sum(mou.spg.kmean$cluster==stages[i]))))
}
```

##human Mac
#Connector lines
```{r}
kmeans.hum.mac.genes <- humMacMouOrth.1to1.in[humMacMouOrth.1to1.in$external_gene_name.x %in% hum.spg.hvg,]
kmeans.hum.mac.genes <- kmeans.hum.mac.genes[kmeans.hum.mac.genes$mmulatta_homolog_associated_gene_name %in% mac.spg.hvg,]
```

```{r}
pos.hum.mac <- c()
for(i in c(1:dim(kmeans.hum.mac.genes)[1])){
  hum.gen <- kmeans.hum.mac.genes$external_gene_name.x[i]
  mac.gen <- kmeans.hum.mac.genes$mmulatta_homolog_associated_gene_name[i]
  x = which(hum.spg.kmeans.genes$genes == hum.gen)
  y = which(mac.spg.kmeans.genes$genes == mac.gen)
  pos.hum.mac <- rbind(pos.hum.mac,c(x,y,hum.gen, mac.gen))
}
pos.hum.mac <- data.matrix(pos.hum.mac)
colnames(pos.hum.mac) = c("hum.pos","mac.pos", "hum.gen","mac.gen")
```

Sort within each cluster
```{r}
rownames(pos.hum.mac) <- pos.hum.mac[,"hum.gen"]
hum.kmeans.genes.ord <- hum.spg.kmeans.genes
for(i in c(1:6)){
  genes <- hum.spg.kmeans.genes$genes[hum.spg.kmeans.genes$stages==i]
  ogenes <- as.character(genes[!genes %in% pos.hum.mac[,"hum.gen"]])
  genes <- pos.hum.mac[,"hum.gen"][pos.hum.mac[,"hum.gen"] %in% genes]
  genes <- genes[order(as.numeric(pos.hum.mac[genes,"mac.pos"]))]
  hum.kmeans.genes.ord$genes[hum.spg.kmeans.genes$stages==i] = c(genes,ogenes)
}
```

```{r}
rownames(pos.hum.mac) <- pos.hum.mac[,"mac.gen"]
mac.kmeans.genes.ord <- mac.spg.kmeans.genes
for(i in c(1:6)){
  genes <- mac.spg.kmeans.genes$genes[mac.spg.kmeans.genes$stages==i]
  ogenes <- as.character(genes[!genes %in% pos.hum.mac[,"mac.gen"]])
  genes <- pos.hum.mac[,"mac.gen"][pos.hum.mac[,"mac.gen"] %in% genes]
  genes <- genes[order(as.numeric(pos.hum.mac[genes,"hum.pos"]))]
  mac.kmeans.genes.ord$genes[mac.spg.kmeans.genes$stages==i] = c(genes,ogenes)
}
```


```{r fig.width=4, fig.height=10}
cols <- c("#fabc3c",
"#a64dd4",
"#14e072",
"#d9014a",
"#abb4ff",
"#b28e00")
plot(1,xlim=c(1,2),ylim=c(0,1000),type="n",axes=FALSE,ann=FALSE)

for(i in c(1:dim(kmeans.hum.mac.genes)[1])){
#for(i in c(1:10)){ 
  x = which(hum.kmeans.genes.ord$genes == kmeans.hum.mac.genes$external_gene_name.x[i])
  y = which(mac.kmeans.genes.ord$genes == kmeans.hum.mac.genes$mmulatta_homolog_associated_gene_name[i])
  #print(kmeans.hum.mac.genes$external_gene_name.x[i])
  #print(kmeans.hum.mac.genes$mmusculus_homolog_associated_gene_name[i])
  x = 1000-x*1000/dim(hum.spg.kmeans.genes)[1]
  y = 1000-y*1000/dim(mac.spg.kmeans.genes)[1]
  #print(x)
  #print(y)
  segments(1,x,2,y,lwd = 0.1)
}
points(data.frame(x=rep(1,dim(hum.spg.kmeans.genes)[1]),y=1000-c(1:dim(hum.spg.kmeans.genes)[1])*1000/dim(hum.spg.kmeans.genes)[1]),col=cols[hum.spg.kmeans.genes$stages], pch=15, cex=0.8)
points(data.frame(x=rep(2,dim(mac.spg.kmeans.genes)[1]),y=1000-c(1:dim(mac.spg.kmeans.genes)[1])*1000/dim(mac.spg.kmeans.genes)[1]),col=cols[mac.spg.kmeans.genes$stages], pch=15, cex=0.8)
#segments(1,0,10,0,col = "red")
```

```{r fig.width=4, fig.height=10}
cols <- c("#fabc3c",
"#a64dd4",
"#14e072",
"#d9014a",
"#abb4ff",
"#b28e00")
ha = rowAnnotation(stages = mac.spg.kmeans.genes$stages, col = list(stages = c("1" = cols[1], "2" = cols[2], "3" = cols[3],"4" = cols[4], "5" = cols[5], "6" = cols[6])))
#ha = HeatmapAnnotation(foo = anno_simple(cbind(1:10, 10:1), pch = 1:10))
Heatmap(mac.spg.cen.std[as.character(mac.spg.kmeans.genes$genes),], cluster_rows = FALSE, cluster_columns = FALSE,show_row_names = FALSE, left_annotation = ha)
```

```{r fig.width=4, fig.height=10}
cols <- c("#fabc3c",
"#a64dd4",
"#14e072",
"#d9014a",
"#abb4ff",
"#b28e00")
ha = rowAnnotation(stages = hum.spg.kmeans.genes$stages, col = list(stages = c("1" = cols[1], "2" = cols[2], "3" = cols[3],"4" = cols[4], "5" = cols[5], "6" = cols[6])))
#ha = HeatmapAnnotation(foo = anno_simple(cbind(1:10, 10:1), pch = 1:10))
Heatmap(hum.spg.cen.std[as.character(hum.spg.kmeans.genes$genes),], cluster_rows = FALSE, cluster_columns = FALSE,show_row_names = FALSE, right_annotation = ha)
```

##human Mou
#Connector lines
```{r}
kmeans.hum.mou.genes <- humMacMouOrth.1to1.in[humMacMouOrth.1to1.in$external_gene_name.x %in% hum.spg.hvg,]
kmeans.hum.mou.genes <- kmeans.hum.mou.genes[kmeans.hum.mou.genes$mmusculus_homolog_associated_gene_name %in% mou.spg.hvg,]
```

```{r}
pos.hum.mou <- c()
for(i in c(1:dim(kmeans.hum.mou.genes)[1])){
  hum.gen <- kmeans.hum.mou.genes$external_gene_name.x[i]
  mou.gen <- kmeans.hum.mou.genes$mmusculus_homolog_associated_gene_name[i]
  x = which(hum.spg.kmeans.genes$genes == hum.gen)
  y = which(mou.spg.kmeans.genes$genes == mou.gen)
  pos.hum.mou <- rbind(pos.hum.mou,c(x,y,hum.gen, mou.gen))
}
pos.hum.mou <- data.matrix(pos.hum.mou)
colnames(pos.hum.mou) = c("hum.pos","mou.pos", "hum.gen","mou.gen")
```

Sort within each cluster
```{r}

rownames(pos.hum.mou) <- pos.hum.mou[,"hum.gen"]
hum.kmeans.genes.ord <- hum.spg.kmeans.genes
for(i in c(1:6)){
  genes <- hum.spg.kmeans.genes$genes[hum.spg.kmeans.genes$stages==i]
  ogenes <- as.character(genes[!genes %in% pos.hum.mou[,"hum.gen"]])
  genes <- pos.hum.mou[,"hum.gen"][pos.hum.mou[,"hum.gen"] %in% genes]
  genes <- genes[order(as.numeric(pos.hum.mou[genes,"mou.pos"]))]
  hum.kmeans.genes.ord$genes[hum.spg.kmeans.genes$stages==i] = c(genes,ogenes)
}
```

```{r}
rownames(pos.hum.mou) <- pos.hum.mou[,"mou.gen"]
mou.kmeans.genes.ord <- mou.spg.kmeans.genes
for(i in c(1:6)){
  genes <- mou.spg.kmeans.genes$genes[mou.spg.kmeans.genes$stages==i]
  ogenes <- as.character(genes[!genes %in% pos.hum.mou[,"mou.gen"]])
  genes <- pos.hum.mou[,"mou.gen"][pos.hum.mou[,"mou.gen"] %in% genes]
  genes <- genes[order(as.numeric(pos.hum.mou[genes,"hum.pos"]))]
  mou.kmeans.genes.ord$genes[mou.spg.kmeans.genes$stages==i] = c(genes,ogenes)
}
```


```{r fig.width=4, fig.height=10}
cols <- c("#fabc3c",
"#a64dd4",
"#14e072",
"#d9014a",
"#abb4ff",
"#b28e00")
plot(1,xlim=c(1,2),ylim=c(0,1000),type="n",axes=FALSE,ann=FALSE)

for(i in c(1:dim(kmeans.hum.mou.genes)[1])){
#for(i in c(1:10)){ 
  x = which(hum.kmeans.genes.ord$genes == kmeans.hum.mou.genes$external_gene_name.x[i])
  y = which(mou.kmeans.genes.ord$genes == kmeans.hum.mou.genes$mmusculus_homolog_associated_gene_name[i])
  #print(kmeans.hum.mou.genes$external_gene_name.x[i])
  #print(kmeans.hum.mou.genes$mmusculus_homolog_associated_gene_name[i])
  x = 1000-x*1000/dim(hum.spg.kmeans.genes)[1]
  y = 1000-y*1000/dim(mou.spg.kmeans.genes)[1]
  #print(x)
  #print(y)
  segments(1,x,2,y,lwd = 0.1)
}
points(data.frame(x=rep(1,dim(hum.spg.kmeans.genes)[1]),y=1000-c(1:dim(hum.spg.kmeans.genes)[1])*1000/dim(hum.spg.kmeans.genes)[1]),col=cols[hum.spg.kmeans.genes$stages], pch=15, cex=0.8)
points(data.frame(x=rep(2,dim(mou.spg.kmeans.genes)[1]),y=1000-c(1:dim(mou.spg.kmeans.genes)[1])*1000/dim(mou.spg.kmeans.genes)[1]),col=cols[mou.spg.kmeans.genes$stages], pch=15, cex=0.8)
#segments(1,0,10,0,col = "red")
```

```{r fig.width=4, fig.height=10}
cols <- c("#fabc3c",
"#a64dd4",
"#14e072",
"#d9014a",
"#abb4ff",
"#b28e00")
ha = rowAnnotation(stages = hum.spg.kmeans.genes$stages, col = list(stages = c("1" = cols[1], "2" = cols[2], "3" = cols[3],"4" = cols[4], "5" = cols[5], "6" = cols[6])))
#ha = HeatmapAnnotation(foo = anno_simple(cbind(1:10, 10:1), pch = 1:10))
Heatmap(hum.spg.cen.std[as.character(hum.spg.kmeans.genes$genes),], cluster_rows = FALSE, cluster_columns = FALSE,show_row_names = FALSE, right_annotation = ha)
```

```{r fig.width=4, fig.height=10}
cols <- c("#fabc3c",
"#a64dd4",
"#14e072",
"#d9014a",
"#abb4ff",
"#b28e00")
ha = rowAnnotation(stages = mou.spg.kmeans.genes$stages, col = list(stages = c("1" = cols[1], "2" = cols[2], "3" = cols[3],"4" = cols[4], "5" = cols[5], "6" = cols[6])))
#ha = HeatmapAnnotation(foo = anno_simple(cbind(1:10, 10:1), pch = 1:10))
Heatmap(mou.spg.cen.std[as.character(mou.spg.kmeans.genes$genes),], cluster_rows = FALSE, cluster_columns = FALSE,show_row_names = FALSE, left_annotation = ha)
```

##Monkey Mou
#Connector lines
```{r}
kmeans.mac.mou.genes <- humMacMouOrth.1to1.in[humMacMouOrth.1to1.in$mmulatta_homolog_associated_gene_name %in% mac.spg.hvg,]
kmeans.mac.mou.genes <- kmeans.mac.mou.genes[kmeans.mac.mou.genes$mmusculus_homolog_associated_gene_name %in% mou.spg.hvg,]
```

```{r}
pos.mac.mou <- c()
for(i in c(1:dim(kmeans.mac.mou.genes)[1])){
  mac.gen <- kmeans.mac.mou.genes$mmulatta_homolog_associated_gene_name[i]
  mou.gen <- kmeans.mac.mou.genes$mmusculus_homolog_associated_gene_name[i]
  x = which(mac.spg.kmeans.genes$genes == mac.gen)
  y = which(mou.spg.kmeans.genes$genes == mou.gen)
  pos.mac.mou <- rbind(pos.mac.mou,c(x,y,mac.gen, mou.gen))
}
pos.mac.mou <- data.matrix(pos.mac.mou)
colnames(pos.mac.mou) = c("mac.pos","mou.pos", "mac.gen","mou.gen")
```

Sort within each cluster
```{r}

rownames(pos.mac.mou) <- pos.mac.mou[,"mac.gen"]
mac.kmeans.genes.ord <- mac.spg.kmeans.genes
for(i in c(1:6)){
  genes <- mac.spg.kmeans.genes$genes[mac.spg.kmeans.genes$stages==i]
  ogenes <- as.character(genes[!genes %in% pos.mac.mou[,"mac.gen"]])
  genes <- pos.mac.mou[,"mac.gen"][pos.mac.mou[,"mac.gen"] %in% genes]
  genes <- genes[order(as.numeric(pos.mac.mou[genes,"mou.pos"]))]
  mac.kmeans.genes.ord$genes[mac.spg.kmeans.genes$stages==i] = c(genes,ogenes)
}
```

```{r}
rownames(pos.mac.mou) <- pos.mac.mou[,"mou.gen"]
mou.kmeans.genes.ord <- mou.spg.kmeans.genes
for(i in c(1:6)){
  genes <- mou.spg.kmeans.genes$genes[mou.spg.kmeans.genes$stages==i]
  ogenes <- as.character(genes[!genes %in% pos.mac.mou[,"mou.gen"]])
  genes <- pos.mac.mou[,"mou.gen"][pos.mac.mou[,"mou.gen"] %in% genes]
  genes <- genes[order(as.numeric(pos.mac.mou[genes,"mac.pos"]))]
  mou.kmeans.genes.ord$genes[mou.spg.kmeans.genes$stages==i] = c(genes,ogenes)
}
```


```{r fig.width=4, fig.height=10}
cols <- c("#fabc3c",
"#a64dd4",
"#14e072",
"#d9014a",
"#abb4ff",
"#b28e00")
plot(1,xlim=c(1,2),ylim=c(0,1000),type="n",axes=FALSE,ann=FALSE)

for(i in c(1:dim(kmeans.mac.mou.genes)[1])){
#for(i in c(1:10)){ 
  x = which(mac.kmeans.genes.ord$genes == kmeans.mac.mou.genes$mmulatta_homolog_associated_gene_name[i])
  y = which(mou.kmeans.genes.ord$genes == kmeans.mac.mou.genes$mmusculus_homolog_associated_gene_name[i])
  #print(kmeans.hum.mou.genes$external_gene_name.x[i])
  #print(kmeans.hum.mou.genes$mmusculus_homolog_associated_gene_name[i])
  x = 1000-x*1000/dim(mac.spg.kmeans.genes)[1]
  y = 1000-y*1000/dim(mou.spg.kmeans.genes)[1]
  #print(x)
  #print(y)
  segments(1,x,2,y,lwd = 0.1)
}
points(data.frame(x=rep(1,dim(mac.spg.kmeans.genes)[1]),y=1000-c(1:dim(mac.spg.kmeans.genes)[1])*1000/dim(mac.spg.kmeans.genes)[1]),col=cols[mac.spg.kmeans.genes$stages], pch=15, cex=0.8)
points(data.frame(x=rep(2,dim(mou.spg.kmeans.genes)[1]),y=1000-c(1:dim(mou.spg.kmeans.genes)[1])*1000/dim(mou.spg.kmeans.genes)[1]),col=cols[mou.spg.kmeans.genes$stages], pch=15, cex=0.8)
#segments(1,0,10,0,col = "red")
```

```{r fig.width=4, fig.height=10}
cols <- c("#fabc3c",
"#a64dd4",
"#14e072",
"#d9014a",
"#abb4ff",
"#b28e00")
ha = rowAnnotation(stages = mou.spg.kmeans.genes$stages, col = list(stages = c("1" = cols[1], "2" = cols[2], "3" = cols[3],"4" = cols[4], "5" = cols[5], "6" = cols[6])))
#ha = HeatmapAnnotation(foo = anno_simple(cbind(1:10, 10:1), pch = 1:10))
Heatmap(mou.spg.cen.std[as.character(mou.spg.kmeans.genes$genes),], cluster_rows = FALSE, cluster_columns = FALSE,show_row_names = FALSE, left_annotation = ha)
```

```{r fig.width=4, fig.height=10}
cols <- c("#fabc3c",
"#a64dd4",
"#14e072",
"#d9014a",
"#abb4ff",
"#b28e00")
ha = rowAnnotation(stages = mac.spg.kmeans.genes$stages, col = list(stages = c("1" = cols[1], "2" = cols[2], "3" = cols[3],"4" = cols[4], "5" = cols[5], "6" = cols[6])))
#ha = HeatmapAnnotation(foo = anno_simple(cbind(1:10, 10:1), pch = 1:10))
Heatmap(mac.spg.cen.std[as.character(mac.spg.kmeans.genes$genes),], cluster_rows = FALSE, cluster_columns = FALSE,show_row_names = FALSE, right_annotation = ha)
```

Write the genes.
```{r}
orth.connect <- data.frame(humgen=humMacMouOrth.1to1.in$external_gene_name.x, macgen = humMacMouOrth.1to1.in$mmulatta_homolog_associated_gene_name, mougen = humMacMouOrth.1to1.in$mmusculus_homolog_associated_gene_name, humclu=rep(0, dim(humMacMouOrth.1to1.in)[1]),macclu=rep(0, dim(humMacMouOrth.1to1.in)[1]),mouclu=rep(0, dim(humMacMouOrth.1to1.in)[1]))
```

```{r}
rownames(orth.connect) = humMacMouOrth.1to1.in$external_gene_name.x
hum.ord = c(3,1,4,2)
for(i in c(1:4)){
  orth.connect[intersect(humMacMouOrth.1to1.in$external_gene_name.x,names(which(hum.spg.kmean$cluster==hum.ord[i]))),"humclu"] = i
}

rownames(orth.connect) = humMacMouOrth.1to1.in$mmulatta_homolog_associated_gene_name
mac.ord = c(2,4,3,1)
for(i in c(1:4)){
  orth.connect[intersect(humMacMouOrth.1to1.in$mmulatta_homolog_associated_gene_name,names(which(mac.spg.kmean$cluster==mac.ord[i]))),"macclu"] = i
}

rownames(orth.connect) = humMacMouOrth.1to1.in$mmusculus_homolog_associated_gene_name
mou.ord = c(3,1,4,2)
for(i in c(1:4)){
  orth.connect[intersect(humMacMouOrth.1to1.in$mmusculus_homolog_associated_gene_name,names(which(mou.spg.kmean$cluster==mou.ord[i]))),"mouclu"] = i
}
```

```{r}
write.table(orth.connect, file = "~/storage/HumanMacaqueMouseCompAnaysis/interdata/SPG_GeneKmeans/OrthologConnection.txt",quote=FALSE,row.names=FALSE,col.names = TRUE,sep="\t")
```

