#1. Read in the datasets
```{r}
spg.cca <- readRDS("~/storage/HumanMacaqueMouseCompAnaysis/interdata/spg.cca.rds")
```

```{r}
library(monocle)
dge.hum.raw2 <- readRDS("~/storage/HumanMacaqueMouseCompAnaysis/interdata/dge.hum.raw2.rds")
dge.mac.raw2 <- readRDS("~/storage/HumanMacaqueMouseCompAnaysis/interdata/dge.mac.raw2.rds")
dge.mou.raw2 <- readRDS("~/storage/HumanMacaqueMouseCompAnaysis/interdata/dge.mou.raw2.rds")
humMacMouOrth.1to1.in <- readRDS("~/storage/HumanMacaqueMouseCompAnaysis/interdata/humMacMouOrth.1to1.in.Rds")
```

###. Raw counts data for three species
```{r}
load("~/storage/HumanMacaqueMouseCompAnaysis/countsOriginalData/human_merged1235_20cpg15upg.robj")
dge.hum.raw <- dge@raw.data
rm(dge)
load("~/storage/HumanMacaqueMouseCompAnaysis/countsOriginalData/mac_merged9_nospikeincells.Robj")
dge.mac.raw <- dge@raw.data
rm(dge)
dge.mou.raw <- read.table("~/storage/HumanMacaqueMouseCompAnaysis/countsOriginalData/GSE112393_MergedAdultMouseST25_DGE.txt",header = TRUE, row.names = 1,sep="\t")
```

Move SPG clus 7 to germ and generate SPG clus 1-6 for three species.
```{r}
hum.spg.cells <- spg.cca@cell.names[spg.cca@meta.data$species=="Human" & spg.cca@meta.data$res.0.6.ord %in% c(1:6)]
mac.spg.cells <- spg.cca@cell.names[spg.cca@meta.data$species=="Monkey" & spg.cca@meta.data$res.0.6.ord %in% c(1:6)]
mou.spg.cells <- spg.cca@cell.names[spg.cca@meta.data$species=="Mouse" & spg.cca@meta.data$res.0.6.ord %in% c(1:6)]
```

```{r}
hum.germ.cells <- c(colnames(dge.hum.raw2), spg.cca@cell.names[spg.cca@meta.data$species=="Human" & spg.cca@meta.data$res.0.6.ord==8])
mac.germ.cells <- c(colnames(dge.mac.raw2), spg.cca@cell.names[spg.cca@meta.data$species=="Monkey" & spg.cca@meta.data$res.0.6.ord==8])
mou.germ.cells <- c(colnames(dge.mou.raw2), spg.cca@cell.names[spg.cca@meta.data$species=="Mouse" & spg.cca@meta.data$res.0.6.ord==8])
```

> length(hum.germ.cells)
[1] 4293
> length(mac.germ.cells)
[1] 12016
> length(mou.germ.cells)
[1] 18537
> length(hum.spg.cells)
[1] 1688
> length(mac.spg.cells)
[1] 747
> length(mou.spg.cells)
[1] 2174

###. Full dataset
```{r}
dge.hum.spg.raw <- dge.hum.raw[, hum.spg.cells]
dge.mac.spg.raw <- dge.mac.raw[, mac.spg.cells]
dge.mou.spg.raw <- dge.mou.raw[, mou.spg.cells]

dge.hum.germ.raw <- dge.hum.raw[, hum.germ.cells]
dge.mac.germ.raw <- dge.mac.raw[, mac.germ.cells]
dge.mou.germ.raw <- dge.mou.raw[, mou.germ.cells]
```


##Pre-runned monocle results
```{r}
mac.monocle <- readRDS("~/storage/HumanMacaqueMouseCompAnaysis/interdata/mac.new.cds.rds")
mac.cds <- mac.monocle$mac.cds
mac.ording_genes <- mac.monocle$mac.ording_genes
rm(mac.monocle)

hum.monocle <- readRDS("~/storage/HumanMacaqueMouseCompAnaysis/interdata/hum.new.cds.rds")
hum.cds <- hum.monocle$hum.cds
hum.ording_genes <- hum.monocle$hum.ording_genes
rm(hum.monocle)

mou.monocle <- readRDS("~/storage/HumanMacaqueMouseCompAnaysis/interdata/mou.new.cds.rds")
mou.cds <- mou.monocle$mou.cds
mou.ording_genes <- mou.monocle$mou.ording_genes
rm(mou.monocle)
```

###run monocle for human
####(Be careful when you need to rerun, may take a large memory and 1-2 hrs and crash your Rstudio)
```{r}
#Human
hum.cds <- newCellDataSet(as.matrix(dge.hum.germ.raw))
hum.cds <- estimateSizeFactors(hum.cds)
hum.cds <- estimateDispersions(hum.cds)
disp_table <- dispersionTable(hum.cds)
unsup_clustering_genes <- subset(disp_table, mean_expression >= 0.1)
hum.cds <- setOrderingFilter(hum.cds, unsup_clustering_genes$gene_id)
plot_ordering_genes(hum.cds)
hum.cds <- reduceDimension(hum.cds, max_components = 2,reduction_method = "tSNE", verbose = T)
hum.cds <- clusterCells(hum.cds, num_clusters = 10)
plot_cell_clusters(hum.cds)
pData(hum.cds)$Cluster <- factor(pData(hum.cds)$Cluster)
hum.cds.deg <- differentialGeneTest(hum.cds, fullModelFormulaStr = "~Cluster", cores = detectCores()-2)
hum.ording_genes <- rownames(hum.cds.deg)[order(hum.cds.deg$qval)][1:2000]
hum.cds <- setOrderingFilter(hum.cds, ordering_genes = c(hum.ording_genes))
hum.cds <- reduceDimension(hum.cds, max_components = 4,scaling = T,norm_method = "log", verbose = T)
hum.cds <- orderCells(hum.cds)
saveRDS(list(hum.cds = hum.cds,hum.ording_genes = hum.ording_genes), file = "~/storage/HumanMacaqueMouseCompAnaysis/interdata/hum.new.cds.rds", compress = FALSE)
plot_cell_trajectory(hum.cds, color_by = "Pseudotime")
```

###run monocle for monkey
```{r}
#Monkey
mac.cds <- newCellDataSet(as.matrix(dge.mac.germ.raw))
mac.cds <- estimateSizeFactors(mac.cds)
mac.cds <- estimateDispersions(mac.cds)
disp_table <- dispersionTable(mac.cds)
unsup_clustering_genes <- subset(disp_table, mean_expression >= 0.1)
mac.cds <- setOrderingFilter(mac.cds, unsup_clustering_genes$gene_id)
plot_ordering_genes(mac.cds)
mac.cds <- reduceDimension(mac.cds, max_components = 2,reduction_method = "tSNE", verbose = T)
mac.cds <- clusterCells(mac.cds, num_clusters = 10)
plot_cell_clusters(mac.cds)
pData(mac.cds)$Cluster <- factor(pData(mac.cds)$Cluster)
mac.cds.deg <- differentialGeneTest(mac.cds, fullModelFormulaStr = "~Cluster", cores = detectCores()-2)
mac.ording_genes <- rownames(mac.cds.deg)[order(mac.cds.deg$qval)][1:2000]
mac.cds <- setOrderingFilter(mac.cds, ordering_genes = c(mac.ording_genes))
mac.cds <- reduceDimension(mac.cds, max_components = 4,scaling = T,norm_method = "log", verbose = T)
mac.cds <- orderCells(mac.cds)
saveRDS(list(mac.cds = mac.cds, mac.ording_genes = mac.ording_genes), file = "~/storage/HumanMacaqueMouseCompAnaysis/interdata/mac.new.cds.rds", compress = FALSE)
plot_cell_trajectory(mac.cds, color_by = "Pseudotime")
```

###run monocle for mouse
```{r}
#Mouse
mou.cds <- newCellDataSet(as.matrix(dge.mou.germ.raw))
mou.cds <- estimateSizeFactors(mou.cds)
mou.cds <- estimateDispersions(mou.cds)
disp_table <- dispersionTable(mou.cds)
unsup_clustering_genes <- subset(disp_table, mean_expression >= 0.1)
mou.cds <- setOrderingFilter(mou.cds, unsup_clustering_genes$gene_id)
plot_ordering_genes(mou.cds)
mou.cds <- reduceDimension(mou.cds, max_components = 2,reduction_method = "tSNE", verbose = T)
mou.cds <- clusterCells(mou.cds, num_clusters = 10)
plot_cell_clusters(mou.cds)
pData(mou.cds)$Cluster <- factor(pData(mou.cds)$Cluster)
mou.cds.deg <- differentialGeneTest(mou.cds, fullModelFormulaStr = "~Cluster", cores = detectCores()-2)
mou.ording_genes <- rownames(mou.cds.deg)[order(mou.cds.deg$qval)][1:2000]
mou.cds <- setOrderingFilter(mou.cds, ordering_genes = c(mou.ording_genes))
mou.cds <- reduceDimension(mou.cds, max_components = 4,scaling = T,norm_method = "log", verbose = T)
mou.cds <- orderCells(mou.cds)
saveRDS(list(mou.cds = mou.cds,mou.ording_genes = mou.ording_genes), file = "~/storage/HumanMacaqueMouseCompAnaysis/interdata/mou.new.cds.rds", compress = FALSE)
plot_cell_trajectory(mou.cds, color_by = "Pseudotime")
```

###plot the trajectory from monocle
```{r fig.width=5,fig.height=5}
plot_cell_trajectory(hum.cds, color_by = "Pseudotime")
plot_cell_trajectory(mac.cds, color_by = "Pseudotime")
plot_cell_trajectory(mou.cds, color_by = "Pseudotime")
```

##plot the expression pattern of genes along the trajecotry to determine biologically meaningful direction.
```{r fig.width=5,fig.height=4}
plot_genes_in_pseudotime(mac.cds["HORMAD1",], color_by = "Pseudotime")
plot_genes_in_pseudotime(hum.cds["HORMAD1",], color_by = "Pseudotime")
plot_genes_in_pseudotime(mou.cds["Hormad1",], color_by = "Pseudotime")

plot_genes_in_pseudotime(hum.cds["PIWIL1",], color_by = "Pseudotime")
plot_genes_in_pseudotime(mac.cds["PIWIL1",], color_by = "Pseudotime")
plot_genes_in_pseudotime(mou.cds["Piwil1",], color_by = "Pseudotime")

plot_genes_in_pseudotime(hum.cds["ACRV1",], color_by = "Pseudotime")
plot_genes_in_pseudotime(mac.cds["ACRV1",], color_by = "Pseudotime")
plot_genes_in_pseudotime(mou.cds["Acrv1",], color_by = "Pseudotime")
```

```{r}
dge.hum.germ.norm <- apply(dge.hum.germ.raw,2,function(x) x/sum(x)*1e4)
dge.mac.germ.norm <- apply(dge.mac.germ.raw,2,function(x) x/sum(x)*1e4)
dge.mou.germ.norm <- apply(dge.mou.germ.raw,2,function(x) x/sum(x)*1e4)
```

Intersection of genes
```{r}
int.genes <- humMacMouOrth.1to1.in[humMacMouOrth.1to1.in$external_gene_name.x %in% hum.ording_genes,]
int.genes <- int.genes[int.genes$mmulatta_homolog_associated_gene_name %in% mac.ording_genes,]
int.genes <- int.genes[int.genes$mmusculus_homolog_associated_gene_name %in% mou.ording_genes,]
```

```{r}
dge.hum.germ.norm.int <- dge.hum.germ.norm[int.genes$external_gene_name.x,]
dge.mac.germ.norm.int <- dge.mac.germ.norm[int.genes$mmulatta_homolog_associated_gene_name,]
dge.mou.germ.norm.int <- dge.mou.germ.norm[int.genes$mmusculus_homolog_associated_gene_name,]
dge.hum.germ.lg.int <- log(dge.hum.germ.norm.int + 1)
dge.mac.germ.lg.int <- log(dge.mac.germ.norm.int + 1)
dge.mou.germ.lg.int <- log(dge.mou.germ.norm.int + 1)
```

### Run cellAlign
```{r}
library(cellAlign)
library(ComplexHeatmap)
library(ggplot2)
```

```{r}
numPts = 200
hum.traj <- as.numeric(hum.cds$Pseudotime)
hum.traj <- 1- hum.traj/max(hum.traj)
names(hum.traj) <- colnames(dge.hum.germ.lg.int)
align.hum <- cellAlign::interWeights(expDataBatch = dge.hum.germ.lg.int, trajCond = hum.traj, winSz = 0.1, numPts = numPts)

mac.traj <- as.numeric(mac.cds$Pseudotime)
mac.traj <- 1- mac.traj/max(mac.traj)
names(mac.traj) <- colnames(dge.mac.germ.lg.int)
align.mac <- cellAlign::interWeights(expDataBatch = dge.mac.germ.lg.int, trajCond = mac.traj, winSz = 0.1, numPts = numPts)

mou.traj <- as.numeric(mou.cds$Pseudotime)
mou.traj <- mou.traj/max(mou.traj)
names(mou.traj) <- colnames(dge.mou.germ.lg.int)
align.mou <- cellAlign::interWeights(expDataBatch = dge.mou.germ.lg.int, trajCond = mou.traj, winSz = 0.1, numPts = numPts)
```

```{r}
align.hum.scaled <- scaleInterpolate(align.hum)
align.mac.scaled <- scaleInterpolate(align.mac)
align.mou.scaled <- scaleInterpolate(align.mou)
```

Align: map monkey to human
```{r}
hummac.align <- globalAlign(align.mac.scaled$scaledData, 
                     align.hum.scaled$scaledData, 
                     scores = list(query = align.mac.scaled$traj, 
                                   ref = align.hum.scaled$traj), 
                     sigCalc = F, numPerm = 20)
plotAlign(hummac.align)
hummac.mapping <- mapRealDataGlobal(hummac.align, intTrajQuery = align.mac.scaled$traj, realTrajQuery = mac.traj, intTrajRef = align.hum.scaled$traj, realTrajRef = hum.traj)
plotMapping(hummac.mapping)
```

Align: map mouse to monkey
```{r}
macmou.align <- globalAlign(align.mou.scaled$scaledData, align.mac.scaled$scaledData, scores = list(query = align.mou.scaled$traj, ref = align.mac.scaled$traj), sigCalc = F, numPerm = 20)
plotAlign(macmou.align)
macmou.mapping <- mapRealDataGlobal(macmou.align, intTrajQuery = align.mou.scaled$traj, realTrajQuery = mou.traj, intTrajRef = align.mac.scaled$traj, realTrajRef = mac.traj)
plotMapping(macmou.mapping)
```

Align: map mouse to human
```{r}
hummou.align <- globalAlign(align.mou.scaled$scaledData, align.hum.scaled$scaledData, scores = list(query = align.mou.scaled$traj, ref = align.hum.scaled$traj), sigCalc = F, numPerm = 20)
plotAlign(hummou.align)
hummou.mapping <- mapRealDataGlobal(hummou.align, intTrajQuery = align.mou.scaled$traj, realTrajQuery = mou.traj, intTrajRef = align.hum.scaled$traj, realTrajRef = hum.traj)
plotMapping(hummou.mapping)
```

##. Calculate the centroids for the pseudotime points.
```{r}
hum.lens <- unlist(lapply(hummac.mapping$refAssign,length))
hum.200.cen <- matrix(0, nrow=dim(dge.hum.germ.norm)[1], ncol=length(hummac.mapping$refAssign))
for(i in c(1:dim(hum.200.cen)[2])){
  if(length(hummac.mapping$refAssign[[i]])==1){
    hum.200.cen[,i] <- dge.hum.germ.norm[,hummac.mapping$refAssign[[i]]]
  }else{
    hum.200.cen[,i] <- apply(dge.hum.germ.norm[,hummac.mapping$refAssign[[i]]],1,mean)
  }
}

mac.lens <- unlist(lapply(hummac.mapping$queryAssign,length))
mac.200.cen <- matrix(0, nrow=dim(dge.mac.germ.norm)[1], ncol=length(hummac.mapping$queryAssign))
for(i in c(1:dim(mac.200.cen)[2])){
  if(length(hummac.mapping$queryAssign[[i]])==1){
    mac.200.cen[,i] <- dge.mac.germ.norm[,hummac.mapping$queryAssign[[i]]]
  }else{
    mac.200.cen[,i] <- apply(dge.mac.germ.norm[,hummac.mapping$queryAssign[[i]]],1,mean)
  }
}

mou.lens <- unlist(lapply(hummou.mapping$queryAssign,length))
mou.200.cen <- matrix(0, nrow=dim(dge.mou.germ.norm)[1], ncol=length(hummou.mapping$queryAssign))
for(i in c(1:dim(mou.200.cen)[2])){
  if(length(hummou.mapping$queryAssign[[i]])==1){
    mou.200.cen[,i] <- dge.mou.germ.norm[,hummou.mapping$queryAssign[[i]]]
  }else{
    mou.200.cen[,i] <- apply(dge.mou.germ.norm[,hummou.mapping$queryAssign[[i]]],1,mean)
  }
}
```

```{r}
rownames(hum.200.cen) = rownames(dge.hum.germ.norm)
rownames(mac.200.cen) = rownames(dge.mac.germ.norm)
rownames(mou.200.cen) = rownames(dge.mou.germ.norm)
```

Centroids for SPG
```{r}
dge.hum.spg.norm <- apply(dge.hum.spg.raw,2,function(x) x/sum(x)*1e4)
dge.mac.spg.norm <- apply(dge.mac.spg.raw,2,function(x) x/sum(x)*1e4)
dge.mou.spg.norm <- apply(dge.mou.spg.raw,2,function(x) x/sum(x)*1e4)
```

```{r}
dge.hum.spg.cen <- apply(dge.hum.spg.norm,1,mean)
dge.mac.spg.cen <- apply(dge.mac.spg.norm,1,mean)
dge.mou.spg.cen <- apply(dge.mou.spg.norm,1,mean)

names(dge.hum.spg.cen) <- rownames(dge.hum.spg.norm)
names(dge.mac.spg.cen) <- rownames(dge.mac.spg.norm)
names(dge.mou.spg.cen) <- rownames(dge.mou.spg.norm)
```

Write the centroids to file
```{r}
write.table(cbind(dge.hum.spg.cen, hum.200.cen), file = "~/storage/HumanMacaqueMouseCompAnaysis/interdata/hum.201.centroid.txt", quote = FALSE, sep="\t",row.names = TRUE, col.names = FALSE)
write.table(cbind(dge.mac.spg.cen, mac.200.cen), file = "~/storage/HumanMacaqueMouseCompAnaysis/interdata/mac.201.centroid.txt", quote = FALSE, sep="\t",row.names = TRUE, col.names = FALSE)
write.table(cbind(dge.mou.spg.cen, mou.200.cen), file = "~/storage/HumanMacaqueMouseCompAnaysis/interdata/mou.201.centroid.txt", quote = FALSE, sep="\t",row.names = TRUE, col.names = FALSE)
```


### Merge the 201 centroids from each species, use the intersection of the HVGs to do PCA
```{r}
ca.cen.int.600 <- cbind(dge.hum.spg.cen[int.genes$external_gene_name.x],hum.200.cen[int.genes$external_gene_name.x,], dge.mac.spg.cen[int.genes$mmulatta_homolog_associated_gene_name], mac.200.cen[int.genes$mmulatta_homolog_associated_gene_name,], dge.mou.spg.cen[int.genes$mmusculus_homolog_associated_gene_name], mou.200.cen[int.genes$mmusculus_homolog_associated_gene_name,])

colnames(ca.cen.int.600) <- c(rep("Human",201),rep("Monkey",201),rep("Mouse",201))
ca.cen.int.600.norm <- apply(ca.cen.int.600,2,function(x) x/sum(x)*1000)
ca.cen.int.600.pca <- prcomp(t(ca.cen.int.600.norm), center = TRUE, scale. = TRUE)
```

```{r fig.width=7,fig.height=5}
ggplot(data.frame(ca.cen.int.600.pca$x, species=factor(colnames(ca.cen.int.600)), num=c(dim(dge.hum.spg.raw)[2],hum.lens, dim(dge.mac.spg.raw)[2],mac.lens, dim(dge.mou.spg.raw)[2], mou.lens),traj=c(2,rep(1,200),rep(c(2,rep(1,200)),2))), aes(PC1,PC2)) + geom_point(aes(colour=species, shape=factor(traj), size=num/100))+ scale_color_manual(values = c("#e35a97","#5a97e3","#5ae3a6"))
ggplot(data.frame(ca.cen.int.600.pca$x, species=factor(colnames(ca.cen.int.600)), num=c(dim(dge.hum.spg.raw)[2],hum.lens, dim(dge.mac.spg.raw)[2],mac.lens, dim(dge.mou.spg.raw)[2], mou.lens),traj=c(2,rep(1,200),rep(c(2,rep(1,200)),2))), aes(PC1,PC3)) + geom_point(aes(colour=species, shape=factor(traj), size=num/100))+ scale_color_manual(values = c("#e35a97","#5a97e3","#5ae3a6"))
```

Redo PCA using top 30 PCs
```{r fig.width=7,fig.height=5}
ca.cen.int.600.repca <- prcomp(ca.cen.int.600.pca$x[,c(1:3)], center = TRUE, scale. = TRUE)
ggplot(data.frame(ca.cen.int.600.repca$x, species=factor(colnames(ca.cen.int.600)), num=c(dim(dge.hum.spg.norm)[2],hum.lens, dim(dge.mac.spg.norm)[2],mac.lens, dim(dge.mou.spg.norm)[2], mou.lens),traj=c(2,rep(1,200),rep(c(2,rep(1,200)),2))), aes(PC1,PC2)) + geom_point(aes(colour=species, shape=factor(traj), size=num/100))+ scale_color_manual(values = c("#e35a97","#5a97e3","#5ae3a6"))
ggplot(data.frame(ca.cen.int.600.repca$x, species=factor(colnames(ca.cen.int.600)), num=c(dim(dge.hum.spg.norm)[2],hum.lens, dim(dge.mac.spg.norm)[2],mac.lens, dim(dge.mou.spg.norm)[2], mou.lens),traj=c(2,rep(1,200),rep(c(2,rep(1,200)),2))), aes(PC1,PC3)) + geom_point(aes(colour=species, shape=factor(traj), size=num/100))+ scale_color_manual(values = c("#e35a97","#5a97e3","#5ae3a6"))
```


##Orthologs gene expression score

###Calculate (the ratio of the genes to be present in orthologs)/(the ratio of the genes to be present in other genes)

```{r}
calcvperc <- function(genelist, normdata, spgdata, maplist){
  perc <- c()
  othergen <- setdiff(rownames(normdata), genelist)
  for(i in c(1:200)){
    if(length(maplist[[i]])==1){
      perc <- c(perc, (sum(normdata[genelist,maplist[[i]]]>0)/length(genelist))/(sum(normdata[othergen,maplist[[i]]]>0)/length(othergen)))
    }else{
      percvec <- lapply(maplist[[i]],function(x) (sum(normdata[genelist,x]>0)/length(genelist))/(sum(normdata[othergen,x]>0)/length(othergen)))
      perc <- c(perc,mean(unlist(percvec)))
    }
  }
  percvec <- lapply(colnames(spgdata),function(x) (sum(spgdata[genelist,x]>0)/length(genelist))/(sum(spgdata[othergen,x]>0)/length(othergen)))
  perc <- c(perc,mean(unlist(percvec)))
  perc
}
```

###Exploratory analysis (line 357-740)
1. 1-1-1 orthologs
```{r fig.width=7,fig.height=5}
hum.201.cen <- cbind(dge.hum.spg.cen,hum.200.cen)
hum.111orth.perc <- apply(hum.201.cen[humMacMouOrth.1to1.in$external_gene_name.x,],2,sum)/apply(hum.201.cen,2,sum)

mac.201.cen <- cbind(dge.mac.spg.cen,mac.200.cen)
mac.111orth.perc <- apply(mac.201.cen[humMacMouOrth.1to1.in$mmulatta_homolog_associated_gene_name,],2,sum)/apply(mac.201.cen,2,sum)

mou.201.cen <- cbind(dge.mou.spg.cen,mou.200.cen)
mou.111orth.perc <- apply(mou.201.cen[humMacMouOrth.1to1.in$mmusculus_homolog_associated_gene_name,],2,sum)/apply(mou.201.cen,2,sum)

ggplot(data.frame(ca.cen.int.600.pca$x, orth=c(hum.111orth.perc, mac.111orth.perc, mou.111orth.perc), num=c(dim(dge.hum.spg.raw)[2],hum.lens, dim(dge.mac.spg.raw)[2],mac.lens, dim(dge.mou.spg.raw)[2], mou.lens),traj=c(2,rep(1,200),rep(c(2,rep(1,200)),2))), aes(PC1,PC3)) + geom_point(aes(colour=orth, shape=factor(traj), size=num/100)) + scale_color_gradientn(colours = c("#0000b3","#ededed","#b30000")) 
```

Separate plotting
```{r fig.width=7,fig.height=5}
dat = data.frame(ca.cen.int.600.pca$x, orth=c(hum.111orth.perc, rep(NA,201), rep(NA,201)), num=c(dim(dge.hum.spg.raw)[2],hum.lens, dim(dge.mac.spg.raw)[2],mac.lens, dim(dge.mou.spg.raw)[2], mou.lens),traj=c(2,rep(1,200),rep(c(2,rep(1,200)),2)))

ggplot(dat[seq(dim(dat)[1],1),], aes(PC1,PC3)) + geom_point(aes(colour=orth, shape=factor(traj), size=num)) + scale_color_gradientn(colours = c("#ededed","#b30000"))

dat = data.frame(ca.cen.int.600.pca$x, orth=c(rep(NA,201), mac.111orth.perc, rep(NA,201)), num=c(dim(dge.hum.spg.raw)[2],hum.lens, dim(dge.mac.spg.raw)[2],mac.lens, dim(dge.mou.spg.raw)[2], mou.lens),traj=c(2,rep(1,200),rep(c(2,rep(1,200)),2)))

ggplot(dat[seq(dim(dat)[1],1),], aes(PC1,PC3)) + geom_point(aes(colour=orth, shape=factor(traj), size=num)) + scale_color_gradientn(colours = c("#ededed","#b30000"))

dat = data.frame(ca.cen.int.600.pca$x, orth=c(rep(NA,201), rep(NA,201), mou.111orth.perc), num=c(dim(dge.hum.spg.raw)[2],hum.lens, dim(dge.mac.spg.raw)[2],mac.lens, dim(dge.mou.spg.raw)[2], mou.lens),traj=c(2,rep(1,200),rep(c(2,rep(1,200)),2)))

ggplot(dat[seq(dim(dat)[1],1),], aes(PC1,PC3)) + geom_point(aes(colour=orth, shape=factor(traj), size=num)) + scale_color_gradientn(colours = c("#ededed","#b30000"))
```

Plot by three rows
```{r fig.width=7,fig.height=5}
ggplot(data.frame(stages=c(1:201), orth=c(hum.111orth.perc, mac.111orth.perc, mou.111orth.perc), num=c(dim(dge.hum.spg.raw)[2],hum.lens, dim(dge.mac.spg.raw)[2],mac.lens, dim(dge.mou.spg.raw)[2], mou.lens),traj=c(2,rep(1,200),rep(c(2,rep(1,200)),2)), spe=c(rep("Human",201),rep("Monkey",201),rep("Mouse",201))), aes(stages,orth)) + geom_point(aes(colour=spe, shape=factor(traj), size=num/100)) + scale_color_manual(values = c("#e35a97","#5a97e3","#5ae3a6")) + xlab("Pseudo points") + ylab("Percentage of total expression") + ggtitle("1-1-1 Orthologs")
```


## Calculate gene detection rate in each cluster.
```{r}
hum.200.gperc <- matrix(0, nrow=dim(dge.hum.germ.norm)[1], ncol=length(hummac.mapping$refAssign))
for(i in c(1:dim(hum.200.gperc)[2])){
  if(length(hummac.mapping$refAssign[[i]])==1){
    hum.200.gperc[,i] <- as.numeric(dge.hum.germ.norm[,hummac.mapping$refAssign[[i]]]>0)
  }else{
    hum.200.gperc[,i] <- apply(dge.hum.germ.norm[,hummac.mapping$refAssign[[i]]],1,function(x) mean(x>0))
  }
}

mac.200.gperc <- matrix(0, nrow=dim(dge.mac.germ.norm)[1], ncol=length(hummac.mapping$queryAssign))
for(i in c(1:dim(mac.200.gperc)[2])){
  if(length(hummac.mapping$queryAssign[[i]])==1){
    mac.200.gperc[,i] <- as.numeric(dge.mac.germ.norm[,hummac.mapping$queryAssign[[i]]]>0)
  }else{
    mac.200.gperc[,i] <- apply(dge.mac.germ.norm[,hummac.mapping$queryAssign[[i]]],1,function(x) mean(x>0))
  }
}

mou.200.gperc <- matrix(0, nrow=dim(dge.mou.germ.norm)[1], ncol=length(hummou.mapping$queryAssign))
for(i in c(1:dim(mou.200.gperc)[2])){
  if(length(hummou.mapping$queryAssign[[i]])==1){
    mou.200.gperc[,i] <- as.numeric(dge.mou.germ.norm[,hummou.mapping$queryAssign[[i]]]>0)
  }else{
    mou.200.gperc[,i] <- apply(dge.mou.germ.norm[,hummou.mapping$queryAssign[[i]]],1,function(x) mean(x>0))
  }
}
```

```{r}
hum.201.gper <- cbind(apply(dge.hum.spg.raw,1,function(x) mean(x>0)),hum.200.gperc)
mac.201.gper <- cbind(apply(dge.mac.spg.raw,1,function(x) mean(x>0)),mac.200.gperc)
mou.201.gper <- cbind(apply(dge.mou.spg.raw,1,function(x) mean(x>0)),mou.200.gperc)
rownames(hum.201.gper) <- rownames(dge.hum.spg.raw)
rownames(mac.201.gper) <- rownames(dge.mac.spg.raw)
rownames(mou.201.gper) <- rownames(dge.mou.spg.raw)
```



```{r fig.width=7,fig.height=5}
hum.111orth.perc <- apply(hum.201.gper[humMacMouOrth.1to1.in$external_gene_name.x,],2,function(x) sum(x>0.05))/apply(hum.201.gper,2,function(x) sum(x>0.05))

mac.111orth.perc <- apply(mac.201.gper[humMacMouOrth.1to1.in$mmulatta_homolog_associated_gene_name,],2,function(x) sum(x>0.05))/apply(mac.201.gper,2,function(x) sum(x>0.05))

mou.111orth.perc <- apply(mou.201.gper[humMacMouOrth.1to1.in$mmusculus_homolog_associated_gene_name,],2,function(x) sum(x>0.05))/apply(mou.201.gper,2,function(x) sum(x>0.05))

ggplot(data.frame(ca.cen.int.600.pca$x, orth=c(hum.111orth.perc, mac.111orth.perc, mou.111orth.perc), num=c(dim(dge.hum.spg.raw)[2],hum.lens, dim(dge.mac.spg.raw)[2],mac.lens, dim(dge.mou.spg.raw)[2], mou.lens),traj=c(2,rep(1,200),rep(c(2,rep(1,200)),2))), aes(PC1,PC3)) + geom_point(aes(colour=orth, shape=factor(traj), size=num)) + scale_color_gradientn(colours = c("#0000b3","#ededed","#b30000")) 
```

```{r fig.width=7,fig.height=5}
dat <- data.frame(ca.cen.int.600.pca$x, orth=c(hum.111orth.perc, rep(NA,201), rep(NA,201)), num=c(dim(dge.hum.spg.raw)[2],hum.lens, dim(dge.mac.spg.raw)[2],mac.lens, dim(dge.mou.spg.raw)[2], mou.lens),traj=c(2,rep(1,200),rep(c(2,rep(1,200)),2)))
ggplot(dat[seq(dim(dat)[1],1),], aes(PC1,PC3)) + geom_point(aes(colour=orth, shape=factor(traj), size=num)) + scale_color_gradientn(colours = c("#ededed","#b30000")) 

dat <- data.frame(ca.cen.int.600.pca$x, orth=c(rep(NA,201), mac.111orth.perc, rep(NA,201)), num=c(dim(dge.hum.spg.raw)[2],hum.lens, dim(dge.mac.spg.raw)[2],mac.lens, dim(dge.mou.spg.raw)[2], mou.lens),traj=c(2,rep(1,200),rep(c(2,rep(1,200)),2)))
ggplot(dat, aes(PC1,PC3)) + geom_point(aes(colour=orth, shape=factor(traj), size=num)) + scale_color_gradientn(colours = c("#ededed","#b30000")) 

dat <- data.frame(ca.cen.int.600.pca$x, orth=c(rep(NA,201), rep(NA,201), mou.111orth.perc), num=c(dim(dge.hum.spg.raw)[2],hum.lens, dim(dge.mac.spg.raw)[2],mac.lens, dim(dge.mou.spg.raw)[2], mou.lens),traj=c(2,rep(1,200),rep(c(2,rep(1,200)),2)))
ggplot(dat, aes(PC1,PC3)) + geom_point(aes(colour=orth, shape=factor(traj), size=num)) + scale_color_gradientn(colours = c("#ededed","#b30000")) 
```

```{r}
library(circlize)
col_func = colorRamp2(c(0,0.3,1),c("blue","grey","red"))
```


###Do Heatmap
```{r fig.width=5, fig.height=7}
hum.orths <- humMacMouOrth.1to1.in$external_gene_name.x[order(apply(hum.201.gper[humMacMouOrth.1to1.in$external_gene_name.x,],1,sum), decreasing = TRUE)]

hum.nonorthos <- setdiff(rownames(hum.201.gper),humMacMouOrth.1to1.in$external_gene_name.x)
hum.nonorthos <- hum.nonorthos[order(apply(hum.201.gper[hum.nonorthos,],1,sum), decreasing = TRUE)]

hm = rowAnnotation(df = data.frame(orth=c(rep("Ortho",length(hum.orths)),rep("NonOrtho",length(hum.nonorthos)))), col=list(orth=c("Ortho"="red","NonOrtho"="blue")))

Heatmap(hum.201.gper[c(hum.orths,hum.nonorthos),], cluster_rows = FALSE, cluster_columns = FALSE, show_row_names = FALSE, show_column_names = FALSE, col = col_func, left_annotation = hm)
```

```{r fig.width=5, fig.height=7}
mac.orths <- humMacMouOrth.1to1.in$mmulatta_homolog_associated_gene_name[order(apply(mac.201.gper[humMacMouOrth.1to1.in$mmulatta_homolog_associated_gene_name,],1,sum), decreasing = TRUE)]

mac.nonorthos <- setdiff(rownames(mac.201.gper),humMacMouOrth.1to1.in$mmulatta_homolog_associated_gene_name)
mac.nonorthos <- mac.nonorthos[order(apply(mac.201.gper[mac.nonorthos,],1,sum), decreasing = TRUE)]

hm = rowAnnotation(df = data.frame(orth=c(rep("Ortho",length(mac.orths)),rep("NonOrtho",length(mac.nonorthos)))), col=list(orth=c("Ortho"="red","NonOrtho"="blue")))
Heatmap(mac.201.gper[c(mac.orths,mac.nonorthos),], cluster_rows = FALSE, cluster_columns = FALSE, show_row_names = FALSE, show_column_names = FALSE, col = col_func, left_annotation = hm)
```

```{r fig.width=5, fig.height=7}
mou.orths <- humMacMouOrth.1to1.in$mmusculus_homolog_associated_gene_name[order(apply(mou.201.gper[humMacMouOrth.1to1.in$mmusculus_homolog_associated_gene_name,],1,sum), decreasing = TRUE)]

mou.nonorthos <- setdiff(rownames(mou.201.gper),humMacMouOrth.1to1.in$mmusculus_homolog_associated_gene_name)
mou.nonorthos <- mou.nonorthos[order(apply(mou.201.gper[mou.nonorthos,],1,sum), decreasing = TRUE)]

hm = rowAnnotation(df = data.frame(orth=c(rep("Ortho",length(mou.orths)),rep("NonOrtho",length(mou.nonorthos)))), col=list(orth=c("Ortho"="red","NonOrtho"="blue")))
Heatmap(mou.201.gper[c(mou.orths,mou.nonorthos),], cluster_rows = FALSE, cluster_columns = FALSE, show_row_names = FALSE, show_column_names = FALSE, col = col_func, left_annotation = hm)
```

###Calculate per cell and then average the percentage.
```{r}
hum.200.111orth.vperc <- c()
for(i in c(1:200)){
  if(length(hummac.mapping$refAssign[[i]])==1){
    hum.200.111orth.vperc <- c(hum.200.111orth.vperc, sum(dge.hum.germ.norm[hum.orths,hummac.mapping$refAssign[[i]]]>0)/sum(dge.hum.germ.norm[,hummac.mapping$refAssign[[i]]]>0))
  }else{
    percvec <- lapply(hummac.mapping$refAssign[[i]],function(x) sum(dge.hum.germ.norm[hum.orths,x]>0)/sum(dge.hum.germ.norm[,x]>0))
    hum.200.111orth.vperc <- c(hum.200.111orth.vperc,mean(unlist(percvec)))
  }
}

mac.200.111orth.vperc <- c()
for(i in c(1:200)){
  if(length(hummac.mapping$queryAssign[[i]])==1){
    mac.200.111orth.vperc <- c(mac.200.111orth.vperc, sum(dge.mac.germ.norm[mac.orths,hummac.mapping$queryAssign[[i]]]>0)/sum(dge.mac.germ.norm[,hummac.mapping$queryAssign[[i]]]>0))
  }else{
    percvec <- lapply(hummac.mapping$queryAssign[[i]],function(x) sum(dge.mac.germ.norm[mac.orths,x]>0)/sum(dge.mac.germ.norm[,x]>0))
    mac.200.111orth.vperc <- c(mac.200.111orth.vperc,mean(unlist(percvec)))
  }
}

mou.200.111orth.vperc <- c()
for(i in c(1:200)){
  if(length(hummou.mapping$refAssign[[i]])==1){
    mou.200.111orth.vperc <- c(mou.200.111orth.vperc, sum(dge.mou.germ.norm[mou.orths,hummou.mapping$queryAssign[[i]]]>0)/sum(dge.mou.germ.norm[,hummou.mapping$queryAssign[[i]]]>0))
  }else{
    percvec <- lapply(hummou.mapping$queryAssign[[i]],function(x) sum(dge.mou.germ.norm[mou.orths,x]>0)/sum(dge.mou.germ.norm[,x]>0))
    mou.200.111orth.vperc <- c(mou.200.111orth.vperc,mean(unlist(percvec)))
  }
}
```

```{r}
percvec <- lapply(colnames(dge.hum.spg.norm),function(x) sum(dge.hum.spg.norm[hum.orths,x]>0)/sum(dge.hum.spg.norm[,x]>0))
hum.201.111orth.vperc <- c(hum.200.111orth.vperc,mean(unlist(percvec)))
    
percvec <- lapply(colnames(dge.mac.spg.norm),function(x) sum(dge.mac.spg.norm[mac.orths,x]>0)/sum(dge.mac.spg.norm[,x]>0))
mac.201.111orth.vperc <- c(mac.200.111orth.vperc,mean(unlist(percvec)))

percvec <- lapply(colnames(dge.mou.spg.norm),function(x) sum(dge.mou.spg.norm[mou.orths,x]>0)/sum(dge.mou.spg.norm[,x]>0))
mou.201.111orth.vperc <- c(mou.200.111orth.vperc,mean(unlist(percvec)))
```


Plot by three rows
```{r fig.width=7,fig.height=5}
ggplot(data.frame(stages=c(1:201), orth=c(hum.201.111orth.vperc, mac.201.111orth.vperc, mou.201.111orth.vperc), num=c(dim(dge.hum.spg.raw)[2],hum.lens, dim(dge.mac.spg.raw)[2],mac.lens, dim(dge.mou.spg.raw)[2], mou.lens),traj=c(2,rep(1,200),rep(c(2,rep(1,200)),2)), spe=c(rep("Human",201),rep("Monkey",201),rep("Mouse",201))), aes(stages,orth)) + geom_point(aes(colour=spe, shape=factor(traj), size=num/100)) + scale_color_manual(values = c("#e35a97","#5a97e3","#5ae3a6")) + xlab("Pseudo points") + ylab("Percentage of expressed genes") + ggtitle("1-1-1 Orthologs")
```

Median instead of mean
```{r}
hum.200.111orth.vperc <- c()
for(i in c(1:200)){
  if(length(hummac.mapping$refAssign[[i]])==1){
    hum.200.111orth.vperc <- c(hum.200.111orth.vperc, sum(dge.hum.germ.norm[hum.orths,hummac.mapping$refAssign[[i]]]>0)/sum(dge.hum.germ.norm[,hummac.mapping$refAssign[[i]]]>0))
  }else{
    percvec <- lapply(hummac.mapping$refAssign[[i]],function(x) sum(dge.hum.germ.norm[hum.orths,x]>0)/sum(dge.hum.germ.norm[,x]>0))
    hum.200.111orth.vperc <- c(hum.200.111orth.vperc,median(unlist(percvec)))
  }
}

mac.200.111orth.vperc <- c()
for(i in c(1:200)){
  if(length(hummac.mapping$queryAssign[[i]])==1){
    mac.200.111orth.vperc <- c(mac.200.111orth.vperc, sum(dge.mac.germ.norm[mac.orths,hummac.mapping$queryAssign[[i]]]>0)/sum(dge.mac.germ.norm[,hummac.mapping$queryAssign[[i]]]>0))
  }else{
    percvec <- lapply(hummac.mapping$queryAssign[[i]],function(x) sum(dge.mac.germ.norm[mac.orths,x]>0)/sum(dge.mac.germ.norm[,x]>0))
    mac.200.111orth.vperc <- c(mac.200.111orth.vperc,median(unlist(percvec)))
  }
}

mou.200.111orth.vperc <- c()
for(i in c(1:200)){
  if(length(hummou.mapping$refAssign[[i]])==1){
    mou.200.111orth.vperc <- c(mou.200.111orth.vperc, sum(dge.mou.germ.norm[mou.orths,hummou.mapping$queryAssign[[i]]]>0)/sum(dge.mou.germ.norm[,hummou.mapping$queryAssign[[i]]]>0))
  }else{
    percvec <- lapply(hummou.mapping$queryAssign[[i]],function(x) sum(dge.mou.germ.norm[mou.orths,x]>0)/sum(dge.mou.germ.norm[,x]>0))
    mou.200.111orth.vperc <- c(mou.200.111orth.vperc,median(unlist(percvec)))
  }
}
```

```{r}
percvec <- lapply(colnames(dge.hum.spg.norm),function(x) sum(dge.hum.spg.norm[hum.orths,x]>0)/sum(dge.hum.spg.norm[,x]>0))
hum.201.111orth.vperc <- c(hum.200.111orth.vperc,median(unlist(percvec)))
    
percvec <- lapply(colnames(dge.mac.spg.norm),function(x) sum(dge.mac.spg.norm[mac.orths,x]>0)/sum(dge.mac.spg.norm[,x]>0))
mac.201.111orth.vperc <- c(mac.200.111orth.vperc,median(unlist(percvec)))

percvec <- lapply(colnames(dge.mou.spg.norm),function(x) sum(dge.mou.spg.norm[mou.orths,x]>0)/sum(dge.mou.spg.norm[,x]>0))
mou.201.111orth.vperc <- c(mou.200.111orth.vperc,median(unlist(percvec)))
```


Plot by three rows
```{r fig.width=7,fig.height=5}
ggplot(data.frame(stages=c(1:201), orth=c(hum.201.111orth.vperc, mac.201.111orth.vperc, mou.201.111orth.vperc), num=c(dim(dge.hum.spg.raw)[2],hum.lens, dim(dge.mac.spg.raw)[2],mac.lens, dim(dge.mou.spg.raw)[2], mou.lens),traj=c(2,rep(1,200),rep(c(2,rep(1,200)),2)), spe=c(rep("Human",201),rep("Monkey",201),rep("Mouse",201))), aes(stages,orth)) + geom_point(aes(colour=spe, shape=factor(traj), size=num/100)) + scale_color_manual(values = c("#e35a97","#5a97e3","#5ae3a6")) + xlab("Pseudo points") + ylab("Percentage of expressed genes") + ggtitle("1-1-1 Orthologs")
```

##Calculate the ratio of the median/mean for each clusters.
```{r fig.width=7,fig.height=5}
hum.201.111orth.hperc <- apply(hum.201.gper[humMacMouOrth.1to1.in$external_gene_name.x,],2,mean)/apply(hum.201.gper,2,mean)

mac.201.111orth.hperc <- apply(mac.201.gper[humMacMouOrth.1to1.in$mmulatta_homolog_associated_gene_name,],2,mean)/apply(mac.201.gper,2,mean)

mou.201.111orth.hperc <- apply(mou.201.gper[humMacMouOrth.1to1.in$mmusculus_homolog_associated_gene_name,],2,mean)/apply(mou.201.gper,2,mean)

ggplot(data.frame(stages=c(1:201), orth=c(hum.201.111orth.hperc, mac.201.111orth.hperc, mou.201.111orth.hperc), num=c(dim(dge.hum.spg.raw)[2],hum.lens, dim(dge.mac.spg.raw)[2],mac.lens, dim(dge.mou.spg.raw)[2], mou.lens),traj=c(2,rep(1,200),rep(c(2,rep(1,200)),2)), spe=c(rep("Human",201),rep("Monkey",201),rep("Mouse",201))), aes(stages,orth)) + geom_point(aes(colour=spe, shape=factor(traj), size=num/100)) + scale_color_manual(values = c("#e35a97","#5a97e3","#5ae3a6")) + xlab("Pseudo points") + ylab("Ratio of the detection rate") + ggtitle("1-1-1 Orthologs")
```

##Median is not a good thing to use because we have too many genes showed ~0 detection rate.
```{r fig.width=7,fig.height=5}
hum.201.111orth.hperc <- apply(hum.201.gper[humMacMouOrth.1to1.in$external_gene_name.x,],2,median)/apply(hum.201.gper,2,median)

mac.201.111orth.hperc <- apply(mac.201.gper[humMacMouOrth.1to1.in$mmulatta_homolog_associated_gene_name,],2,median)/apply(mac.201.gper,2,median)

mou.201.111orth.hperc <- apply(mou.201.gper[humMacMouOrth.1to1.in$mmusculus_homolog_associated_gene_name,],2,median)/apply(mou.201.gper,2,median)

ggplot(data.frame(stages=c(1:201), orth=c(hum.201.111orth.hperc, mac.201.111orth.hperc, mou.201.111orth.hperc), num=c(dim(dge.hum.spg.raw)[2],hum.lens, dim(dge.mac.spg.raw)[2],mac.lens, dim(dge.mou.spg.raw)[2], mou.lens),traj=c(2,rep(1,200),rep(c(2,rep(1,200)),2)), spe=c(rep("Human",201),rep("Monkey",201),rep("Mouse",201))), aes(stages,orth)) + geom_point(aes(colour=spe, shape=factor(traj), size=num/100)) + scale_color_manual(values = c("#e35a97","#5a97e3","#5ae3a6")) + xlab("Pseudo points") + ylab("Ratio of the detection rate") + ggtitle("1-1-1 Orthologs")
```

```{r}
hum.spg.mean <- apply(dge.hum.spg.norm,1,mean)
hum.spg.per0 <- apply(dge.hum.spg.norm,1,function(x) mean(x>0))
hum.spg.cv <- apply(dge.hum.spg.norm,1,function(x) sqrt(var(x))/mean(x))
hum.germ.mean <- apply(dge.hum.germ.norm,1,mean)
hum.germ.per0 <- apply(dge.hum.germ.norm,1,function(x) mean(x>0))
hum.germ.cv <- apply(dge.hum.germ.norm,1,function(x) sqrt(var(x))/mean(x))
```

```{r}
hum.germ.f10.norm <- c()
hum.germ.b10.norm <- c()
hum.germ.m10.norm <- c()
for(i in c(1:10)){
  hum.germ.f10.norm <- cbind(hum.germ.f10.norm, dge.hum.germ.norm[, hummac.mapping$refAssign[[i]]])
  hum.germ.b10.norm <- cbind(hum.germ.b10.norm, dge.hum.germ.norm[, hummac.mapping$refAssign[[201-i]]])
  hum.germ.m10.norm <- cbind(hum.germ.m10.norm, dge.hum.germ.norm[, hummac.mapping$refAssign[[95+i]]])
}
hum.germ.f10.mean <- apply(hum.germ.f10.norm,1,mean)
hum.germ.f10.per0 <- apply(hum.germ.f10.norm,1,function(x) mean(x>0))
hum.germ.f10.cv <- apply(hum.germ.f10.norm,1,function(x) sqrt(var(x))/mean(x))

hum.germ.b10.mean <- apply(hum.germ.b10.norm,1,mean)
hum.germ.b10.per0 <- apply(hum.germ.b10.norm,1,function(x) mean(x>0))
hum.germ.b10.cv <- apply(hum.germ.b10.norm,1,function(x) sqrt(var(x))/mean(x))

hum.germ.m10.mean <- apply(hum.germ.m10.norm,1,mean)
hum.germ.m10.per0 <- apply(hum.germ.m10.norm,1,function(x) mean(x>0))
hum.germ.m10.cv <- apply(hum.germ.m10.norm,1,function(x) sqrt(var(x))/mean(x))
```


```{r fig.height=5,fig.width=6}
plot(log(hum.spg.mean),hum.spg.per0,pch=16,cex=0.5,xlab="Mean expression", ylab="Percentage of NonZero")
points(log(hum.germ.mean),hum.germ.per0, col="red",pch=16,cex=0.5)

plot(log(hum.germ.b10.mean),hum.germ.b10.per0,col="grey",pch=16,cex=0.5,xlab="Mean expression", ylab="Percentage of NonZero")
points(log(hum.germ.m10.mean),hum.germ.m10.per0,col="grey",pch=16,cex=0.5)
points(log(hum.germ.f10.mean),hum.germ.f10.per0,col="yellow",pch=16,cex=0.5)

plot(log(hum.germ.f10.mean),hum.germ.f10.per0,col="grey",pch=16,cex=0.5,xlab="Mean expression", ylab="Percentage of NonZero")
points(log(hum.germ.b10.mean),hum.germ.b10.per0,col="grey",pch=16,cex=0.5)
points(log(hum.germ.m10.mean),hum.germ.m10.per0,col="green",pch=16,cex=0.5)

plot(log(hum.germ.f10.mean),hum.germ.f10.per0,col="grey",pch=16,cex=0.5,xlab="Mean expression", ylab="Percentage of NonZero")
points(log(hum.germ.m10.mean),hum.germ.m10.per0,col="grey",pch=16,cex=0.5)
points(log(hum.germ.b10.mean),hum.germ.b10.per0,col="blue",pch=16,cex=0.5)


plot(log(hum.spg.mean),log(hum.spg.cv),pch=16,cex=0.5,xlab="Mean expression", ylab="Percentage of NonZero")
points(log(hum.germ.mean),log(hum.germ.cv), col="red",pch=16,cex=0.5)


plot(log(hum.germ.b10.mean),log(hum.germ.b10.cv),col="grey",pch=16,cex=0.5,xlab="Mean expression", ylab="CV")
points(log(hum.germ.m10.mean),log(hum.germ.m10.cv),col="grey",pch=16,cex=0.5)
points(log(hum.germ.f10.mean),log(hum.germ.f10.cv),col="yellow",pch=16,cex=0.5)

plot(log(hum.germ.f10.mean),log(hum.germ.f10.cv),col="grey",pch=16,cex=0.5,xlab="Mean expression", ylab="CV")
points(log(hum.germ.b10.mean),log(hum.germ.b10.cv),col="grey",pch=16,cex=0.5)
points(log(hum.germ.m10.mean),log(hum.germ.m10.cv),col="green",pch=16,cex=0.5)

plot(log(hum.germ.f10.mean),log(hum.germ.f10.cv),col="grey",pch=16,cex=0.5,xlab="Mean expression", ylab="CV")
points(log(hum.germ.m10.mean),log(hum.germ.m10.cv),col="grey",pch=16,cex=0.5)
points(log(hum.germ.b10.mean),log(hum.germ.b10.cv),col="blue",pch=16,cex=0.5)


```

```{r}
names(hum.spg.mean) <- rownames(dge.hum.spg.norm)
names(hum.spg.per0) <- rownames(dge.hum.spg.norm)
names(hum.spg.cv) <- rownames(dge.hum.spg.norm)

names(hum.germ.mean) <- rownames(dge.hum.germ.norm)
names(hum.germ.per0) <- rownames(dge.hum.germ.norm)
names(hum.germ.cv) <- rownames(dge.hum.germ.norm)

names(hum.germ.b10.mean) <- rownames(dge.hum.germ.norm)
names(hum.germ.b10.per0) <- rownames(dge.hum.germ.norm)
names(hum.germ.b10.cv) <- rownames(dge.hum.germ.norm)

names(hum.germ.m10.mean) <- rownames(dge.hum.germ.norm)
names(hum.germ.m10.per0) <- rownames(dge.hum.germ.norm)
names(hum.germ.m10.cv) <- rownames(dge.hum.germ.norm)

names(hum.germ.f10.mean) <- rownames(dge.hum.germ.norm)
names(hum.germ.f10.per0) <- rownames(dge.hum.germ.norm)
names(hum.germ.f10.cv) <- rownames(dge.hum.germ.norm)
```


```{r fig.height=5,fig.width=6}
plot(log(hum.spg.mean[humMacMouOrth.1to1.in$external_gene_name.x]),hum.spg.per0[humMacMouOrth.1to1.in$external_gene_name.x],pch=16,cex=0.5,xlab = "Mean expression", ylab = "Percentage of NonZero")
points(log(hum.germ.mean[humMacMouOrth.1to1.in$external_gene_name.x]),hum.germ.per0[humMacMouOrth.1to1.in$external_gene_name.x], col="red",pch=16,cex=0.5)

plot(log(hum.germ.b10.mean[humMacMouOrth.1to1.in$external_gene_name.x]),hum.germ.b10.per0[humMacMouOrth.1to1.in$external_gene_name.x],col="grey",pch=16,cex=0.5,xlab = "Mean expression", ylab = "Percentage of NonZero")
points(log(hum.germ.m10.mean[humMacMouOrth.1to1.in$external_gene_name.x]),hum.germ.m10.per0[humMacMouOrth.1to1.in$external_gene_name.x],col="grey",pch=16,cex=0.5)
points(log(hum.germ.f10.mean[humMacMouOrth.1to1.in$external_gene_name.x]),hum.germ.f10.per0[humMacMouOrth.1to1.in$external_gene_name.x],col="yellow",pch=16,cex=0.5)

plot(log(hum.germ.f10.mean[humMacMouOrth.1to1.in$external_gene_name.x]),hum.germ.f10.per0[humMacMouOrth.1to1.in$external_gene_name.x],col="grey",pch=16,cex=0.5,xlab = "Mean expression", ylab = "Percentage of NonZero")
points(log(hum.germ.b10.mean[humMacMouOrth.1to1.in$external_gene_name.x]),hum.germ.b10.per0[humMacMouOrth.1to1.in$external_gene_name.x],col="grey",pch=16,cex=0.5)
points(log(hum.germ.m10.mean[humMacMouOrth.1to1.in$external_gene_name.x]),hum.germ.m10.per0[humMacMouOrth.1to1.in$external_gene_name.x],col="green",pch=16,cex=0.5)

plot(log(hum.germ.f10.mean[humMacMouOrth.1to1.in$external_gene_name.x]),hum.germ.f10.per0[humMacMouOrth.1to1.in$external_gene_name.x],col="grey",pch=16,cex=0.5,xlab = "Mean expression", ylab = "Percentage of NonZero")
points(log(hum.germ.m10.mean[humMacMouOrth.1to1.in$external_gene_name.x]),hum.germ.m10.per0[humMacMouOrth.1to1.in$external_gene_name.x],col="grey",pch=16,cex=0.5)
points(log(hum.germ.b10.mean[humMacMouOrth.1to1.in$external_gene_name.x]),hum.germ.b10.per0[humMacMouOrth.1to1.in$external_gene_name.x],col="blue",pch=16,cex=0.5)


plot(log(hum.spg.mean[humMacMouOrth.1to1.in$external_gene_name.x]),log(hum.spg.cv[humMacMouOrth.1to1.in$external_gene_name.x]),pch=16,cex=0.5,xlab = "Mean expression", ylab = "CV")
points(log(hum.germ.mean[humMacMouOrth.1to1.in$external_gene_name.x]),log(hum.germ.cv[humMacMouOrth.1to1.in$external_gene_name.x]), col="red",pch=16,cex=0.5)


plot(log(hum.germ.b10.mean[humMacMouOrth.1to1.in$external_gene_name.x]),log(hum.germ.b10.cv[humMacMouOrth.1to1.in$external_gene_name.x]),col="grey",pch=16,cex=0.5,xlab = "Mean expression", ylab = "CV")
points(log(hum.germ.m10.mean[humMacMouOrth.1to1.in$external_gene_name.x]),log(hum.germ.m10.cv[humMacMouOrth.1to1.in$external_gene_name.x]),col="grey",pch=16,cex=0.5)
points(log(hum.germ.f10.mean[humMacMouOrth.1to1.in$external_gene_name.x]),log(hum.germ.f10.cv[humMacMouOrth.1to1.in$external_gene_name.x]),col="yellow",pch=16,cex=0.5)

plot(log(hum.germ.f10.mean[humMacMouOrth.1to1.in$external_gene_name.x]),log(hum.germ.f10.cv[humMacMouOrth.1to1.in$external_gene_name.x]),col="grey",pch=16,cex=0.5,xlab = "Mean expression", ylab = "CV")
points(log(hum.germ.b10.mean[humMacMouOrth.1to1.in$external_gene_name.x]),log(hum.germ.b10.cv[humMacMouOrth.1to1.in$external_gene_name.x]),col="grey",pch=16,cex=0.5)
points(log(hum.germ.m10.mean[humMacMouOrth.1to1.in$external_gene_name.x]),log(hum.germ.m10.cv[humMacMouOrth.1to1.in$external_gene_name.x]),col="green",pch=16,cex=0.5)

plot(log(hum.germ.f10.mean[humMacMouOrth.1to1.in$external_gene_name.x]),log(hum.germ.f10.cv[humMacMouOrth.1to1.in$external_gene_name.x]),col="grey",pch=16,cex=0.5,xlab = "Mean expression", ylab = "CV")
points(log(hum.germ.m10.mean[humMacMouOrth.1to1.in$external_gene_name.x]),log(hum.germ.m10.cv[humMacMouOrth.1to1.in$external_gene_name.x]),col="grey",pch=16,cex=0.5)
points(log(hum.germ.b10.mean[humMacMouOrth.1to1.in$external_gene_name.x]),log(hum.germ.b10.cv[humMacMouOrth.1to1.in$external_gene_name.x]),col="blue",pch=16,cex=0.5)

```

###1-1-1 orthologues
####percentage of total expression
```{r fig.width=7,fig.height=5}
hum.111orth.perc <- apply(hum.201.cen[humMacMouOrth.1to1.in$external_gene_name.x,],2,sum)/apply(hum.201.cen,2,sum)
mac.111orth.perc <- apply(mac.201.cen[humMacMouOrth.1to1.in$mmulatta_homolog_associated_gene_name,],2,sum)/apply(mac.201.cen,2,sum)
mou.111orth.perc <- apply(mou.201.cen[humMacMouOrth.1to1.in$mmusculus_homolog_associated_gene_name,],2,sum)/apply(mou.201.cen,2,sum)

pdf("~/storage/HumanMacaqueMouseCompAnaysis/interdata/plots/111orth.exp.pdf",width = 7, height = 5)
ggplot(data.frame(stages=c(1:201), orth=c(hum.111orth.perc, mac.111orth.perc, mou.111orth.perc), num=c(dim(dge.hum.spg.raw)[2],hum.lens, dim(dge.mac.spg.raw)[2],mac.lens, dim(dge.mou.spg.raw)[2], mou.lens),traj=c(2,rep(1,200),rep(c(2,rep(1,200)),2)), spe=c(rep("Human",201),rep("Monkey",201),rep("Mouse",201))), aes(stages,orth)) + geom_point(aes(colour=spe, shape=factor(traj), size=sqrt(num))) + scale_color_manual(values = c("#e35a97","#5a97e3","#5ae3a6")) + xlab("Pseudo points") + ylab("Percentage of total expression") + ggtitle("1-1-1 Orthologues")
dev.off()
```

###1-1-1 orthologues
####ratio of expressed genes
```{r fig.width=7,fig.height=5}
hum.111orth.201.vperc <- calcvperc(humMacMouOrth.1to1.in$external_gene_name.x,dge.hum.germ.norm, dge.hum.spg.norm,hummac.mapping$refAssign)
mac.111orth.201.vperc <- calcvperc(humMacMouOrth.1to1.in$mmulatta_homolog_associated_gene_name,dge.mac.germ.norm,dge.mac.spg.norm,hummac.mapping$queryAssign)
mou.111orth.201.vperc <- calcvperc(humMacMouOrth.1to1.in$mmusculus_homolog_associated_gene_name,dge.mou.germ.norm,dge.mou.spg.norm,hummou.mapping$queryAssign)

pdf("~/storage/HumanMacaqueMouseCompAnaysis/interdata/plots/111orth.genratio.pdf",width = 7, height = 5)
ggplot(data.frame(stages=c(1:201), orth=c(hum.111orth.201.vperc, mac.111orth.201.vperc, mou.111orth.201.vperc), num=c(dim(dge.hum.spg.raw)[2],hum.lens, dim(dge.mac.spg.raw)[2],mac.lens, dim(dge.mou.spg.raw)[2], mou.lens),traj=c(2,rep(1,200),rep(c(2,rep(1,200)),2)), spe=c(rep("Human",201),rep("Monkey",201),rep("Mouse",201))), aes(stages,orth)) + geom_point(aes(colour=spe, shape=factor(traj), size=sqrt(num))) + scale_color_manual(values = c("#e35a97","#5a97e3","#5ae3a6")) + xlab("Pseudo points") + ylab("Percentage of expressed genes") + ggtitle("1-1-1 Orthologues")
dev.off()
```

### retrieve human ensembl database and extract human monkey orthologues
```{r}
library(biomaRt) #Ensemble Gene 94
ensembl <- useMart(host='http://oct2018.archive.ensembl.org', 
                     biomart='ENSEMBL_MART_ENSEMBL')
ensembl = useDataset("hsapiens_gene_ensembl",mart=ensembl)

attributes = c("ensembl_gene_id","external_gene_name","mmulatta_homolog_ensembl_gene","mmulatta_homolog_associated_gene_name","mmulatta_homolog_perc_id","mmulatta_homolog_perc_id_r1","mmulatta_homolog_goc_score","mmulatta_homolog_wga_coverage","mmulatta_homolog_dn","mmulatta_homolog_ds","mmulatta_homolog_orthology_confidence","mmulatta_homolog_orthology_type")

#Get all Human-Macaque gene pairs
humMac <- getBM(attributes = attributes,mart = ensembl)
dim(humMac)

#Get all Human-Macaque gene pairs
hum_type <- getBM(attributes = c("ensembl_gene_id","external_gene_name","gene_biotype"),mart = ensembl)
```

###. extract Human-Mouse orthologues and information
```{r}
attributes = c("ensembl_gene_id","external_gene_name","mmusculus_homolog_ensembl_gene","mmusculus_homolog_associated_gene_name", "mmusculus_homolog_perc_id","mmusculus_homolog_perc_id_r1","mmusculus_homolog_goc_score","mmusculus_homolog_wga_coverage","mmusculus_homolog_dn","mmusculus_homolog_ds","mmusculus_homolog_orthology_confidence","mmusculus_homolog_orthology_type")

humMou <- getBM(attributes = attributes,mart = ensembl)
dim(humMou)
```

###. merge human-monkey and human-mouse orthologues using the ensembl gene id as the key and filter protein coding genes
```{r}
humgene <- unique(humMac$ensembl_gene_id)
humMacMou.AllGene <- merge(humMac, humMou, by = "ensembl_gene_id")
rownames(hum_type) = hum_type$ensembl_gene_id
hum.spe <- humMacMou.AllGene[humMacMou.AllGene$mmulatta_homolog_ensembl_gene=="" & humMacMou.AllGene$mmusculus_homolog_ensembl_gene=="",]
hum.spe.coding <- hum.spe[hum_type[hum.spe$ensembl_gene_id,]$gene_biotype=="protein_coding",]
```

### Human specific genes
```{r}
#gave ensembl id to gene symbol if the symbol doesn't exist.
hum.spe <- intersect(hum.spe$external_gene_name.x, rownames(dge.hum.germ.norm))
hum.spe.coding <- intersect(hum.spe.coding$external_gene_name.x, rownames(dge.hum.germ.norm))
```

### retrieve mouse ensembl database 
```{r}
ensembl = useDataset("mmusculus_gene_ensembl",mart=ensembl)

attributes = c("ensembl_gene_id","external_gene_name","hsapiens_homolog_ensembl_gene","hsapiens_homolog_associated_gene_name","hsapiens_homolog_perc_id","hsapiens_homolog_perc_id_r1","hsapiens_homolog_goc_score","hsapiens_homolog_wga_coverage","hsapiens_homolog_dn","hsapiens_homolog_ds","hsapiens_homolog_orthology_confidence","hsapiens_homolog_orthology_type")

#Get all Human-Macaque gene pairs
mouHum <- getBM(attributes = attributes,mart = ensembl)
dim(mouHum)

mou_type <- getBM(attributes = c("ensembl_gene_id","external_gene_name","gene_biotype"),mart = ensembl)

###. Mouse-Macaque
attributes = c("ensembl_gene_id","external_gene_name","mmulatta_homolog_ensembl_gene","mmulatta_homolog_associated_gene_name", "mmulatta_homolog_perc_id","mmulatta_homolog_perc_id_r1","mmulatta_homolog_goc_score","mmulatta_homolog_wga_coverage","mmulatta_homolog_dn","mmulatta_homolog_ds","mmulatta_homolog_orthology_confidence","mmulatta_homolog_orthology_type")

#Get all Human-Macaque gene pairs
mouMac <- getBM(attributes = attributes,mart = ensembl)
dim(mouMac)

```

##merge mouse-human and mouse-monkey orthologues by using ensembl gene id as the kay and filter protein coding genes.
```{r}
mouHumMac.AllGene <- merge(mouHum, mouMac, by = "ensembl_gene_id")
rownames(mou_type) = mou_type$ensembl_gene_id
mou.spe <- mouHumMac.AllGene[mouHumMac.AllGene$mmulatta_homolog_ensembl_gene=="" & mouHumMac.AllGene$hsapiens_homolog_ensembl_gene=="",]
mou.spe.coding <- mou.spe[mou_type[mou.spe$ensembl_gene_id,]$gene_biotype=="protein_coding",]
```

### Mouse specific genes
```{r}
mou.spe <- intersect(mou.spe$external_gene_name.x, rownames(dge.mou.germ.norm))
mou.spe.coding <- intersect(mou.spe.coding$external_gene_name.x, rownames(dge.mou.germ.norm))
```

###. retrieve Monkey orthologues information from ensembl database
```{r}
ensembl = useDataset("mmulatta_gene_ensembl",mart=ensembl)
#listFilters(ensembl) #265 with_mmulatta_homolog
#searchAttributes(mart = ensembl, pattern = "mmulatta")

attributes = c("ensembl_gene_id","external_gene_name","hsapiens_homolog_ensembl_gene","hsapiens_homolog_associated_gene_name","hsapiens_homolog_perc_id","hsapiens_homolog_perc_id_r1","hsapiens_homolog_goc_score","hsapiens_homolog_wga_coverage","hsapiens_homolog_dn","hsapiens_homolog_ds","hsapiens_homolog_orthology_confidence","hsapiens_homolog_orthology_type")

#Get all Human-Macaque gene pairs
macHum <- getBM(attributes = attributes,mart = ensembl)
dim(macHum)

mac_type <- getBM(attributes = c("ensembl_gene_id","external_gene_name","gene_biotype"),mart = ensembl)

###. Mouse-Macaque
attributes = c("ensembl_gene_id","external_gene_name","mmusculus_homolog_ensembl_gene","mmusculus_homolog_associated_gene_name", "mmusculus_homolog_perc_id","mmusculus_homolog_perc_id_r1","mmusculus_homolog_goc_score","mmusculus_homolog_wga_coverage","mmusculus_homolog_dn","mmusculus_homolog_ds","mmusculus_homolog_orthology_confidence","mmusculus_homolog_orthology_type")

#Get all Human-Macaque gene pairs
macMou <- getBM(attributes = attributes,mart = ensembl)
dim(macMou)
```

### merge monkey-human and monkey-mouse orthologues by using ensembl gene id as the key and filter by protein coding genes.
```{r}
macHumMou.AllGene <- merge(macHum, macMou, by = "ensembl_gene_id")
rownames(mac_type) = mac_type$ensembl_gene_id
mac.spe <- macHumMou.AllGene[macHumMou.AllGene$mmusculus_homolog_ensembl_gene=="" & macHumMou.AllGene$hsapiens_homolog_ensembl_gene=="",]
mac.spe.coding <- mac.spe[mac_type[mac.spe$ensembl_gene_id,]$gene_biotype=="protein_coding",]
```

### Monkey specific genes
```{r}
##replace empty gene name with corresponding ensembl gene id.
mac.spe$external_gene_name.x[mac.spe$external_gene_name.x==""] = mac.spe$ensembl_gene_id[mac.spe$external_gene_name.x==""]
mac.spe.coding$external_gene_name.x[mac.spe.coding$external_gene_name.x==""] = mac.spe.coding$ensembl_gene_id[mac.spe.coding$external_gene_name.x==""]

mac.spe <- intersect(mac.spe$external_gene_name.x, rownames(dge.mac.germ.norm))
mac.spe.coding <- intersect(mac.spe.coding$external_gene_name.x, rownames(dge.mac.germ.norm))
```

#####. primate-specific genes (shared by human and monkey not in mouse)
```{r}
hummac.spe <- humMacMou.AllGene[humMacMou.AllGene$mmulatta_homolog_ensembl_gene!="" & humMacMou.AllGene$mmusculus_homolog_ensembl_gene=="",]
hummac.spe.coding <- hummac.spe[hum_type[hummac.spe$ensembl_gene_id, ]$gene_biotype=="protein_coding",]
```

### filter by genes existing in our dataset.
```{r}
##for monkey replace empty gene name with corresponding ensembl gene id.
hummac.spe.coding$mmulatta_homolog_associated_gene_name[hummac.spe.coding$mmulatta_homolog_associated_gene_name==""]=hummac.spe.coding$mmulatta_homolog_ensembl_gene[hummac.spe.coding$mmulatta_homolog_associated_gene_name==""]

hum.hummac.spe.coding <- intersect(hummac.spe.coding$external_gene_name.x,rownames(dge.hum.germ.norm))
mac.hummac.spe.coding <-intersect(hummac.spe.coding$mmulatta_homolog_associated_gene_name, rownames(dge.mac.germ.norm))
```

###. primate-specific genes
####. percentage of total expression
```{r fig.width=7,fig.height=5}
hum.hummac.coding.perc <- apply(hum.201.cen[hum.hummac.spe.coding,],2,sum)/apply(hum.201.cen,2,sum)
mac.hummac.coding.perc <- apply(mac.201.cen[mac.hummac.spe.coding,],2,sum)/apply(mac.201.cen,2,sum)

pdf("~/storage/HumanMacaqueMouseCompAnaysis/interdata/plots/primates.exp.pdf",width = 7, height = 5)
ggplot(data.frame(stages=c(1:201), orth=c(hum.hummac.coding.perc, mac.hummac.coding.perc), num=c(dim(dge.hum.spg.raw)[2],hum.lens, dim(dge.mac.spg.raw)[2],mac.lens),traj=rep(c(2,rep(1,200)),2), spe=c(rep("Human",201),rep("Monkey",201))), aes(stages,orth)) + geom_point(aes(colour=spe, shape=factor(traj), size=sqrt(num))) + scale_color_manual(values = c("#e35a97","#5a97e3")) + xlab("Pseudo points") + ylab("Percentage of total expression") + ggtitle("Primates specific")
dev.off()
```

###. primate-specific genes
####. ratio of expressed genes
```{r}
hum.hummac.201.perc <- calcvperc(hum.hummac.spe.coding, dge.hum.germ.norm, dge.hum.spg.norm, hummac.mapping$refAssign)

mac.hummac.201.perc <- calcvperc(mac.hummac.spe.coding, dge.mac.germ.norm, dge.mac.spg.norm, hummac.mapping$queryAssign)
```


```{r fig.width=7,fig.height=5}
pdf("~/storage/HumanMacaqueMouseCompAnaysis/interdata/plots/primates.genratio.pdf",width = 7, height = 5)
ggplot(data.frame(stages=c(1:201), orth=c(hum.hummac.201.perc, mac.hummac.201.perc), num=c(dim(dge.hum.spg.raw)[2],hum.lens, dim(dge.mac.spg.raw)[2],mac.lens),traj=c(rep(c(2,rep(1,200)),2)), spe=c(rep("Human",201),rep("Monkey",201))), aes(stages,orth)) + geom_point(aes(colour=spe, shape=factor(traj), size=sqrt(num))) + scale_color_manual(values = c("#e35a97","#5a97e3")) + xlab("Pseudo points") + ylab("Percentage of expressed genes") + ggtitle("Primate Specific")
dev.off()
```

#####. moumac-specific (orthologues between mouse and monkey no orthologues in mouse)
```{r}
moumac.spe <- mouHumMac.AllGene[mouHumMac.AllGene$hsapiens_homolog_ensembl_gene=="" & mouHumMac.AllGene$mmulatta_homolog_ensembl_gene!="",]
moumac.spe.coding <- moumac.spe[mou_type[moumac.spe$ensembl_gene_id, ]$gene_biotype=="protein_coding",]
```

### moumac specific genes
```{r}
##replace empty monkey gene name with corresponding ensembl gene id 
moumac.spe.coding$mmulatta_homolog_associated_gene_name[moumac.spe.coding$mmulatta_homolog_associated_gene_name==""] = moumac.spe.coding$mmulatta_homolog_ensembl_gene[moumac.spe.coding$mmulatta_homolog_associated_gene_name==""]
mou.moumac.spe.coding <- intersect(moumac.spe.coding$external_gene_name.x,rownames(dge.mou.germ.norm))
mac.moumac.spe.coding <-intersect(moumac.spe.coding$mmulatta_homolog_associated_gene_name, rownames(dge.mac.germ.norm))
```

###Mouse-monkey specific genes
#### percentage of total expression
```{r fig.width=7,fig.height=5}
mou.moumac.coding.perc <- apply(mou.201.cen[mou.moumac.spe.coding,],2,sum)/apply(mou.201.cen,2,sum)
mac.moumac.coding.perc <- apply(mac.201.cen[mac.moumac.spe.coding,],2,sum)/apply(mac.201.cen,2,sum)

pdf("~/storage/HumanMacaqueMouseCompAnaysis/interdata/plots/moumon.exp.pdf",width = 7, height = 5)
ggplot(data.frame(stages=c(1:201), orth=c(mac.moumac.coding.perc, mou.moumac.coding.perc), num=c(dim(dge.mac.spg.raw)[2],mac.lens, dim(dge.mou.spg.raw)[2], mou.lens),traj=rep(c(2,rep(1,200)),2), spe=c(rep("Monkey",201),rep("Mouse",201))), aes(stages,orth)) + geom_point(aes(colour=spe, shape=factor(traj), size=sqrt(num))) + scale_color_manual(values = c("#5a97e3","#5ae3a6")) + xlab("Pseudo points") + ylab("Percentage of total expression") + ggtitle("Mouse-Monkey specific")
dev.off()
```

###MOuse-monkey specific genes
#### Ratio of expressed genes
```{r}
mac.moumac.201.perc <- calcvperc(mac.moumac.spe.coding, dge.mac.germ.norm, dge.mac.spg.norm, hummac.mapping$queryAssign)

mou.moumac.201.perc <- calcvperc(mou.moumac.spe.coding, dge.mou.germ.norm, dge.mou.spg.norm, hummou.mapping$queryAssign)
```

```{r fig.width=7,fig.height=5}
pdf("~/storage/HumanMacaqueMouseCompAnaysis/interdata/plots/moumon.genratio.pdf",width = 7, height = 5)
ggplot(data.frame(stages=c(1:201), orth=c(mac.moumac.201.perc, mou.moumac.201.perc), num=c(dim(dge.mac.spg.raw)[2],mac.lens, dim(dge.mou.spg.raw)[2], mou.lens),traj=c(rep(c(2,rep(1,200)),2)), spe=c(rep("Monkey",201),rep("Mouse",201))), aes(stages,orth)) + geom_point(aes(colour=spe, shape=factor(traj), size=sqrt(num))) + scale_color_manual(values = c("#5a97e3","#5ae3a6")) + xlab("Pseudo points") + ylab("Percentage of expressed genes") + ggtitle("Mouse and Monkey specific")
dev.off()
```

#####. hummou-specific genes (orthologues between human and mouse, no orthologues in monkey)
```{r}
hummou.spe <- humMacMou.AllGene[humMacMou.AllGene$mmulatta_homolog_ensembl_gene=="" & humMacMou.AllGene$mmusculus_homolog_ensembl_gene!="", ]
hummou.spe.coding <- hummou.spe[hum_type[hummou.spe$ensembl_gene_id, ]$gene_biotype=="protein_coding",]
```

####filtered by genes that have expression in our datasets.
```{r}
mou.hummou.spe.coding <- intersect(hummou.spe.coding$mmusculus_homolog_associated_gene_name,rownames(dge.mou.germ.norm))
hum.hummou.spe.coding <-intersect(hummou.spe.coding$external_gene_name.x, rownames(dge.hum.germ.norm))
```

####human-mouse specific genes
#### percentage of total expression.
```{r fig.width=7,fig.height=5}
mou.hummou.coding.perc <- apply(mou.201.cen[mou.hummou.spe.coding,],2,sum)/apply(mou.201.cen,2,sum)
hum.hummou.coding.perc <- apply(hum.201.cen[hum.hummou.spe.coding,],2,sum)/apply(hum.201.cen,2,sum)

pdf("~/storage/HumanMacaqueMouseCompAnaysis/interdata/plots/hummou.exp.pdf",width = 7, height = 5)
ggplot(data.frame(stages=c(1:201), orth=c(hum.hummou.coding.perc, mou.hummou.coding.perc), num=c(dim(dge.hum.spg.raw)[2],hum.lens, dim(dge.mou.spg.raw)[2], mou.lens),traj=rep(c(2,rep(1,200)),2), spe=c(rep("Human",201),rep("Mouse",201))), aes(stages,orth)) + geom_point(aes(colour=spe, shape=factor(traj), size=sqrt(num))) + scale_color_manual(values = c("#e35a97","#5ae3a6")) + xlab("Pseudo points") + ylab("Percentage of total expression") + ggtitle("Human-Mouse specific ")
dev.off()
```

####human-mouse specific genes
####ratio of expressed genes
```{r}
hum.hummou.201.perc <- calcvperc(unique(hum.hummou.spe.coding), dge.hum.germ.norm, dge.hum.spg.norm, hummac.mapping$refAssign)

mou.hummou.201.perc <- calcvperc(unique(mou.hummou.spe.coding), dge.mou.germ.norm, dge.mou.spg.norm, hummou.mapping$queryAssign)
```

```{r fig.width=7,fig.height=5}
pdf("~/storage/HumanMacaqueMouseCompAnaysis/interdata/plots/hummou.genratio.pdf",width = 7, height = 5)
ggplot(data.frame(stages=c(1:201), orth=c(hum.hummou.201.perc, mou.hummou.201.perc), num=c(dim(dge.hum.spg.raw)[2],hum.lens, dim(dge.mou.spg.raw)[2], mou.lens),traj=c(rep(c(2,rep(1,200)),2)), spe=c(rep("Human",201),rep("Mouse",201))), aes(stages,orth)) + geom_point(aes(colour=spe, shape=factor(traj), size=sqrt(num))) + scale_color_manual(values = c("#e35a97","#5ae3a6")) + xlab("Pseudo points") + ylab("Percentage of expressed genes") + ggtitle("Human and Mouse specific")
dev.off()
```

### Other types of orthologs (genes that have orthologues in all three species but not in 1-1-1 relationship)
```{r}
#all detected human coding genes
hum.all.det.coding <- intersect(hum_type$external_gene_name[hum_type$gene_biotype=="protein_coding"], rownames(dge.hum.germ.norm))
#remove human specific genes
hum.other.othos <- setdiff(hum.all.det.coding, hum.spe.coding)
#remove hum-mac specific genes
hum.other.othos <- setdiff(hum.other.othos, hum.hummac.spe.coding)
#remove hum-mou specific genes
hum.other.othos <- setdiff(hum.other.othos, hum.hummou.spe.coding)
#remove 1-1-1 orthologs
hum.other.othos <- setdiff(hum.other.othos, humMacMouOrth.1to1.in$external_gene_name.x)
```

```{r}
#all detected monkey coding genes
mac.all.det.coding <- intersect(mac_type$external_gene_name[mac_type$gene_biotype=="protein_coding"], rownames(dge.mac.germ.norm))
#remove monkey specific genes
mac.other.othos <- setdiff(mac.all.det.coding, mac.spe.coding)
#remove hum-mac specific genes
mac.other.othos <- setdiff(mac.other.othos, mac.hummac.spe.coding)
#remove mac-mou specific genes
mac.other.othos <- setdiff(mac.other.othos, mac.moumac.spe.coding)
#remove 1-1-1 orthologs
mac.other.othos <- setdiff(mac.other.othos, humMacMouOrth.1to1.in$mmulatta_homolog_associated_gene_name)
```

```{r}
#all detected mouse coding genes
mou.all.det.coding <- intersect(mou_type$external_gene_name[mou_type$gene_biotype=="protein_coding"], rownames(dge.mou.germ.norm))
#remove mouse specific genes
mou.other.othos <- setdiff(mou.all.det.coding, mou.spe.coding)
#remove hum-mou specific genes
mou.other.othos <- setdiff(mou.other.othos, mou.hummou.spe.coding)
#remove mac-mou specific genes
mou.other.othos <- setdiff(mou.other.othos, mou.moumac.spe.coding)
#remove 1-1-1 orthologs
mou.other.othos <- setdiff(mou.other.othos, humMacMouOrth.1to1.in$mmusculus_homolog_associated_gene_name)
```

### other types of orthologues
#### percentage of total expression
```{r fig.width=7,fig.height=5}
hum.other.othos.perc <- apply(hum.201.cen[hum.other.othos,],2,sum)/apply(hum.201.cen,2,sum)
mac.other.othos.perc <- apply(mac.201.cen[mac.other.othos,],2,sum)/apply(mac.201.cen,2,sum)
mou.other.othos.perc <- apply(mou.201.cen[mou.other.othos,],2,sum)/apply(mou.201.cen,2,sum)

pdf("~/storage/HumanMacaqueMouseCompAnaysis/interdata/plots/otherortho.exp.pdf",width = 7, height = 5)
ggplot(data.frame(stages=c(1:201), orth=c(hum.other.othos.perc, mac.other.othos.perc, mou.other.othos.perc), num=c(dim(dge.hum.spg.raw)[2],hum.lens, dim(dge.mac.spg.raw)[2],mac.lens, dim(dge.mou.spg.raw)[2], mou.lens),traj=c(2,rep(1,200),rep(c(2,rep(1,200)),2)), spe=c(rep("Human",201),rep("Monkey",201),rep("Mouse",201))), aes(stages,orth)) + geom_point(aes(colour=spe, shape=factor(traj), size=sqrt(num))) + scale_color_manual(values = c("#e35a97","#5a97e3","#5ae3a6")) + xlab("Pseudo points") + ylab("Percentage of total expression") + ggtitle("Other types of three species Orthologs")
dev.off()
```

### other types of orthologues
#### ratio of expressed genes
```{r fig.width=7,fig.height=5}
hum.other.othos.201.perc <- calcvperc(hum.other.othos,dge.hum.germ.norm, dge.hum.spg.norm,hummac.mapping$refAssign)
mac.other.othos.201.perc <- calcvperc(mac.other.othos,dge.mac.germ.norm,dge.mac.spg.norm,hummac.mapping$queryAssign)
mou.other.othos.201.perc <- calcvperc(mou.other.othos,dge.mou.germ.norm,dge.mou.spg.norm,hummou.mapping$queryAssign)

pdf("~/storage/HumanMacaqueMouseCompAnaysis/interdata/plots/otherortho.genratio.pdf",width = 7, height = 5)
ggplot(data.frame(stages=c(1:201), orth=c(hum.other.othos.201.perc, mac.other.othos.201.perc, mou.other.othos.201.perc), num=c(dim(dge.hum.spg.raw)[2],hum.lens, dim(dge.mac.spg.raw)[2],mac.lens, dim(dge.mou.spg.raw)[2], mou.lens),traj=c(2,rep(1,200),rep(c(2,rep(1,200)),2)), spe=c(rep("Human",201),rep("Monkey",201),rep("Mouse",201))), aes(stages,orth)) + geom_point(aes(colour=spe, shape=factor(traj), size=sqrt(num))) + scale_color_manual(values = c("#e35a97","#5a97e3","#5ae3a6")) + xlab("Pseudo points") + ylab("Percentage of expressed genes") + ggtitle("Other types of three species Orthologs")
dev.off()
```

### int.genes
```{r fig.width=7,fig.height=5}
hum.int.perc <- apply(hum.201.cen[int.genes$external_gene_name.x,],2,sum)/apply(hum.201.cen,2,sum)

mac.int.perc <- apply(mac.201.cen[int.genes$mmulatta_homolog_associated_gene_name,],2,sum)/apply(mac.201.cen,2,sum)

mou.int.perc <- apply(mou.201.cen[int.genes$mmusculus_homolog_associated_gene_name,],2,sum)/apply(mou.201.cen,2,sum)

dat <- data.frame(ca.cen.int.600.pca$x, orth=c(hum.int.perc, mac.int.perc, mou.int.perc), num=c(dim(dge.hum.spg.raw)[2],hum.lens, dim(dge.mac.spg.raw)[2],mac.lens, dim(dge.mou.spg.raw)[2], mou.lens),traj=c(2,rep(1,200),rep(c(2,rep(1,200)),2)))
ggplot(dat, aes(PC1,PC3)) + geom_point(aes(colour=orth, shape=factor(traj), size=num/100)) + scale_color_gradientn(colours = c("#0000b3","#ededed","#b30000")) 

dat <- data.frame(ca.cen.int.600.pca$x, orth=c(hum.int.perc, rep(NA,201), rep(NA,201)), num=c(dim(dge.hum.spg.raw)[2],hum.lens, dim(dge.mac.spg.raw)[2],mac.lens, dim(dge.mou.spg.raw)[2], mou.lens),traj=c(2,rep(1,200),rep(c(2,rep(1,200)),2)))
ggplot(dat[seq(dim(dat)[1],1),], aes(PC1,PC3)) + geom_point(aes(colour=orth, shape=factor(traj), size=num/100)) + scale_color_gradientn(colours = c("#0000b3","#ededed","#b30000")) 

dat <- data.frame(ca.cen.int.600.pca$x, orth=c(rep(NA,201), mac.int.perc, rep(NA,201)), num=c(dim(dge.hum.spg.raw)[2],hum.lens, dim(dge.mac.spg.raw)[2],mac.lens, dim(dge.mou.spg.raw)[2], mou.lens),traj=c(2,rep(1,200),rep(c(2,rep(1,200)),2)))
ggplot(dat, aes(PC1,PC3)) + geom_point(aes(colour=orth, shape=factor(traj), size=num/100)) + scale_color_gradientn(colours = c("#0000b3","#ededed","#b30000")) 

dat <- data.frame(ca.cen.int.600.pca$x, orth=c(rep(NA,201), rep(NA,201), mou.int.perc), num=c(dim(dge.hum.spg.raw)[2],hum.lens, dim(dge.mac.spg.raw)[2],mac.lens, dim(dge.mou.spg.raw)[2], mou.lens),traj=c(2,rep(1,200),rep(c(2,rep(1,200)),2)))
ggplot(dat, aes(PC1,PC3)) + geom_point(aes(colour=orth, shape=factor(traj), size=num/100)) + scale_color_gradientn(colours = c("#0000b3","#ededed","#b30000")) 
```


Three species specific 
```{r}
hum.spe.201.perc <- calcvperc(hum.spe.coding, dge.hum.germ.norm, dge.hum.spg.norm, hummac.mapping$refAssign)
mac.spe.201.perc <- calcvperc(mac.spe.coding, dge.mac.germ.norm, dge.mac.spg.norm, hummac.mapping$queryAssign)
mou.spe.201.perc <- calcvperc(mou.spe.coding, dge.mou.germ.norm, dge.mou.spg.norm, hummou.mapping$queryAssign)
```

Plot by three rows
```{r fig.width=7,fig.height=5}
pdf("~/storage/HumanMacaqueMouseCompAnaysis/interdata/plots/spe.specific.genratio.pdf",width = 7, height = 5)
ggplot(data.frame(stages=c(1:201), orth=c(hum.spe.201.perc, mac.spe.201.perc, mou.spe.201.perc), num=c(dim(dge.hum.spg.raw)[2],hum.lens, dim(dge.mac.spg.raw)[2],mac.lens, dim(dge.mou.spg.raw)[2], mou.lens),traj=c(2,rep(1,200),rep(c(2,rep(1,200)),2)), spe=c(rep("Human",201),rep("Monkey",201),rep("Mouse",201))), aes(stages,orth)) + geom_point(aes(colour=spe, shape=factor(traj), size=sqrt(num))) + scale_color_manual(values = c("#e35a97","#5a97e3","#5ae3a6")) + xlab("Pseudo points") + ylab("Percentage of expressed genes") + ggtitle("Species specific")
dev.off()
```

```{r fig.width=7,fig.height=5}
hum.spe.coding.perc <- apply(hum.201.cen[hum.spe.coding,],2,sum)/apply(hum.201.cen,2,sum)
mac.spe.coding.perc <- apply(mac.201.cen[mac.spe.coding,],2,sum)/apply(mac.201.cen,2,sum)
mou.spe.coding.perc <- apply(mou.201.cen[mou.spe.coding,],2,sum)/apply(mou.201.cen,2,sum)

pdf("~/storage/HumanMacaqueMouseCompAnaysis/interdata/plots/spe.specific.exp.pdf",width = 7, height = 5)
ggplot(data.frame(stages=c(1:201), orth=c(hum.spe.coding.perc, mac.spe.coding.perc, mou.spe.coding.perc), num=c(dim(dge.hum.spg.raw)[2],hum.lens, dim(dge.mac.spg.raw)[2],mac.lens, dim(dge.mou.spg.raw)[2], mou.lens),traj=c(2,rep(1,200),rep(c(2,rep(1,200)),2)), spe=c(rep("Human",201),rep("Monkey",201),rep("Mouse",201))), aes(stages,orth)) + geom_point(aes(colour=spe, shape=factor(traj), size=num/100)) + scale_color_manual(values = c("#e35a97","#5a97e3","#5ae3a6")) + xlab("Pseudo points") + ylab("Percentage of total expression") + ggtitle("Species specific")
dev.off()
```

```{r}
humMacMouOrth.1to1.in.coding <- humMacMouOrth.1to1.in[hum_type[humMacMouOrth.1to1.in$ensembl_gene_id,]$gene_biotype=="protein_coding" & mac_type[humMacMouOrth.1to1.in$mmulatta_homolog_ensembl_gene,]$gene_biotype=="protein_coding" & mou_type[humMacMouOrth.1to1.in$mmusculus_homolog_ensembl_gene,]$gene_biotype=="protein_coding",]
```

```{r fig.width=7,fig.height=5}
hum.111orth.coding.perc <- apply(hum.201.cen[humMacMouOrth.1to1.in.coding$external_gene_name.x,],2,sum)/apply(hum.201.cen,2,sum)
mac.111orth.coding.perc <- apply(mac.201.cen[humMacMouOrth.1to1.in.coding$mmulatta_homolog_associated_gene_name,],2,sum)/apply(mac.201.cen,2,sum)
mou.111orth.coding.perc <- apply(mou.201.cen[humMacMouOrth.1to1.in.coding$mmusculus_homolog_associated_gene_name,],2,sum)/apply(mou.201.cen,2,sum)

pdf("~/storage/HumanMacaqueMouseCompAnaysis/interdata/plots/111orth.coding.exp.pdf",width = 7, height = 5)
ggplot(data.frame(stages=c(1:201), orth=c(hum.111orth.coding.perc, mac.111orth.coding.perc, mou.111orth.coding.perc), num=c(dim(dge.hum.spg.raw)[2],hum.lens, dim(dge.mac.spg.raw)[2],mac.lens, dim(dge.mou.spg.raw)[2], mou.lens),traj=c(2,rep(1,200),rep(c(2,rep(1,200)),2)), spe=c(rep("Human",201),rep("Monkey",201),rep("Mouse",201))), aes(stages,orth)) + geom_point(aes(colour=spe, shape=factor(traj), size=sqrt(num))) + scale_color_manual(values = c("#e35a97","#5a97e3","#5ae3a6")) + xlab("Pseudo points") + ylab("Percentage of total expression") + ggtitle("1-1-1 Orthologues (Coding)")
dev.off()
```

```{r fig.width=7,fig.height=5}
hum.111orth.coding.201.vperc <- calcvperc(humMacMouOrth.1to1.in$external_gene_name.x,dge.hum.germ.norm, dge.hum.spg.norm,hummac.mapping$refAssign)
mac.111orth.coding.201.vperc <- calcvperc(humMacMouOrth.1to1.in$mmulatta_homolog_associated_gene_name,dge.mac.germ.norm,dge.mac.spg.norm,hummac.mapping$queryAssign)
mou.111orth.coding.201.vperc <- calcvperc(humMacMouOrth.1to1.in$mmusculus_homolog_associated_gene_name,dge.mou.germ.norm,dge.mou.spg.norm,hummou.mapping$queryAssign)

pdf("~/storage/HumanMacaqueMouseCompAnaysis/interdata/plots/111orth.coding.genratio.pdf",width = 7, height = 5)
ggplot(data.frame(stages=c(1:201), orth=c(hum.111orth.coding.201.vperc, mac.111orth.coding.201.vperc, mou.111orth.coding.201.vperc), num=c(dim(dge.hum.spg.raw)[2],hum.lens, dim(dge.mac.spg.raw)[2],mac.lens, dim(dge.mou.spg.raw)[2], mou.lens),traj=c(2,rep(1,200),rep(c(2,rep(1,200)),2)), spe=c(rep("Human",201),rep("Monkey",201),rep("Mouse",201))), aes(stages,orth)) + geom_point(aes(colour=spe, shape=factor(traj), size=sqrt(num))) + scale_color_manual(values = c("#e35a97","#5a97e3","#5ae3a6")) + xlab("Pseudo points") + ylab("Percentage of expressed genes") + ggtitle("1-1-1 Orthologues (Coding)")
dev.off()
```

Correlation matrix
```{r fig.height=8, fig.width=9}
ca.cen.int.cor <- cor <- cor(ca.cen.int.600.norm[, c(rep(TRUE,201),rep(TRUE,143),rep(FALSE,58), rep(TRUE,201))],method = "spearman")
rownames(ca.cen.int.cor) <- colnames(ca.cen.int.cor) <- c(paste0("H",c(1:201)), paste0("M",c(1:143)), paste0("m",c(1:201)))
Heatmap(ca.cen.int.cor, cluster_rows = FALSE, cluster_columns = FALSE)
```

###load package for principle curve
```{r}
library(princurve)
```

### principle curve fitting for three trajectories removing monkey tail.
```{r}
rmmac <- c(FALSE,rep(TRUE,200),FALSE, rep(TRUE,142),rep(FALSE,58), FALSE, rep(TRUE,200))
x <- ca.cen.int.600.pca$x[rmmac,c(1,3)]
fit <- principal_curve(x)
```

### plot the fitted principle fit to the PCA trajectory alignemtn plot
```{r fig.width=7,fig.height=5}
ggplot(data.frame(ca.cen.int.600.pca$x, species=factor(colnames(ca.cen.int.600)), num=c(dim(dge.hum.spg.norm)[2],hum.lens, dim(dge.mac.spg.norm)[2],mac.lens, dim(dge.mou.spg.norm)[2], mou.lens),traj=c(2,rep(1,200),2,rep(1,142), rep(3,58),2, rep(1,200))), aes(PC1,PC3)) + geom_point(aes(colour=species, shape=factor(traj), size=num/100))+ scale_color_manual(values = c("#e35a97","#5a97e3","#5ae3a6")) + scale_shape_manual(values = c(20, 17, 3)) + geom_line(data=data.frame(fit$s), aes(PC1, PC3))
```

### Plot the mapping for each dot onto the central principle curve.
```{r fig.width=7,fig.height=5}
ggplot(data.frame(ca.cen.int.600.pca$x, species=factor(colnames(ca.cen.int.600)), num=c(dim(dge.hum.spg.norm)[2],hum.lens, dim(dge.mac.spg.norm)[2],mac.lens, dim(dge.mou.spg.norm)[2], mou.lens),traj=c(2,rep(1,200),2,rep(1,142), rep(3,58),2, rep(1,200))), aes(PC1,PC3)) + geom_point(aes(colour=species, shape=factor(traj), size=num/100))+ scale_color_manual(values = c("#e35a97","#5a97e3","#5ae3a6")) + scale_shape_manual(values = c(20, 17, 3)) + geom_line(data=data.frame(fit$s), aes(PC1, PC3)) + geom_segment(aes(x=x,y=y,xend=xend,yend=yend), colour="grey", size=0.2, data=data.frame(x=ca.cen.int.600.pca$x[rmmac,"PC1"], y=ca.cen.int.600.pca$x[rmmac,"PC3"], xend=fit$s[,"PC1"], yend=fit$s[,"PC3"]))
```

##20 cluster centoids by cutting the principle curve to 20 segments with equal length. Then, 20 cluster centroids are generated for each species.
```{r}
pc.len <- max(fit$lambda)
pc.step <- (pc.len + 0.01)/20 ##add 0.01 to include the last point 
hum.lambda <- fit$lambda[1:200]
hum.20.warped.cen <- matrix(0, nrow=dim(dge.hum.germ.norm)[1], ncol=20)
hum.20.lens <- c()
for(i in c(1:20)){
  segs <- which(hum.lambda >= pc.step*(i-1) & hum.lambda < pc.step*i)
  cells <- unlist(lapply(segs,function(x) hummac.mapping$refAssign[[x]]))
  hum.20.lens <- c(hum.20.lens, length(cells))
  hum.20.warped.cen[,i] <- apply(dge.hum.germ.norm[,cells],1,mean)
}

mac.lambda <- fit$lambda[201:342]
mac.20.warped.cen <- matrix(0, nrow=dim(dge.mac.germ.norm)[1], ncol=20)
mac.20.lens <- c()
for(i in c(1:20)){
  segs <- which(mac.lambda >= pc.step*(i-1) & mac.lambda < pc.step*i)
  cells <- unlist(lapply(segs,function(x) hummac.mapping$queryAssign[[x]]))
  mac.20.lens <- c(mac.20.lens, length(cells))
  mac.20.warped.cen[,i] <- apply(dge.mac.germ.norm[,cells],1,mean)
}

mou.lambda <- fit$lambda[343:542]
mou.20.warped.cen <- matrix(0, nrow=dim(dge.mou.germ.norm)[1], ncol=20)
mou.20.lens <- c()
for(i in c(1:20)){
  segs <- which(mou.lambda >= pc.step*(i-1) & mou.lambda < pc.step*i)
  cells <- unlist(lapply(segs,function(x) hummou.mapping$queryAssign[[x]]))
  mou.20.lens <- c(mou.20.lens, length(cells))
  mou.20.warped.cen[,i] <- apply(dge.mou.germ.norm[,cells],1,mean)
}
```

```{r}
rownames(hum.20.warped.cen) <- rownames(dge.hum.germ.norm)
rownames(mac.20.warped.cen) <- rownames(dge.mac.germ.norm)
rownames(mou.20.warped.cen) <- rownames(dge.mou.germ.norm)
```


#### Write the 20 centroids to file
```{r}
write.table(rbind("NumberCells"=hum.20.lens, hum.20.warped.cen), file = "~/storage/HumanMacaqueMouseCompAnaysis/interdata/hum.20.warpped.centroids.text", quote = FALSE, row.names = TRUE, col.names = FALSE,sep="\t")
write.table(rbind("NumberCells"=mac.20.lens, mac.20.warped.cen), file = "~/storage/HumanMacaqueMouseCompAnaysis/interdata/mac.20.warpped.centroids.text", quote = FALSE, row.names = TRUE, col.names = FALSE,sep="\t")
write.table(rbind("NumberCells"=mou.20.lens, mou.20.warped.cen), file = "~/storage/HumanMacaqueMouseCompAnaysis/interdata/mou.20.warpped.centroids.text", quote = FALSE, row.names = TRUE, col.names = FALSE,sep="\t")
```


#### Violin plot for number of cells across the trajectory (20 stages)
```{r fig.width=8, fig.height=4}
dat <- data.frame(traj = c(unlist(lapply(c(1:20),function(x) rep(x, hum.20.lens[x]))), unlist(lapply(c(1:20),function(x) rep(x, mac.20.lens[x]))) ,unlist(lapply(c(1:20),function(x) rep(x, mou.20.lens[x])))), spe=c(unlist(lapply(c(1:20),function(x) rep("Human", hum.20.lens[x]))), unlist(lapply(c(1:20),function(x) rep("Monkey", mac.20.lens[x]))) ,unlist(lapply(c(1:20),function(x) rep("Mouse", mou.20.lens[x])))))
ggplot(dat, aes(factor(spe), traj)) + geom_violin(aes(fill=factor(spe))) + coord_flip() + labs(y="Pseudo-time from principle curve",x="Species")
```

###Calculate the number of cells only from ST1-6 batch (those batches are not specifically enriched for SPG)
```{r}
mou.lambda <- fit$lambda[343:542]
mou.20.st6.lens <- c()
for(i in c(1:20)){
  segs <- which(mou.lambda >= pc.step*(i-1) & mou.lambda < pc.step*i)
  cells <- unlist(lapply(segs,function(x) hummou.mapping$queryAssign[[x]]))
  cells <- cells[grepl("ST[1-6]",cells)]
  mou.20.st6.lens <- c(mou.20.st6.lens, length(cells))
}
```

```{r}
hum.spg.len <- dim(dge.hum.spg.norm)[2]
mac.spg.len <- dim(dge.mac.spg.norm)[2]
mou.spg.len <- sum(grepl("ST[1-6]", colnames(dge.mou.spg.norm)))
hum.21.lens <- c(hum.spg.len, hum.20.lens)
mac.21.lens <- c(mac.spg.len, mac.20.lens)
mou.21.lens <- c(mou.spg.len, mou.20.st6.lens)
```

###Violin plot adding SPG as one cluster
```{r fig.width=6, fig.height=4}
dat <- data.frame(traj = c(unlist(lapply(c(0:20),function(x) rep(x, hum.21.lens[x+1]))), unlist(lapply(c(0:20),function(x) rep(x, mac.21.lens[x+1]))) ,unlist(lapply(c(0:20),function(x) rep(x, mou.21.lens[x+1])))), spe=c(unlist(lapply(c(0:20),function(x) rep("Human", hum.21.lens[x+1]))), unlist(lapply(c(0:20),function(x) rep("Monkey", mac.21.lens[x+1]))) ,unlist(lapply(c(0:20),function(x) rep("Mouse", mou.21.lens[x+1])))))
ggplot(dat, aes(factor(spe), traj)) + geom_violin(aes(fill=factor(spe))) + coord_flip() + labs(y="Pseudo-time from principle curve",x="Species")+ geom_hline(yintercept = 1)+ theme(legend.position = "none")
```

Read in SPG dataset and Expand spg into 6 clusters
```{r}
spg.cca <- readRDS("~/storage/HumanMacaqueMouseCompAnaysis/interdata/spg.cca.rds")
```

```{r}
hum.spg.6.lens <- table(spg.cca@meta.data$res.0.6.ord[spg.cca@meta.data$species=="Human"])[1:6]
mac.spg.6.lens <- table(spg.cca@meta.data$res.0.6.ord[spg.cca@meta.data$species=="Monkey"])[1:6]
mou.spg.6.lens <- table(spg.cca@meta.data$res.0.6.ord[spg.cca@meta.data$species=="Mouse" & grepl("ST[1-6]", spg.cca@cell.names)])[1:6]
```

```{r}
hum.26.lens <- c(hum.spg.6.lens, hum.20.lens)
mac.26.lens <- c(mac.spg.6.lens, mac.20.lens)
mou.26.lens <- c(mou.spg.6.lens, mou.20.st6.lens)
```

###Violin plot adding SPG as 6 clusters
```{r fig.width=6, fig.height=4}
dat <- data.frame(traj = c(unlist(lapply(c(1:26),function(x) rep(x, hum.26.lens[x]))), unlist(lapply(c(1:26),function(x) rep(x, mac.26.lens[x]))) ,unlist(lapply(c(1:26),function(x) rep(x, mou.26.lens[x])))), spe=c(unlist(lapply(c(1:26),function(x) rep("Human", hum.26.lens[x]))), unlist(lapply(c(1:26),function(x) rep("Monkey", mac.26.lens[x]))) ,unlist(lapply(c(1:26),function(x) rep("Mouse", mou.26.lens[x])))))
ggplot(dat, aes(factor(spe), traj)) + geom_violin(aes(fill=factor(spe))) + coord_flip() + labs(y="Pseudo-time from principle curve",x="Species") + geom_hline(yintercept = 6)+ theme(legend.position = "none")
```

###Dot plot adding SPG as 6 clusters
```{r fig.width=8, fig.height=4}
dat <- data.frame(traj = c(hum.26.lens,mac.26.lens,mou.26.lens), spe=c(rep("Human",26),rep("Monkey",26),rep("Mouse",26)), pt=rep(c(1:26),3))
ggplot(dat, aes(pt, factor(spe))) + scale_size_continuous(range=c(1,10)) + geom_point(aes(col=factor(spe),size=sqrt(traj))) + labs(y="Pseudo-time from principle curve",x="Species") + geom_vline(xintercept = 6.5)
```

###Dot plot adding SPG as 1 cluster
```{r fig.width=8, fig.height=4}
dat <- data.frame(traj = c(hum.21.lens,mac.21.lens,mou.21.lens), spe=c(rep("Human",21),rep("Monkey",21),rep("Mouse",21)), pt=rep(c(1:21),3))
ggplot(dat, aes(pt, factor(spe))) + scale_size_continuous(range=c(1,10)) + geom_point(aes(col=factor(spe),size=sqrt(traj))) + labs(y="Pseudo-time from principle curve",x="Species") + geom_vline(xintercept = 1.5)
```

###PCA for aligned 20 clusters
```{r}
aligned.20.cen <- cbind(hum.20.warped.cen[int.genes$external_gene_name.x,], mac.20.warped.cen[int.genes$mmulatta_homolog_associated_gene_name,], mou.20.warped.cen[int.genes$mmusculus_homolog_associated_gene_name,])
```

```{r fig.width=7,fig.height=5}
aligned.20.cen.norm <- apply(aligned.20.cen,2, function(x) x/sum(x)*1e4)
aligned.20.cen.pca <- prcomp(t(log(aligned.20.cen.norm+1)), scale. = TRUE, center = TRUE)

ggplot(data.frame(aligned.20.cen.pca$x, species=factor(c(rep("Human",20),rep("Monkey",20),rep("Mouse",20))),num=c(hum.20.lens,mac.20.lens,mou.20.lens)), aes(PC1,PC2)) + geom_point(aes(colour=species, size=num))+ scale_color_manual(values = c("#e35a97","#5a97e3","#5ae3a6"))
```

With SPG
```{r}
aligned.21.cen <- cbind(dge.hum.spg.cen[int.genes$external_gene_name.x],hum.20.warped.cen[int.genes$external_gene_name.x,], dge.mac.spg.cen[int.genes$mmulatta_homolog_associated_gene_name],mac.20.warped.cen[int.genes$mmulatta_homolog_associated_gene_name,], dge.mou.spg.cen[int.genes$mmusculus_homolog_associated_gene_name],mou.20.warped.cen[int.genes$mmusculus_homolog_associated_gene_name,])
```

```{r fig.width=7,fig.height=5}
aligned.21.cen.norm <- apply(aligned.21.cen,2, function(x) x/sum(x)*1e4)
aligned.21.cen.pca <- prcomp(t(log(aligned.21.cen.norm+1)), scale. = TRUE, center = TRUE)
#aligned.21.cen.repca <- prcomp(aligned.21.cen.pca$x[,c(1:2)], scale. = TRUE, center = TRUE)

ggplot(data.frame(aligned.21.cen.pca$x, species=factor(c(rep("Human",21),rep("Monkey",21),rep("Mouse",21))),num=c(dim(dge.hum.spg.raw)[2],hum.20.lens,dim(dge.mac.spg.raw)[2],mac.20.lens,dim(dge.mou.spg.raw)[2],mou.20.lens),traj=rep(c(2,rep(1,20)),3)), aes(PC1,PC3)) + geom_point(aes(colour=species, size=num, shape=factor(traj)))+ scale_color_manual(values = c("#e35a97","#5a97e3","#5ae3a6"))
```

```{r}
library(circlize)
col_fun = colorRamp2(c(-2, 1, 3.5), c("blue", "white", "red"))
```

### 20 x 20 clusters rank correlation heatmap
```{r}
aligned.20.cen.cor <- cor(log(aligned.20.cen+1),method="spearman")
Heatmap(aligned.20.cen.cor, cluster_rows = FALSE, cluster_columns = FALSE,col = col_fun(seq(-3,4)))
Heatmap(aligned.20.cen.cor, cluster_rows = FALSE, cluster_columns = FALSE)
```

##K-means genes clustering

####add SPG centroid in to the 20 aligned cluster centroids
```{r}
hum.cen <- cbind(dge.hum.spg.cen[humMacMouOrth.1to1.in$external_gene_name.x], aligned.20.all.cen[,c(1:20)])
mac.cen <- cbind(dge.mac.spg.cen[humMacMouOrth.1to1.in$mmulatta_homolog_associated_gene_name], aligned.20.all.cen[,c(21:40)])
mou.cen <- cbind(dge.mou.spg.cen[humMacMouOrth.1to1.in$mmusculus_homolog_associated_gene_name], aligned.20.all.cen[,c(41:60)])
```

```{r}
hum.cen.norm.lg <- log(apply(hum.cen,2,function(x) x/sum(x)*1e4)+1)
mac.cen.norm.lg <- log(apply(mac.cen,2,function(x) x/sum(x)*1e4)+1)
mou.cen.norm.lg <- log(apply(mou.cen,2,function(x) x/sum(x)*1e4)+1)

hum.cen.std <- t(apply(hum.cen.norm.lg,1,function(x) (x-mean(x))/sd(x)))
mac.cen.std <- t(apply(mac.cen.norm.lg,1,function(x) (x-mean(x))/sd(x)))
mou.cen.std <- t(apply(mou.cen.norm.lg,1,function(x) (x-mean(x))/sd(x)))
```

## function used for plotting dots with density as heat
```{r}
get_density <- function(x, y, ...) {
  dens <- MASS::kde2d(x, y, ...)
  ix <- findInterval(x, dens$x)
  iy <- findInterval(y, dens$y)
  ii <- cbind(ix, iy)
  return(dens$z[ii])
}
library(viridis)
library(ggExtra)
library(gridExtra)
```


1. HVGs from 1-1-1 Orthologs
```{r fig.width=5,fig.height=5}
g.mean <- apply(hum.cen.norm.lg,1,mean)
g.var <- apply(hum.cen.norm.lg,1,var)

g.var <- g.var[g.mean!=0]
g.mean <- g.mean[g.mean!=0]
hum.hvgs <- names(g.mean)[g.mean > 0.2 & g.var/g.mean > 0.2]
length(hum.hvgs)
hum.gen <- data.frame(mean = g.mean, varmean = g.var/g.mean)
hum.gen$den <- get_density(hum.gen$mean, hum.gen$varmean, n=100)

p <- ggplot(hum.gen) + geom_point(aes(x=mean, y=varmean, color = log(den+1)),size=0.5) + scale_color_viridis()+ ggtitle("Human HVGs") + theme(legend.position="none") + xlab("Mean") + ylab("Variation/Mean") + theme(plot.title = element_text(hjust = 0.5, face = "bold")) + geom_hline(yintercept = 0.2) + geom_vline(xintercept = 0.2)
p1 <- ggMarginal(p, type="density")
grid.arrange(p1, ncol=1)
```

```{r fig.width=5,fig.height=5}
g.mean <- apply(mac.cen.norm.lg,1,mean)
g.var <- apply(mac.cen.norm.lg,1,var)
g.var <- g.var[g.mean!=0]
g.mean <- g.mean[g.mean!=0]

mac.hvgs <- names(g.mean)[g.mean > 0.2 & g.var/g.mean > 0.2]
length(mac.hvgs)
mac.gen <- data.frame(mean = g.mean, varmean = g.var/g.mean)
mac.gen$den <- get_density(mac.gen$mean, mac.gen$varmean, n=100)

p <- ggplot(mac.gen) + geom_point(aes(x=mean, y=varmean, color = log(den+1)),size=0.5) + scale_color_viridis()+ ggtitle("Monkey HVGs") + theme(legend.position="none") + xlab("Mean") + ylab("Variation/Mean") + theme(plot.title = element_text(hjust = 0.5, face = "bold")) + geom_hline(yintercept = 0.2) + geom_vline(xintercept = 0.2)
p2 <- ggMarginal(p, type="density")
grid.arrange(p2, ncol=1)
```

```{r fig.width=5,fig.height=5}
g.mean <- apply(mou.cen.norm.lg,1,mean)
g.var <- apply(mou.cen.norm.lg,1,var)
g.var <- g.var[g.mean!=0]
g.mean <- g.mean[g.mean!=0]

mou.hvgs <- names(g.mean)[g.mean > 0.2 & g.var/g.mean > 0.2]
length(mou.hvgs)

mou.gen <- data.frame(mean = g.mean, varmean = g.var/g.mean)
mou.gen$den <- get_density(mou.gen$mean, mou.gen$varmean, n=100)

p <- ggplot(mou.gen) + geom_point(aes(x=mean, y=varmean, color = log(den+1)),size=0.5) + scale_color_viridis()+ ggtitle("Mouse HVGs") + theme(legend.position="none") + xlab("Mean") + ylab("Variation/Mean") + theme(plot.title = element_text(hjust = 0.5, face = "bold")) + geom_hline(yintercept = 0.2) + geom_vline(xintercept = 0.2)
p2 <- ggMarginal(p, type="density")
grid.arrange(p2, ncol=1)
```

```{r}
set.seed(123)
colnames(hum.cen.std) <- c(1:21)
hum.hvg.warped.20.kmeans <- kmeans(hum.cen.std[hum.hvgs, ],6)
table(hum.hvg.warped.20.kmeans$cluster)
for(i in c(1:6)){
dr = Heatmap(hum.cen.std[names(which(hum.hvg.warped.20.kmeans$cluster==i)),], cluster_rows = FALSE, cluster_columns = FALSE,show_row_names = FALSE,row_title = i)
draw(dr)
}
```

```{r}
set.seed(123)
colnames(mac.cen.std) <- c(1:21)
mac.hvg.warped.20.kmeans <- kmeans(mac.cen.std[mac.hvgs, ],6)
table(mac.hvg.warped.20.kmeans$cluster)
for(i in c(1:6)){
dr = Heatmap(mac.cen.std[names(which(mac.hvg.warped.20.kmeans$cluster==i)),], cluster_rows = FALSE, cluster_columns = FALSE,show_row_names = FALSE,row_title = i)
draw(dr)
}
```

```{r}
set.seed(100)
colnames(mou.cen.std) <- c(1:21)
mou.hvg.warped.20.kmeans <- kmeans(mou.cen.std[mou.hvgs, ],6)
table(mou.hvg.warped.20.kmeans$cluster)
for(i in c(1:6)){
dr = Heatmap(mou.cen.std[names(which(mou.hvg.warped.20.kmeans$cluster==i)),], cluster_rows = FALSE, cluster_columns = FALSE,show_row_names = FALSE,row_title = i)
draw(dr)
}
```

Write table
```{r}
len <- max(max(table(hum.hvg.warped.20.kmeans$cluster)))
hum.hvgs.kmean.write <- c()
stages <- c(1,3,6,4,5,2)
for(i in stages){
  genes <- names(which(hum.hvg.warped.20.kmeans$cluster==i))
  hum.hvgs.kmean.write <- cbind(hum.hvgs.kmean.write, c(genes, rep("", len-length(genes) )))
}
write.table(hum.hvgs.kmean.write, file="~/storage/HumanMacaqueMouseCompAnaysis/interdata/Germ_kmeans/hum.germ.kmean.ortho.list.txt", quote = FALSE, row.names = FALSE,sep = "\t")

len <- max(max(table(mac.hvg.warped.20.kmeans$cluster)))
mac.hvgs.kmean.write <- c()
stages <- c(6,3,2,5,1,4)
for(i in stages){
  genes <- names(which(mac.hvg.warped.20.kmeans$cluster==i))
  mac.hvgs.kmean.write <- cbind(mac.hvgs.kmean.write, c(genes, rep("", len-length(genes) )))
}
write.table(mac.hvgs.kmean.write, file="~/storage/HumanMacaqueMouseCompAnaysis/interdata/Germ_kmeans/mac.germ.kmean.ortho.list.txt", quote = FALSE, row.names = FALSE,sep = "\t")

len <- max(max(table(mou.hvg.warped.20.kmeans$cluster)))
mou.hvgs.kmean.write <- c()
stages <- c(2,5,6,4,3,1)
for(i in stages){
  genes <- names(which(mou.hvg.warped.20.kmeans$cluster==i))
  mou.hvgs.kmean.write <- cbind(mou.hvgs.kmean.write, c(genes, rep("", len-length(genes) )))
}
write.table(mou.hvgs.kmean.write, file="~/storage/HumanMacaqueMouseCompAnaysis/interdata/Germ_kmeans/mou.germ.kmean.ortho.list.txt", quote = FALSE, row.names = FALSE,sep = "\t")
```

##.Connecting plot
```{r}
hum.hvg.warped.kmeans.genes <- c()
stages <- c(1,3,6,4,5,2)
for(i in c(1:6)){
  hum.hvg.warped.kmeans.genes <- rbind(hum.hvg.warped.kmeans.genes, data.frame(genes = names(which(hum.hvg.warped.20.kmeans$cluster==stages[i])), stages=rep(i, sum(hum.hvg.warped.20.kmeans$cluster==stages[i]))))
}

mac.hvg.warped.kmeans.genes <- c()
stages <- c(6,3,2,5,1,4)
for(i in c(1:6)){
  mac.hvg.warped.kmeans.genes <- rbind(mac.hvg.warped.kmeans.genes, data.frame(genes = names(which(mac.hvg.warped.20.kmeans$cluster==stages[i])), stages=rep(i, sum(mac.hvg.warped.20.kmeans$cluster==stages[i]))))
}

mou.hvg.warped.kmeans.genes <- c()
stages <- c(2,5,6,4,3,1)
for(i in c(1:6)){
  mou.hvg.warped.kmeans.genes <- rbind(mou.hvg.warped.kmeans.genes, data.frame(genes = names(which(mou.hvg.warped.20.kmeans$cluster==stages[i])), stages=rep(i, sum(mou.hvg.warped.20.kmeans$cluster==stages[i]))))
}
```

##human Mac
#Connector lines
```{r}
kmeans.hum.mac.genes <- humMacMouOrth.1to1.in[humMacMouOrth.1to1.in$external_gene_name.x %in% hum.hvgs,]
kmeans.hum.mac.genes <- kmeans.hum.mac.genes[kmeans.hum.mac.genes$mmulatta_homolog_associated_gene_name %in% mac.hvgs,]

pos.hum.mac <- c()
for(i in c(1:dim(kmeans.hum.mac.genes)[1])){
  hum.gen <- kmeans.hum.mac.genes$external_gene_name.x[i]
  mac.gen <- kmeans.hum.mac.genes$mmulatta_homolog_associated_gene_name[i]
  x = which(hum.hvg.warped.kmeans.genes$genes == hum.gen)
  y = which(mac.hvg.warped.kmeans.genes$genes == mac.gen)
  pos.hum.mac <- rbind(pos.hum.mac,c(x,y,hum.gen, mac.gen))
}
pos.hum.mac <- data.matrix(pos.hum.mac)
colnames(pos.hum.mac) = c("hum.pos","mac.pos", "hum.gen","mac.gen")


#Sort within each cluster

rownames(pos.hum.mac) <- pos.hum.mac[,"hum.gen"]
hum.kmeans.genes.ord <- hum.hvg.warped.kmeans.genes
for(i in c(1:6)){
  genes <- hum.hvg.warped.kmeans.genes$genes[hum.hvg.warped.kmeans.genes$stages==i]
  ogenes <- as.character(genes[!genes %in% pos.hum.mac[,"hum.gen"]])
  genes <- pos.hum.mac[,"hum.gen"][pos.hum.mac[,"hum.gen"] %in% genes]
  genes <- genes[order(as.numeric(pos.hum.mac[genes,"mac.pos"]))]
  hum.kmeans.genes.ord$genes[hum.hvg.warped.kmeans.genes$stages==i] = c(genes,ogenes)
}

rownames(pos.hum.mac) <- pos.hum.mac[,"mac.gen"]
mac.kmeans.genes.ord <- mac.hvg.warped.kmeans.genes
for(i in c(1:6)){
  genes <- mac.hvg.warped.kmeans.genes$genes[mac.hvg.warped.kmeans.genes$stages==i]
  ogenes <- as.character(genes[!genes %in% pos.hum.mac[,"mac.gen"]])
  genes <- pos.hum.mac[,"mac.gen"][pos.hum.mac[,"mac.gen"] %in% genes]
  genes <- genes[order(as.numeric(pos.hum.mac[genes,"hum.pos"]))]
  mac.kmeans.genes.ord$genes[mac.hvg.warped.kmeans.genes$stages==i] = c(genes,ogenes)
}
```

```{r fig.width=4, fig.height=10}
cols <- c("#fabc3c",
"#a64dd4",
"#14e072",
"#d9014a",
"#abb4ff",
"#b28e00")
plot(1,xlim=c(1,2),ylim=c(0,1000),type="n",axes=FALSE,ann=FALSE)

for(i in c(1:dim(kmeans.hum.mac.genes)[1])){
#for(i in c(1:10)){ 
  x = which(hum.kmeans.genes.ord$genes == kmeans.hum.mac.genes$external_gene_name.x[i])
  y = which(mac.kmeans.genes.ord$genes == kmeans.hum.mac.genes$mmulatta_homolog_associated_gene_name[i])
  #print(kmeans.hum.mac.genes$external_gene_name.x[i])
  #print(kmeans.hum.mac.genes$mmusculus_homolog_associated_gene_name[i])
  x = 1000-x*1000/dim(hum.hvg.warped.kmeans.genes)[1]
  y = 1000-y*1000/dim(mac.hvg.warped.kmeans.genes)[1]
  #print(x)
  #print(y)
  segments(1,x,2,y,lwd = 0.1)
}
points(data.frame(x=rep(1,dim(hum.hvg.warped.kmeans.genes)[1]),y=1000-c(1:dim(hum.hvg.warped.kmeans.genes)[1])*1000/dim(hum.hvg.warped.kmeans.genes)[1]),col=cols[hum.hvg.warped.kmeans.genes$stages], pch=15, cex=0.8)
points(data.frame(x=rep(2,dim(mac.hvg.warped.kmeans.genes)[1]),y=1000-c(1:dim(mac.hvg.warped.kmeans.genes)[1])*1000/dim(mac.hvg.warped.kmeans.genes)[1]),col=cols[mac.hvg.warped.kmeans.genes$stages], pch=15, cex=0.8)
#segments(1,0,10,0,col = "red")
```

```{r fig.width=4, fig.height=10}
cols <- c("#fabc3c",
"#a64dd4",
"#14e072",
"#d9014a",
"#abb4ff",
"#b28e00")
ha = rowAnnotation(stages = mac.hvg.warped.kmeans.genes$stages, col = list(stages = c("1" = cols[1], "2" = cols[2], "3" = cols[3],"4" = cols[4], "5" = cols[5], "6" = cols[6])))
#ha = HeatmapAnnotation(foo = anno_simple(cbind(1:10, 10:1), pch = 1:10))
Heatmap(mac.cen.std[as.character(mac.hvg.warped.kmeans.genes$genes),], cluster_rows = FALSE, cluster_columns = FALSE,show_row_names = FALSE, left_annotation = ha)
```

```{r fig.width=4, fig.height=10}
cols <- c("#fabc3c",
"#a64dd4",
"#14e072",
"#d9014a",
"#abb4ff",
"#b28e00")
ha = rowAnnotation(stages = hum.hvg.warped.kmeans.genes$stages, col = list(stages = c("1" = cols[1], "2" = cols[2], "3" = cols[3],"4" = cols[4], "5" = cols[5], "6" = cols[6])))
#ha = HeatmapAnnotation(foo = anno_simple(cbind(1:10, 10:1), pch = 1:10))
Heatmap(hum.cen.std[as.character(hum.hvg.warped.kmeans.genes$genes),], cluster_rows = FALSE, cluster_columns = FALSE,show_row_names = FALSE, right_annotation = ha)
```

##human Mou
#Connector lines
```{r}
kmeans.hum.mou.genes <- humMacMouOrth.1to1.in[humMacMouOrth.1to1.in$external_gene_name.x %in% hum.hvgs,]
kmeans.hum.mou.genes <- kmeans.hum.mou.genes[kmeans.hum.mou.genes$mmusculus_homolog_associated_gene_name %in% mou.hvgs,]

pos.hum.mou <- c()
for(i in c(1:dim(kmeans.hum.mou.genes)[1])){
  hum.gen <- kmeans.hum.mou.genes$external_gene_name.x[i]
  mou.gen <- kmeans.hum.mou.genes$mmusculus_homolog_associated_gene_name[i]
  x = which(hum.hvg.warped.kmeans.genes$genes == hum.gen)
  y = which(mou.hvg.warped.kmeans.genes$genes == mou.gen)
  pos.hum.mou <- rbind(pos.hum.mou,c(x,y,hum.gen, mou.gen))
}
pos.hum.mou <- data.matrix(pos.hum.mou)
colnames(pos.hum.mou) = c("hum.pos","mou.pos", "hum.gen","mou.gen")

rownames(pos.hum.mou) <- pos.hum.mou[,"hum.gen"]
hum.kmeans.genes.ord <- hum.hvg.warped.kmeans.genes
for(i in c(1:6)){
  genes <- hum.hvg.warped.kmeans.genes$genes[hum.hvg.warped.kmeans.genes$stages==i]
  ogenes <- as.character(genes[!genes %in% pos.hum.mou[,"hum.gen"]])
  genes <- pos.hum.mou[,"hum.gen"][pos.hum.mou[,"hum.gen"] %in% genes]
  genes <- genes[order(as.numeric(pos.hum.mou[genes,"mou.pos"]))]
  hum.kmeans.genes.ord$genes[hum.hvg.warped.kmeans.genes$stages==i] = c(genes,ogenes)
}

rownames(pos.hum.mou) <- pos.hum.mou[,"mou.gen"]
mou.kmeans.genes.ord <- mou.hvg.warped.kmeans.genes
for(i in c(1:6)){
  genes <- mou.hvg.warped.kmeans.genes$genes[mou.hvg.warped.kmeans.genes$stages==i]
  ogenes <- as.character(genes[!genes %in% pos.hum.mou[,"mou.gen"]])
  genes <- pos.hum.mou[,"mou.gen"][pos.hum.mou[,"mou.gen"] %in% genes]
  genes <- genes[order(as.numeric(pos.hum.mou[genes,"hum.pos"]))]
  mou.kmeans.genes.ord$genes[mou.hvg.warped.kmeans.genes$stages==i] = c(genes,ogenes)
}
```


```{r fig.width=4, fig.height=10}
cols <- c("#fabc3c",
"#a64dd4",
"#14e072",
"#d9014a",
"#abb4ff",
"#b28e00")
plot(1,xlim=c(1,2),ylim=c(0,1000),type="n",axes=FALSE,ann=FALSE)

for(i in c(1:dim(kmeans.hum.mou.genes)[1])){
#for(i in c(1:10)){ 
  x = which(hum.kmeans.genes.ord$genes == kmeans.hum.mou.genes$external_gene_name.x[i])
  y = which(mou.kmeans.genes.ord$genes == kmeans.hum.mou.genes$mmusculus_homolog_associated_gene_name[i])
  #print(kmeans.hum.mou.genes$external_gene_name.x[i])
  #print(kmeans.hum.mou.genes$mmusculus_homolog_associated_gene_name[i])
  x = 1000-x*1000/dim(hum.hvg.warped.kmeans.genes)[1]
  y = 1000-y*1000/dim(mou.hvg.warped.kmeans.genes)[1]
  #print(x)
  #print(y)
  segments(1,x,2,y,lwd = 0.1)
}
points(data.frame(x=rep(1,dim(hum.hvg.warped.kmeans.genes)[1]),y=1000-c(1:dim(hum.hvg.warped.kmeans.genes)[1])*1000/dim(hum.hvg.warped.kmeans.genes)[1]),col=cols[hum.hvg.warped.kmeans.genes$stages], pch=15, cex=0.8)
points(data.frame(x=rep(2,dim(mou.hvg.warped.kmeans.genes)[1]),y=1000-c(1:dim(mou.hvg.warped.kmeans.genes)[1])*1000/dim(mou.hvg.warped.kmeans.genes)[1]),col=cols[mou.hvg.warped.kmeans.genes$stages], pch=15, cex=0.8)
#segments(1,0,10,0,col = "red")
```

```{r fig.width=4, fig.height=10}
cols <- c("#fabc3c",
"#a64dd4",
"#14e072",
"#d9014a",
"#abb4ff",
"#b28e00")
ha = rowAnnotation(stages = hum.hvg.warped.kmeans.genes$stages, col = list(stages = c("1" = cols[1], "2" = cols[2], "3" = cols[3],"4" = cols[4], "5" = cols[5], "6" = cols[6])))
#ha = HeatmapAnnotation(foo = anno_simple(cbind(1:10, 10:1), pch = 1:10))
Heatmap(hum.cen.std[as.character(hum.hvg.warped.kmeans.genes$genes),], cluster_rows = FALSE, cluster_columns = FALSE,show_row_names = FALSE, right_annotation = ha)
```

```{r fig.width=4, fig.height=10}
cols <- c("#fabc3c",
"#a64dd4",
"#14e072",
"#d9014a",
"#abb4ff",
"#b28e00")
ha = rowAnnotation(stages = mou.hvg.warped.kmeans.genes$stages, col = list(stages = c("1" = cols[1], "2" = cols[2], "3" = cols[3],"4" = cols[4], "5" = cols[5], "6" = cols[6])))
#ha = HeatmapAnnotation(foo = anno_simple(cbind(1:10, 10:1), pch = 1:10))
Heatmap(mou.cen.std[as.character(mou.hvg.warped.kmeans.genes$genes),], cluster_rows = FALSE, cluster_columns = FALSE,show_row_names = FALSE, left_annotation = ha)
```

##Monkey Mou
#Connector lines
```{r}
kmeans.mac.mou.genes <- humMacMouOrth.1to1.in[humMacMouOrth.1to1.in$mmulatta_homolog_associated_gene_name %in% mac.hvgs,]
kmeans.mac.mou.genes <- kmeans.mac.mou.genes[kmeans.mac.mou.genes$mmusculus_homolog_associated_gene_name %in% mou.hvgs,]

pos.mac.mou <- c()
for(i in c(1:dim(kmeans.mac.mou.genes)[1])){
  mac.gen <- kmeans.mac.mou.genes$mmulatta_homolog_associated_gene_name[i]
  mou.gen <- kmeans.mac.mou.genes$mmusculus_homolog_associated_gene_name[i]
  x = which(mac.hvg.warped.kmeans.genes$genes == mac.gen)
  y = which(mou.hvg.warped.kmeans.genes$genes == mou.gen)
  pos.mac.mou <- rbind(pos.mac.mou,c(x,y,mac.gen, mou.gen))
}
pos.mac.mou <- data.matrix(pos.mac.mou)
colnames(pos.mac.mou) = c("mac.pos","mou.pos", "mac.gen","mou.gen")


rownames(pos.mac.mou) <- pos.mac.mou[,"mac.gen"]
mac.kmeans.genes.ord <- mac.hvg.warped.kmeans.genes
for(i in c(1:6)){
  genes <- mac.hvg.warped.kmeans.genes$genes[mac.hvg.warped.kmeans.genes$stages==i]
  ogenes <- as.character(genes[!genes %in% pos.mac.mou[,"mac.gen"]])
  genes <- pos.mac.mou[,"mac.gen"][pos.mac.mou[,"mac.gen"] %in% genes]
  genes <- genes[order(as.numeric(pos.mac.mou[genes,"mou.pos"]))]
  mac.kmeans.genes.ord$genes[mac.hvg.warped.kmeans.genes$stages==i] = c(genes,ogenes)
}

rownames(pos.mac.mou) <- pos.mac.mou[,"mou.gen"]
mou.kmeans.genes.ord <- mou.hvg.warped.kmeans.genes
for(i in c(1:6)){
  genes <- mou.hvg.warped.kmeans.genes$genes[mou.hvg.warped.kmeans.genes$stages==i]
  ogenes <- as.character(genes[!genes %in% pos.mac.mou[,"mou.gen"]])
  genes <- pos.mac.mou[,"mou.gen"][pos.mac.mou[,"mou.gen"] %in% genes]
  genes <- genes[order(as.numeric(pos.mac.mou[genes,"mac.pos"]))]
  mou.kmeans.genes.ord$genes[mou.hvg.warped.kmeans.genes$stages==i] = c(genes,ogenes)
}
```


```{r fig.width=4, fig.height=10}
cols <- c("#fabc3c",
"#a64dd4",
"#14e072",
"#d9014a",
"#abb4ff",
"#b28e00")
plot(1,xlim=c(1,2),ylim=c(0,1000),type="n",axes=FALSE,ann=FALSE)

for(i in c(1:dim(kmeans.mac.mou.genes)[1])){
#for(i in c(1:10)){ 
  x = which(mac.kmeans.genes.ord$genes == kmeans.mac.mou.genes$mmulatta_homolog_associated_gene_name[i])
  y = which(mou.kmeans.genes.ord$genes == kmeans.mac.mou.genes$mmusculus_homolog_associated_gene_name[i])
  #print(kmeans.hum.mou.genes$external_gene_name.x[i])
  #print(kmeans.hum.mou.genes$mmusculus_homolog_associated_gene_name[i])
  x = 1000-x*1000/dim(mac.hvg.warped.kmeans.genes)[1]
  y = 1000-y*1000/dim(mou.hvg.warped.kmeans.genes)[1]
  #print(x)
  #print(y)
  segments(1,x,2,y,lwd = 0.1)
}
points(data.frame(x=rep(1,dim(mac.hvg.warped.kmeans.genes)[1]),y=1000-c(1:dim(mac.hvg.warped.kmeans.genes)[1])*1000/dim(mac.hvg.warped.kmeans.genes)[1]),col=cols[mac.hvg.warped.kmeans.genes$stages], pch=15, cex=0.8)
points(data.frame(x=rep(2,dim(mou.hvg.warped.kmeans.genes)[1]),y=1000-c(1:dim(mou.hvg.warped.kmeans.genes)[1])*1000/dim(mou.hvg.warped.kmeans.genes)[1]),col=cols[mou.hvg.warped.kmeans.genes$stages], pch=15, cex=0.8)
#segments(1,0,10,0,col = "red")
```

```{r fig.width=4, fig.height=10}
cols <- c("#fabc3c",
"#a64dd4",
"#14e072",
"#d9014a",
"#abb4ff",
"#b28e00")
ha = rowAnnotation(stages = mou.hvg.warped.kmeans.genes$stages, col = list(stages = c("1" = cols[1], "2" = cols[2], "3" = cols[3],"4" = cols[4], "5" = cols[5], "6" = cols[6])))
#ha = HeatmapAnnotation(foo = anno_simple(cbind(1:10, 10:1), pch = 1:10))
Heatmap(mou.cen.std[as.character(mou.hvg.warped.kmeans.genes$genes),], cluster_rows = FALSE, cluster_columns = FALSE,show_row_names = FALSE, left_annotation = ha)
```


```{r fig.width=4, fig.height=10}
cols <- c("#fabc3c",
"#a64dd4",
"#14e072",
"#d9014a",
"#abb4ff",
"#b28e00")
ha = rowAnnotation(stages = mac.hvg.warped.kmeans.genes$stages, col = list(stages = c("1" = cols[1], "2" = cols[2], "3" = cols[3],"4" = cols[4], "5" = cols[5], "6" = cols[6])))
#ha = HeatmapAnnotation(foo = anno_simple(cbind(1:10, 10:1), pch = 1:10))
Heatmap(mac.cen.std[as.character(mac.hvg.warped.kmeans.genes$genes),], cluster_rows = FALSE, cluster_columns = FALSE,show_row_names = FALSE, right_annotation = ha)
```


Monkey extra 5 clusters
```{r}
mac.extra.cen <- matrix(0,nrow=dim(dge.mac.germ.norm)[1],ncol=5)
mac.extra.cen[,1] <- apply(dge.mac.germ.norm[,unlist(lapply(c(143:154),function(x) hummac.mapping$queryAssign[[x]]))],1,mean)
mac.extra.cen[,2] <- apply(dge.mac.germ.norm[,unlist(lapply(c(155:166),function(x) hummac.mapping$queryAssign[[x]]))],1,mean)
mac.extra.cen[,3] <- apply(dge.mac.germ.norm[,unlist(lapply(c(167:178),function(x) hummac.mapping$queryAssign[[x]]))],1,mean)
mac.extra.cen[,4] <- apply(dge.mac.germ.norm[,unlist(lapply(c(179:189),function(x) hummac.mapping$queryAssign[[x]]))],1,mean)
mac.extra.cen[,5] <- apply(dge.mac.germ.norm[,unlist(lapply(c(190:200),function(x) hummac.mapping$queryAssign[[x]]))],1,mean)
```

```{r}
rownames(mac.extra.cen) <- rownames(dge.mac.spg.norm)
mac.cen.longer <- cbind(mac.cen, mac.extra.cen[humMacMouOrth.1to1.in$mmulatta_homolog_associated_gene_name,])
mac.cen.longer.lg <- log(mac.cen.longer+1)
mac.cen.longer.std <- t(apply(mac.cen.longer.lg,1,function(x) (x-mean(x))/sd(x)))
```

```{r fig.width=5,fig.height=5}
g.mean <- apply(mac.cen.longer.lg,1,mean)
g.var <- apply(mac.cen.longer.lg,1,var)
plot(g.mean, g.var/g.mean, cex=0.5)
mac.hvgs.longer <- names(g.mean)[g.mean > 0.2 & g.var/g.mean > 0.2]
abline(h=0.2)
abline(v=0.2)
length(mac.hvgs.longer)
```


```{r fig.width=5,fig.height=5}
g.mean <- apply(mac.cen.longer.lg,1,mean)
g.var <- apply(mac.cen.longer.lg,1,var)
g.var <- g.var[g.mean!=0]
g.mean <- g.mean[g.mean!=0]

mac.hvgs.longer <- names(g.mean)[g.mean > 0.2 & g.var/g.mean > 0.15]
length(mac.hvgs.longer)
mac.gen <- data.frame(mean = g.mean, varmean = g.var/g.mean)
mac.gen$den <- get_density(mac.gen$mean, mac.gen$varmean, n=100)

p <- ggplot(mac.gen) + geom_point(aes(x=mean, y=varmean, color = log(den+1)),size=0.5) + scale_color_viridis()+ ggtitle("Monkey HVGs") + theme(legend.position="none") + xlab("Mean") + ylab("Variation/Mean") + theme(plot.title = element_text(hjust = 0.5, face = "bold")) + geom_hline(yintercept = 0.15) + geom_vline(xintercept = 0.2)
p2 <- ggMarginal(p, type="density")
grid.arrange(p2, ncol=1)
```

```{r}
set.seed(123)
colnames(mac.cen.longer.std) <- c(1:26)
mac.hvg.warped.25.kmeans <- kmeans(mac.cen.longer.std[mac.hvgs.longer, ],7)
table(mac.hvg.warped.25.kmeans$cluster)
for(i in c(1:7)){
dr = Heatmap(mac.cen.longer.std[names(which(mac.hvg.warped.25.kmeans$cluster==i)),], cluster_rows = FALSE, cluster_columns = FALSE,show_row_names = FALSE,row_title = i)
draw(dr)
}
```

```{r}
len <- max(max(table(mac.hvg.warped.25.kmeans$cluster)))
mac.hvgs.kmean.write <- c()
stages <- c(3,4,5,2,7,1,6)
for(i in stages){
  genes <- names(which(mac.hvg.warped.25.kmeans$cluster==i))
  mac.hvgs.kmean.write <- cbind(mac.hvgs.kmean.write, c(genes, rep("", len-length(genes) )))
}
write.table(mac.hvgs.kmean.write, file="~/storage/HumanMacaqueMouseCompAnaysis/interdata/Germ_kmeans/mac.germ.longer.kmean.ortho.list.txt", quote = FALSE, row.names = FALSE,sep = "\t")
```


##.Connecting plot
```{r}
mac.hvg.warped.kmeans.genes <- c()
stages <- c(3,4,5,2,7,1,6)
for(i in c(1:7)){
  mac.hvg.warped.kmeans.genes <- rbind(mac.hvg.warped.kmeans.genes, data.frame(genes = names(which(mac.hvg.warped.25.kmeans$cluster==stages[i])), stages=rep(i, sum(mac.hvg.warped.25.kmeans$cluster==stages[i]))))
}
```

##human Mac
#Connector lines
```{r}
kmeans.hum.mac.genes <- humMacMouOrth.1to1.in[humMacMouOrth.1to1.in$external_gene_name.x %in% hum.hvgs,]
kmeans.hum.mac.genes <- kmeans.hum.mac.genes[kmeans.hum.mac.genes$mmulatta_homolog_associated_gene_name %in% mac.hvgs.longer,]
```

```{r}
pos.hum.mac <- c()
for(i in c(1:dim(kmeans.hum.mac.genes)[1])){
  hum.gen <- kmeans.hum.mac.genes$external_gene_name.x[i]
  mac.gen <- kmeans.hum.mac.genes$mmulatta_homolog_associated_gene_name[i]
  x = which(hum.hvg.warped.kmeans.genes$genes == hum.gen)
  y = which(mac.hvg.warped.kmeans.genes$genes == mac.gen)
  pos.hum.mac <- rbind(pos.hum.mac,c(x,y,hum.gen, mac.gen))
}
pos.hum.mac <- data.matrix(pos.hum.mac)
colnames(pos.hum.mac) = c("hum.pos","mac.pos", "hum.gen","mac.gen")


rownames(pos.hum.mac) <- pos.hum.mac[,"hum.gen"]
hum.kmeans.genes.ord <- hum.hvg.warped.kmeans.genes
for(i in c(1:6)){
  genes <- hum.hvg.warped.kmeans.genes$genes[hum.hvg.warped.kmeans.genes$stages==i]
  ogenes <- as.character(genes[!genes %in% pos.hum.mac[,"hum.gen"]])
  genes <- pos.hum.mac[,"hum.gen"][pos.hum.mac[,"hum.gen"] %in% genes]
  genes <- genes[order(as.numeric(pos.hum.mac[genes,"mac.pos"]))]
  hum.kmeans.genes.ord$genes[hum.hvg.warped.kmeans.genes$stages==i] = c(genes,ogenes)
}

rownames(pos.hum.mac) <- pos.hum.mac[,"mac.gen"]
mac.kmeans.genes.ord <- mac.hvg.warped.kmeans.genes
for(i in c(1:7)){
  genes <- mac.hvg.warped.kmeans.genes$genes[mac.hvg.warped.kmeans.genes$stages==i]
  ogenes <- as.character(genes[!genes %in% pos.hum.mac[,"mac.gen"]])
  genes <- pos.hum.mac[,"mac.gen"][pos.hum.mac[,"mac.gen"] %in% genes]
  genes <- genes[order(as.numeric(pos.hum.mac[genes,"hum.pos"]))]
  mac.kmeans.genes.ord$genes[mac.hvg.warped.kmeans.genes$stages==i] = c(genes,ogenes)
}
```


```{r fig.width=4, fig.height=10}
cols <- c("#fabc3c",
"#a64dd4",
"#14e072",
"#d9014a",
"#abb4ff",
"#b28e00",
"#e35a97")
plot(1,xlim=c(1,2),ylim=c(0,1000),type="n",axes=FALSE,ann=FALSE)

for(i in c(1:dim(kmeans.hum.mac.genes)[1])){
#for(i in c(1:10)){ 
  x = which(hum.kmeans.genes.ord$genes == kmeans.hum.mac.genes$external_gene_name.x[i])
  y = which(mac.kmeans.genes.ord$genes == kmeans.hum.mac.genes$mmulatta_homolog_associated_gene_name[i])
  #print(kmeans.hum.mac.genes$external_gene_name.x[i])
  #print(kmeans.hum.mac.genes$mmusculus_homolog_associated_gene_name[i])
  x = 1000-x*1000/dim(hum.hvg.warped.kmeans.genes)[1]
  y = 1000-y*1000/dim(mac.hvg.warped.kmeans.genes)[1]
  #print(x)
  #print(y)
  segments(1,x,2,y,lwd = 0.1)
}
points(data.frame(x=rep(1,dim(hum.hvg.warped.kmeans.genes)[1]),y=1000-c(1:dim(hum.hvg.warped.kmeans.genes)[1])*1000/dim(hum.hvg.warped.kmeans.genes)[1]),col=cols[hum.hvg.warped.kmeans.genes$stages], pch=15, cex=0.8)
points(data.frame(x=rep(2,dim(mac.hvg.warped.kmeans.genes)[1]),y=1000-c(1:dim(mac.hvg.warped.kmeans.genes)[1])*1000/dim(mac.hvg.warped.kmeans.genes)[1]),col=cols[mac.hvg.warped.kmeans.genes$stages], pch=15, cex=0.8)
#segments(1,0,10,0,col = "red")
```

```{r fig.width=4, fig.height=10}
cols <- c("#fabc3c",
"#a64dd4",
"#14e072",
"#d9014a",
"#abb4ff",
"#b28e00",
"#e35a97")
ha = rowAnnotation(stages = mac.hvg.warped.kmeans.genes$stages, col = list(stages = c("1" = cols[1], "2" = cols[2], "3" = cols[3],"4" = cols[4], "5" = cols[5], "6" = cols[6], "7" = cols[7])))
#ha = HeatmapAnnotation(foo = anno_simple(cbind(1:10, 10:1), pch = 1:10))
Heatmap(mac.cen.std[as.character(mac.hvg.warped.kmeans.genes$genes),], cluster_rows = FALSE, cluster_columns = FALSE,show_row_names = FALSE, left_annotation = ha)
```

```{r fig.width=4, fig.height=10}
cols <- c("#fabc3c",
"#a64dd4",
"#14e072",
"#d9014a",
"#abb4ff",
"#b28e00")
ha = rowAnnotation(stages = hum.hvg.warped.kmeans.genes$stages, col = list(stages = c("1" = cols[1], "2" = cols[2], "3" = cols[3],"4" = cols[4], "5" = cols[5], "6" = cols[6])))
#ha = HeatmapAnnotation(foo = anno_simple(cbind(1:10, 10:1), pch = 1:10))
Heatmap(hum.cen.std[as.character(hum.hvg.warped.kmeans.genes$genes),], cluster_rows = FALSE, cluster_columns = FALSE,show_row_names = FALSE, right_annotation = ha)
```

##Monkey Mou
#Connector lines
```{r}
kmeans.mac.mou.genes <- humMacMouOrth.1to1.in[humMacMouOrth.1to1.in$mmulatta_homolog_associated_gene_name %in% mac.hvgs.longer,]
kmeans.mac.mou.genes <- kmeans.mac.mou.genes[kmeans.mac.mou.genes$mmusculus_homolog_associated_gene_name %in% mou.hvgs,]

pos.mac.mou <- c()
for(i in c(1:dim(kmeans.mac.mou.genes)[1])){
  mac.gen <- kmeans.mac.mou.genes$mmulatta_homolog_associated_gene_name[i]
  mou.gen <- kmeans.mac.mou.genes$mmusculus_homolog_associated_gene_name[i]
  x = which(mac.hvg.warped.kmeans.genes$genes == mac.gen)
  y = which(mou.hvg.warped.kmeans.genes$genes == mou.gen)
  pos.mac.mou <- rbind(pos.mac.mou,c(x,y,mac.gen, mou.gen))
}
pos.mac.mou <- data.matrix(pos.mac.mou)
colnames(pos.mac.mou) = c("mac.pos","mou.pos", "mac.gen","mou.gen")


rownames(pos.mac.mou) <- pos.mac.mou[,"mac.gen"]
mac.kmeans.genes.ord <- mac.hvg.warped.kmeans.genes
for(i in c(1:7)){
  genes <- mac.hvg.warped.kmeans.genes$genes[mac.hvg.warped.kmeans.genes$stages==i]
  ogenes <- as.character(genes[!genes %in% pos.mac.mou[,"mac.gen"]])
  genes <- pos.mac.mou[,"mac.gen"][pos.mac.mou[,"mac.gen"] %in% genes]
  genes <- genes[order(as.numeric(pos.mac.mou[genes,"mou.pos"]))]
  mac.kmeans.genes.ord$genes[mac.hvg.warped.kmeans.genes$stages==i] = c(genes,ogenes)
}

rownames(pos.mac.mou) <- pos.mac.mou[,"mou.gen"]
mou.kmeans.genes.ord <- mou.hvg.warped.kmeans.genes
for(i in c(1:6)){
  genes <- mou.hvg.warped.kmeans.genes$genes[mou.hvg.warped.kmeans.genes$stages==i]
  ogenes <- as.character(genes[!genes %in% pos.mac.mou[,"mou.gen"]])
  genes <- pos.mac.mou[,"mou.gen"][pos.mac.mou[,"mou.gen"] %in% genes]
  genes <- genes[order(as.numeric(pos.mac.mou[genes,"mac.pos"]))]
  mou.kmeans.genes.ord$genes[mou.hvg.warped.kmeans.genes$stages==i] = c(genes,ogenes)
}
```


```{r fig.width=4, fig.height=10}
cols <- c("#fabc3c",
"#a64dd4",
"#14e072",
"#d9014a",
"#abb4ff",
"#b28e00",
"#e35a97")
plot(1,xlim=c(1,2),ylim=c(0,1000),type="n",axes=FALSE,ann=FALSE)

for(i in c(1:dim(kmeans.mac.mou.genes)[1])){
#for(i in c(1:10)){ 
  x = which(mac.kmeans.genes.ord$genes == kmeans.mac.mou.genes$mmulatta_homolog_associated_gene_name[i])
  y = which(mou.kmeans.genes.ord$genes == kmeans.mac.mou.genes$mmusculus_homolog_associated_gene_name[i])
  #print(kmeans.hum.mou.genes$external_gene_name.x[i])
  #print(kmeans.hum.mou.genes$mmusculus_homolog_associated_gene_name[i])
  x = 1000-x*1000/dim(mac.hvg.warped.kmeans.genes)[1]
  y = 1000-y*1000/dim(mou.hvg.warped.kmeans.genes)[1]
  #print(x)
  #print(y)
  segments(1,x,2,y,lwd = 0.1)
}
points(data.frame(x=rep(1,dim(mac.hvg.warped.kmeans.genes)[1]),y=1000-c(1:dim(mac.hvg.warped.kmeans.genes)[1])*1000/dim(mac.hvg.warped.kmeans.genes)[1]),col=cols[mac.hvg.warped.kmeans.genes$stages], pch=15, cex=0.8)
points(data.frame(x=rep(2,dim(mou.hvg.warped.kmeans.genes)[1]),y=1000-c(1:dim(mou.hvg.warped.kmeans.genes)[1])*1000/dim(mou.hvg.warped.kmeans.genes)[1]),col=cols[mou.hvg.warped.kmeans.genes$stages], pch=15, cex=0.8)
#segments(1,0,10,0,col = "red")
```

```{r fig.width=4, fig.height=10}
cols <- c("#fabc3c",
"#a64dd4",
"#14e072",
"#d9014a",
"#abb4ff",
"#b28e00")
ha = rowAnnotation(stages = mou.hvg.warped.kmeans.genes$stages, col = list(stages = c("1" = cols[1], "2" = cols[2], "3" = cols[3],"4" = cols[4], "5" = cols[5], "6" = cols[6])))
#ha = HeatmapAnnotation(foo = anno_simple(cbind(1:10, 10:1), pch = 1:10))
Heatmap(mou.cen.std[as.character(mou.hvg.warped.kmeans.genes$genes),], cluster_rows = FALSE, cluster_columns = FALSE,show_row_names = FALSE, left_annotation = ha)
```


```{r fig.width=4, fig.height=10}
cols <- c("#fabc3c",
"#a64dd4",
"#14e072",
"#d9014a",
"#abb4ff",
"#b28e00",
"#e25a97")
ha = rowAnnotation(stages = mac.hvg.warped.kmeans.genes$stages, col = list(stages = c("1" = cols[1], "2" = cols[2], "3" = cols[3],"4" = cols[4], "5" = cols[5], "6" = cols[6], "7" = cols[7])))
#ha = HeatmapAnnotation(foo = anno_simple(cbind(1:10, 10:1), pch = 1:10))
Heatmap(mac.cen.std[as.character(mac.hvg.warped.kmeans.genes$genes),], cluster_rows = FALSE, cluster_columns = FALSE,show_row_names = FALSE, right_annotation = ha)
```


1. Read in the datasets for germ cells and spg (4-62)
2. Read in/Run monocle for three species (67-175)
3. Run cellAlign to get 200 pseudo-time points (178-257)
4. Caclulate 200 centroids (260-320)
5. PCA on centroids (325-344)
6. Analysis for gene expression pattern changes along the trajectory for multiple categories of orthologues. (347-1142)
7. Trajectory alignment using Principle curve on the PC1-PC3 space for 200*3 pseduotime points. (1144-1171)
8. Generate 20 clusters for aligned trajectories (1173-1220)
9. Trajectory flow width (number of cells change along the trajectory) (1223-1289)
10. PCA and heatmap for post-aligned 20-cluster-trajectories (1291-1326)
11. K-means gene clustering for three trajectories (1328-1475)
12. Orthologues connecting plots for k-means gene clusters (1475-2029)
